// xont-ventura-services/xont-ventura-services.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@angular/http'), require('rxjs/add/operator/map'), require('rxjs/add/operator/catch'), require('rxjs/Rx'), require('xont-ventura-services')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/common', '@angular/http', 'rxjs/add/operator/map', 'rxjs/add/operator/catch', 'rxjs/Rx', 'xont-ventura-services'], factory) :
	(factory((global['xont-ventura-services'] = {}),global.core,global.common,global.http,null,null,global.Rx,global.xontVenturaServices));
}(this, (function (exports,core,common,http,map,_catch,Rx,xontVenturaServices) { 'use strict';

var DatetimeService$1 = (function () {
    /**
     * @param {?} http
     */
    function DatetimeService$$1(http$$1) {
        this.http = http$$1;
        this.format = '';
        this.format = localStorage.getItem('ClientDateFormat');
    }
    /**
     * @param {?} date
     * @return {?}
     */
    DatetimeService$$1.prototype.getDisplayDate = function (date) {
        var /** @type {?} */ year = date.getFullYear();
        var /** @type {?} */ yearPrefix = "";
        for (var /** @type {?} */ i = 0; i < (4 - year.toString().length); i++) {
            yearPrefix += "0";
        }
        year = yearPrefix + year;
        var /** @type {?} */ month = (1 + date.getMonth()).toString();
        month = month.length > 1 ? month : '0' + month;
        var /** @type {?} */ date = date.getDate().toString();
        date = date.length > 1 ? date : '0' + date;
        var /** @type {?} */ output = '';
        if (this.format == 'yyyy/mm/dd') {
            output = year + '/' + month + '/' + date;
        }
        else if (this.format == 'yyyy/dd/mm') {
            output = year + '/' + date + '/' + month;
        }
        else if (this.format == 'mm/yyyy/dd') {
            output = month + '/' + year + '/' + date;
        }
        else if (this.format == 'mm/dd/yyyy') {
            output = month + '/' + date + '/' + year;
        }
        else if (this.format == 'dd/yyyy/mm') {
            output = date + '/' + year + '/' + month;
        }
        else if (this.format == 'dd/mm/yyyy') {
            output = date + '/' + month + '/' + year;
        }
        else if (this.format == 'yyyy.mm.dd') {
            output = year + '.' + month + '.' + date;
        }
        else if (this.format == 'yyyy.dd.mm') {
            output = year + '.' + date + '.' + month;
        }
        else if (this.format == 'mm.yyyy.dd') {
            output = month + '.' + year + '.' + date;
        }
        else if (this.format == 'mm.dd.yyyy') {
            output = month + '.' + date + '.' + year;
        }
        else if (this.format == 'dd.yyyy.mm') {
            output = date + '.' + year + '.' + month;
        }
        else if (this.format == 'dd.mm.yyyy') {
            output = date + '.' + month + '.' + year;
        }
        else if (this.format == 'yyyy-mm-dd') {
            output = year + '-' + month + '-' + date;
        }
        else if (this.format == 'yyyy-dd-mm') {
            output = year + '-' + date + '-' + month;
        }
        else if (this.format == 'mm-yyyy-dd') {
            output = month + '-' + year + '-' + date;
        }
        else if (this.format == 'mm-dd-yyyy') {
            output = month + '-' + date + '-' + year;
        }
        else if (this.format == 'dd-yyyy-mm') {
            output = date + '-' + year + '-' + month;
        }
        else if (this.format == 'dd-mm-yyyy') {
            output = date + '-' + month + '-' + year;
        }
        else {
            alert('Invalid date format in "ventura.config.json"');
        }
        return output;
    };
    /**
     * @param {?} text
     * @return {?}
     */
    DatetimeService$$1.prototype.isValidDateString = function (text) {
        if (text == '') {
            return false;
        }
        else {
            var /** @type {?} */ year = '';
            var /** @type {?} */ month = '';
            var /** @type {?} */ date = '';
            var /** @type {?} */ array = [];
            if (this.format == 'yyyy/mm/dd') {
                array = text.split('/');
                if (array.length == 3) {
                    year = array[0];
                    month = array[1];
                    date = array[2];
                    return this.validateYearMonthDate(year, month, date);
                }
                else {
                    return false;
                }
            }
            else if (this.format == 'yyyy/dd/mm') {
                array = text.split('/');
                if (array.length == 3) {
                    year = array[0];
                    month = array[2];
                    date = array[1];
                    return this.validateYearMonthDate(year, month, date);
                }
                else {
                    return false;
                }
            }
            else if (this.format == 'mm/yyyy/dd') {
                array = text.split('/');
                if (array.length == 3) {
                    year = array[1];
                    month = array[0];
                    date = array[2];
                    return this.validateYearMonthDate(year, month, date);
                }
                else {
                    return false;
                }
            }
            else if (this.format == 'mm/dd/yyyy') {
                array = text.split('/');
                if (array.length == 3) {
                    year = array[2];
                    month = array[0];
                    date = array[1];
                    return this.validateYearMonthDate(year, month, date);
                }
                else {
                    return false;
                }
            }
            else if (this.format == 'dd/yyyy/mm') {
                array = text.split('/');
                if (array.length == 3) {
                    year = array[1];
                    month = array[2];
                    date = array[0];
                    return this.validateYearMonthDate(year, month, date);
                }
                else {
                    return false;
                }
            }
            else if (this.format == 'dd/mm/yyyy') {
                array = text.split('/');
                if (array.length == 3) {
                    year = array[2];
                    month = array[1];
                    date = array[0];
                    return this.validateYearMonthDate(year, month, date);
                }
                else {
                    return false;
                }
            }
            else if (this.format == 'yyyy.mm.dd') {
                array = text.split('.');
                if (array.length == 3) {
                    year = array[0];
                    month = array[1];
                    date = array[2];
                    return this.validateYearMonthDate(year, month, date);
                }
                else {
                    return false;
                }
            }
            else if (this.format == 'yyyy.dd.mm') {
                array = text.split('.');
                if (array.length == 3) {
                    year = array[0];
                    month = array[2];
                    date = array[1];
                    return this.validateYearMonthDate(year, month, date);
                }
                else {
                    return false;
                }
            }
            else if (this.format == 'mm.yyyy.dd') {
                array = text.split('.');
                if (array.length == 3) {
                    year = array[1];
                    month = array[0];
                    date = array[2];
                    return this.validateYearMonthDate(year, month, date);
                }
                else {
                    return false;
                }
            }
            else if (this.format == 'mm.dd.yyyy') {
                array = text.split('.');
                if (array.length == 3) {
                    year = array[2];
                    month = array[0];
                    date = array[1];
                    return this.validateYearMonthDate(year, month, date);
                }
                else {
                    return false;
                }
            }
            else if (this.format == 'dd.yyyy.mm') {
                array = text.split('.');
                if (array.length == 3) {
                    year = array[1];
                    month = array[2];
                    date = array[0];
                    return this.validateYearMonthDate(year, month, date);
                }
                else {
                    return false;
                }
            }
            else if (this.format == 'dd.mm.yyyy') {
                array = text.split('.');
                if (array.length == 3) {
                    year = array[2];
                    month = array[1];
                    date = array[0];
                    return this.validateYearMonthDate(year, month, date);
                }
                else {
                    return false;
                }
            }
            else if (this.format == 'yyyy-mm-dd') {
                array = text.split('-');
                if (array.length == 3) {
                    year = array[0];
                    month = array[1];
                    date = array[2];
                    return this.validateYearMonthDate(year, month, date);
                }
                else {
                    return false;
                }
            }
            else if (this.format == 'yyyy-dd-mm') {
                array = text.split('-');
                if (array.length == 3) {
                    year = array[0];
                    month = array[2];
                    date = array[1];
                    return this.validateYearMonthDate(year, month, date);
                }
                else {
                    return false;
                }
            }
            else if (this.format == 'mm-yyyy-dd') {
                array = text.split('-');
                if (array.length == 3) {
                    year = array[1];
                    month = array[0];
                    date = array[2];
                    return this.validateYearMonthDate(year, month, date);
                }
                else {
                    return false;
                }
            }
            else if (this.format == 'mm-dd-yyyy') {
                array = text.split('-');
                if (array.length == 3) {
                    year = array[2];
                    month = array[0];
                    date = array[1];
                    return this.validateYearMonthDate(year, month, date);
                }
                else {
                    return false;
                }
            }
            else if (this.format == 'dd-yyyy-mm') {
                array = text.split('-');
                if (array.length == 3) {
                    year = array[1];
                    month = array[2];
                    date = array[0];
                    return this.validateYearMonthDate(year, month, date);
                }
                else {
                    return false;
                }
            }
            else if (this.format == 'dd-mm-yyyy') {
                array = text.split('-');
                if (array.length == 3) {
                    year = array[2];
                    month = array[1];
                    date = array[0];
                    return this.validateYearMonthDate(year, month, date);
                }
                else {
                    return false;
                }
            }
            else {
                alert('Invalid date format in "ventura.config.json"');
            }
        }
    };
    /**
     * @param {?} year
     * @param {?} month
     * @param {?} date
     * @return {?}
     */
    DatetimeService$$1.prototype.validateYearMonthDate = function (year, month, date) {
        var /** @type {?} */ invalid = false;
        if (year.length != year.trim().length) {
            return invalid;
        }
        if (!year.match(/^\d{4}$/)) {
            return invalid;
        }
        if (parseInt(year) < 1753) {
            return invalid;
        }
        if (month.length != month.trim().length) {
            return invalid;
        }
        if (!month.match(/^\d{2}$/)) {
            return invalid;
        }
        if (parseInt(month) < 1 || parseInt(month) > 12) {
            return invalid;
        }
        if (date.length != date.trim().length) {
            return invalid;
        }
        if (!date.match(/^\d{2}$/)) {
            return invalid;
        }
        var /** @type {?} */ maxDate = new Date(parseInt(year), parseInt(month), 0).getDate();
        if (parseInt(date) < 1 || parseInt(date) > maxDate) {
            return invalid;
        }
        return true;
    };
    /**
     * @param {?} text
     * @return {?}
     */
    DatetimeService$$1.prototype.getDateTimeForString = function (text) {
        if (text.includes("T")) {
            var /** @type {?} */ fake = new Date(text);
            return new Date(fake.getTime() + (60 * 1000 * -1 * fake.getTimezoneOffset())); //apply GMT time offset.
        }
        else {
            var /** @type {?} */ year = '';
            var /** @type {?} */ month = '';
            var /** @type {?} */ date = '';
            var /** @type {?} */ array = [];
            if (this.format == 'yyyy/mm/dd') {
                array = text.split('/');
                year = array[0];
                month = array[1];
                date = array[2];
            }
            else if (this.format == 'yyyy/dd/mm') {
                array = text.split('/');
                year = array[0];
                month = array[2];
                date = array[1];
            }
            else if (this.format == 'mm/yyyy/dd') {
                array = text.split('/');
                year = array[1];
                month = array[0];
                date = array[2];
            }
            else if (this.format == 'mm/dd/yyyy') {
                array = text.split('/');
                year = array[2];
                month = array[0];
                date = array[1];
            }
            else if (this.format == 'dd/yyyy/mm') {
                array = text.split('/');
                year = array[1];
                month = array[2];
                date = array[0];
            }
            else if (this.format == 'dd/mm/yyyy') {
                array = text.split('/');
                year = array[2];
                month = array[1];
                date = array[0];
            }
            else if (this.format == 'yyyy.mm.dd') {
                array = text.split('.');
                year = array[0];
                month = array[1];
                date = array[2];
            }
            else if (this.format == 'yyyy.dd.mm') {
                array = text.split('.');
                year = array[0];
                month = array[2];
                date = array[1];
            }
            else if (this.format == 'mm.yyyy.dd') {
                array = text.split('.');
                year = array[1];
                month = array[0];
                date = array[2];
            }
            else if (this.format == 'mm.dd.yyyy') {
                array = text.split('.');
                year = array[2];
                month = array[0];
                date = array[1];
            }
            else if (this.format == 'dd.yyyy.mm') {
                array = text.split('.');
                year = array[1];
                month = array[2];
                date = array[0];
            }
            else if (this.format == 'dd.mm.yyyy') {
                array = text.split('.');
                year = array[2];
                month = array[1];
                date = array[0];
            }
            else if (this.format == 'yyyy-mm-dd') {
                array = text.split('-');
                year = array[0];
                month = array[1];
                date = array[2];
            }
            else if (this.format == 'yyyy-dd-mm') {
                array = text.split('-');
                year = array[0];
                month = array[2];
                date = array[1];
            }
            else if (this.format == 'mm-yyyy-dd') {
                array = text.split('-');
                year = array[1];
                month = array[0];
                date = array[2];
            }
            else if (this.format == 'mm-dd-yyyy') {
                array = text.split('-');
                year = array[2];
                month = array[0];
                date = array[1];
            }
            else if (this.format == 'dd-yyyy-mm') {
                array = text.split('-');
                year = array[1];
                month = array[2];
                date = array[0];
            }
            else if (this.format == 'dd-mm-yyyy') {
                array = text.split('-');
                year = array[2];
                month = array[1];
                date = array[0];
            }
            else {
                alert('Invalid date format in "ventura.config.json"');
                return null;
            }
            var /** @type {?} */ fake = new Date(parseInt(year), parseInt(month) - 1, parseInt(date));
            return new Date(fake.getTime() + (60 * 1000 * -1 * fake.getTimezoneOffset())); //apply GMT time offset.
        }
    };
    /**
     * @return {?}
     */
    DatetimeService$$1.prototype.getClientDateFormat = function () {
        return localStorage.getItem('ClientDateFormat');
    };
    /**
     * @param {?} date
     * @return {?}
     */
    DatetimeService$$1.prototype.get12HourTime = function (date) {
        var /** @type {?} */ hours = date.getHours();
        var /** @type {?} */ minutes = date.getMinutes();
        var /** @type {?} */ ampm = hours >= 12 ? 'PM' : 'AM';
        hours = hours % 12;
        hours = hours ? hours : 12; // the hour '0' should be '12'
        var /** @type {?} */ newHour = hours < 10 ? '0' + hours : hours;
        var /** @type {?} */ newMinutes = minutes < 10 ? '0' + minutes : minutes;
        var /** @type {?} */ strTime = newHour + ':' + newMinutes + ' ' + ampm;
        return strTime;
    };
    /**
     * @param {?} date
     * @return {?}
     */
    DatetimeService$$1.prototype.get24HourTime = function (date) {
        var /** @type {?} */ hours = date.getHours();
        var /** @type {?} */ minutes = date.getMinutes();
        var /** @type {?} */ strTime = hours + ':' + minutes;
        return strTime;
    };
    /**
     * @param {?} fromDate
     * @param {?} toDate
     * @return {?}
     */
    DatetimeService$$1.prototype.validateFromToDates = function (fromDate, toDate) {
        var /** @type {?} */ f = this.getDateTimeForString(fromDate);
        var /** @type {?} */ t = this.getDateTimeForString(toDate);
        if (f.getFullYear() < t.getFullYear()) {
            return true;
        }
        else if (f.getFullYear() > t.getFullYear()) {
            return false;
        }
        else {
            if (f.getMonth() < t.getMonth()) {
                return true;
            }
            else if (f.getMonth() > t.getMonth()) {
                return false;
            }
            else {
                if (f.getDate() <= t.getDate()) {
                    return true;
                }
                else {
                    return false;
                }
            }
        }
    };
    return DatetimeService$$1;
}());
DatetimeService$1.decorators = [
    { type: core.Injectable },
];
/**
 * @nocollapse
 */
DatetimeService$1.ctorParameters = function () { return [
    { type: http.Http, },
]; };

var CommonService = (function () {
    /**
     * @param {?} http
     * @param {?} location
     */
    function CommonService(http$$1, location) {
        this.http = http$$1;
        this.location = location;
    }
    /**
     * @param {?=} taskCode
     * @return {?}
     */
    CommonService.prototype.getAPIPrefix = function (taskCode) {
        //var url = this.location.prepareExternalUrl("");
        //var array = url.split('/');
        //if (array[1] == taskCode) {
        //    return "";
        //} else {
        //    return "/" + array[1];
        //}
        return this.getRootURL();
    };
    /**
     * @param {?} amount
     * @return {?}
     */
    CommonService.prototype.convertAmountToNumber = function (amount) {
        var /** @type {?} */ text = amount.trim().replace(/,/g, "");
        if (text == "") {
            return 0;
        }
        else {
            return parseFloat(text);
        }
    };
    /**
     * @param {?} num
     * @param {?} noOfMinimumDecimalPlaces
     * @return {?}
     */
    CommonService.prototype.convertNumberToAmount = function (num, noOfMinimumDecimalPlaces) {
        var /** @type {?} */ array = num.toString().split('.');
        var /** @type {?} */ prefix = '';
        var /** @type {?} */ suffix = '';
        if (array.length > 1) {
            suffix = array[1];
        }
        prefix = array[0];
        var /** @type {?} */ result = '';
        prefix = prefix.replace(/\B(?=(\d{3})+(?!\d))/g, ",");
        if (array.length > 1) {
            var /** @type {?} */ suffixLen = suffix.length;
            if (suffixLen < noOfMinimumDecimalPlaces) {
                suffix = suffix + this.getDecimalOffset(noOfMinimumDecimalPlaces - suffixLen);
            }
            result = prefix + "." + suffix;
        }
        else {
            if (noOfMinimumDecimalPlaces > 0) {
                result = prefix + "." + this.getDecimalOffset(noOfMinimumDecimalPlaces);
            }
            else {
                result = prefix;
            }
        }
        return result;
    };
    /**
     * @param {?} noOfRemainingDecimalPoints
     * @return {?}
     */
    CommonService.prototype.getDecimalOffset = function (noOfRemainingDecimalPoints) {
        var /** @type {?} */ offset = '';
        for (var /** @type {?} */ i = 0; i < noOfRemainingDecimalPoints; i++) {
            offset = offset + '0';
        }
        return offset;
    };
    /**
     * @return {?}
     */
    CommonService.prototype.isInternetExplorer = function () {
        var /** @type {?} */ ua = window.navigator.userAgent;
        var /** @type {?} */ msie = ua.indexOf('MSIE ');
        if (msie > 0) {
            // IE 10 or older 
            return true;
        }
        var /** @type {?} */ trident = ua.indexOf('Trident/');
        if (trident > 0) {
            // IE 11 
            return true;
        }
        var /** @type {?} */ edge = ua.indexOf('Edge/');
        if (edge > 0) {
            // Edge (IE 12+) 
            return true;
        }
        // other browser
        return false;
    };
    /**
     * @param {?} htmlCode
     * @param {?} filename
     * @return {?}
     */
    CommonService.prototype.generateExcel = function (htmlCode, filename) {
        //var xml = '<html xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:x="urn:schemas-microsoft-com:office:excel" xmlns="http://www.w3.org/TR/REC-html40"><head><!--[if gte mso 9]><xml><x:ExcelWorkbook><x:ExcelWorksheets><x:ExcelWorksheet><x:Name>' + filename + '</x:Name><x:WorksheetOptions><x:DisplayGridlines/></x:WorksheetOptions></x:ExcelWorksheet></x:ExcelWorksheets></x:ExcelWorkbook></xml><![endif]--></head><body>' + htmlCode + '</body></html>';//V3003
        filename += '.xls';
        if (this.isInternetExplorer()) {
            if (window.navigator.msSaveBlob) {
                var /** @type {?} */ blob = new Blob(["\ufeff", htmlCode], { type: 'text/html' });
                navigator.msSaveBlob(blob, filename);
            }
        }
        else {
            //V3003 del
            //var uri = 'data:application/vnd.ms-excel,' + encodeURI(xml);
            //var downloadLink = document.createElement("a");
            //downloadLink.href = uri;
            //downloadLink.download = filename;
            //document.body.appendChild(downloadLink);
            //downloadLink.click();
            //document.body.removeChild(downloadLink);
            //V3003
            //V3003 add
            var /** @type {?} */ myBlob = new Blob(["\ufeff", htmlCode], { type: 'text/html' });
            var /** @type {?} */ url = window.URL.createObjectURL(myBlob);
            var /** @type {?} */ a = document.createElement("a");
            document.body.appendChild(a);
            a.href = url;
            a.download = filename;
            a.click();
            setTimeout(function () { window.URL.revokeObjectURL(url); }, 0);
            //V3003
        }
    };
    /**
     * @return {?}
     */
    CommonService.prototype.getRootURL = function () {
        //let fullPath: string = window.location.href;
        //let array: any[] = fullPath.split("/" + taskCode);
        //return array[0];
        var /** @type {?} */ a = parent.document.getElementsByClassName("navbar"); //V3002
        //if (document.location.hostname == "localhost") {//V3002
        if (a.length == 0) {
            return document.location.origin;
        }
        else {
            var /** @type {?} */ array = document.location.pathname.split("/");
            return document.location.origin + "/" + array[1];
        }
    };
    /**
     * @param {?} taskCode
     * @return {?}
     */
    CommonService.prototype.getPageSize = function (taskCode) {
        var /** @type {?} */ data = JSON.parse(localStorage.getItem(taskCode + '_MasterControlData'));
        if (data.AllowPaging == "1") {
            return data.PageSize;
        }
        else {
            return data.ExtendedPageSize;
        }
    };
    /**
     * @param {?} componentName
     * @param {?} taskCode
     * @return {?}
     */
    CommonService.prototype.fieldLevelAuthentication = function (componentName, taskCode) {
        var _this = this;
        return this.http.get(this.getAPIPrefix(taskCode) + '/api/FieldLevelAuthentication/GetAuthenticationData?formName=' + componentName + '&taskCode=' + taskCode)
            .map(function (response) { return _this.restrictSpecificControls(response.json()); })
            .catch(function (error) { return Rx.Observable.throw(error.json()); });
    };
    /**
     * @param {?} list
     * @return {?}
     */
    CommonService.prototype.restrictSpecificControls = function (list) {
        var /** @type {?} */ message = "Field Level Authentication was applied for ";
        var /** @type {?} */ count = 0;
        for (var /** @type {?} */ i = 0; i < list.length; i++) {
            var /** @type {?} */ control = document.getElementById(list[i].ControlName);
            if (control) {
                if (list[i].Flag == 1) {
                    control.disabled = true;
                    $(control).find('*').attr('disabled', true);
                    count++;
                    message += list[i].ControlName;
                    if (i < (list.length - 1))
                        message += ",";
                }
                else if (list[i].Flag == 2) {
                    control.style.display = "none";
                    count++;
                    message += list[i].ControlName;
                    if (i < (list.length - 1))
                        message += ",";
                }
            }
            if (i == (list.length - 1))
                message += ".";
        }
        if (count > 0)
            console.log(message);
    };
    /**
     * @return {?}
     */
    CommonService.prototype.showLoadingIcon = function () {
        var /** @type {?} */ form = document.getElementsByTagName('body')[0];
        var /** @type {?} */ divOuter = document.createElement('div');
        divOuter.id = 'divProgressBack';
        var /** @type {?} */ divInner = document.createElement('div');
        divInner.id = 'dvProgress';
        divInner.setAttribute('class', 'loading');
        divOuter.appendChild(divInner);
        form.appendChild(divOuter);
    };
    return CommonService;
}());
CommonService.decorators = [
    { type: core.Injectable },
];
/**
 * @nocollapse
 */
CommonService.ctorParameters = function () { return [
    { type: http.Http, },
    { type: common.Location, },
]; };

var DisplayDatePipe = (function () {
    /**
     * @param {?} datetimeService
     */
    function DisplayDatePipe(datetimeService) {
        this.datetimeService = datetimeService;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    DisplayDatePipe.prototype.transform = function (value) {
        var /** @type {?} */ result = '';
        if (!value) {
            return result;
        }
        var /** @type {?} */ text = value.toString().trim();
        if (value instanceof Date) {
            if (value.toString() != 'Invalid Date') {
                result = this.datetimeService.getDisplayDate(value);
            }
        }
        else if (text != '') {
            var /** @type {?} */ d = new Date(text);
            if (d.toString() != 'Invalid Date') {
                result = this.datetimeService.getDisplayDate(d);
            }
        }
        return result;
    };
    return DisplayDatePipe;
}());
DisplayDatePipe.decorators = [
    { type: core.Pipe, args: [{
                name: 'displayDate'
            },] },
];
/**
 * @nocollapse
 */
DisplayDatePipe.ctorParameters = function () { return [
    { type: DatetimeService$1, },
]; };

var FormatMinusPipe = (function () {
    function FormatMinusPipe() {
    }
    /**
     * @param {?} value
     * @return {?}
     */
    FormatMinusPipe.prototype.transform = function (value) {
        //let result: string = '';
        if (value == null) {
            return "";
        }
        var /** @type {?} */ trimVal = value.toString().trim();
        if (trimVal == "") {
            return "";
        }
        //if (/^[+-]?\d+(\.\d+)?$/.test(value.toString().trim())) {
        //    let num: number = parseFloat(value.toString().trim());
        //    if (num < 0) {
        //        let positiveNum: number = (-1 * num);
        //        return "(" + positiveNum.toString() + ")";
        //    } else {
        //        return num.toString();
        //    }
        //} else {
        //    return value.toString();
        //}
        if (trimVal.charAt(0) == "-") {
            return "(" + trimVal.substring(1) + ")";
        }
        return value.toString();
    };
    return FormatMinusPipe;
}());
FormatMinusPipe.decorators = [
    { type: core.Pipe, args: [{
                name: 'formatMinus'
            },] },
];
/**
 * @nocollapse
 */
FormatMinusPipe.ctorParameters = function () { return []; };

var BlockOnPasteDirective = (function () {
    function BlockOnPasteDirective() {
    }
    /**
     * @param {?} e
     * @return {?}
     */
    BlockOnPasteDirective.prototype.blockPaste = function (e) {
        if (this.value == true) {
            e.preventDefault();
        }
    };
    return BlockOnPasteDirective;
}());
BlockOnPasteDirective.decorators = [
    { type: core.Directive, args: [{
                selector: '[blockOnPaste]'
            },] },
];
/**
 * @nocollapse
 */
BlockOnPasteDirective.ctorParameters = function () { return []; };
BlockOnPasteDirective.propDecorators = {
    'value': [{ type: core.Input, args: ["blockOnPaste",] },],
    'blockPaste': [{ type: core.HostListener, args: ['paste', ['$event'],] },],
};

var LightSearchPipe = (function () {
    /**
     * @param {?} datetimeService
     */
    function LightSearchPipe(datetimeService) {
        this.datetimeService = datetimeService;
    }
    /**
     * @param {?} table
     * @param {?} keyword
     * @param {?} fields
     * @return {?}
     */
    LightSearchPipe.prototype.transform = function (table, keyword, fields) {
        var _this = this;
        keyword = keyword.trim().toUpperCase();
        var /** @type {?} */ columns = [];
        var /** @type {?} */ patt = /(\d{4})-(\d{2})-(\d{2})/; //yyyy-mm-dd
        if (table && keyword != "") {
            if (fields) {
                for (var /** @type {?} */ i = 0; i < fields.length; i++) {
                    if (table[0][fields[i]] != undefined && table[0][fields[i]] != null) {
                        if (patt.test(table[0][fields[i]].toString())) {
                            columns.push({ Name: fields[i], DateField: true });
                        }
                        else {
                            columns.push({ Name: fields[i], DateField: false });
                        }
                    }
                    else {
                        columns.push({ Name: fields[i], DateField: false });
                    }
                }
                //columns = fields;
            }
            else {
                for (var /** @type {?} */ key in table[0]) {
                    if (table[0][key] != undefined && table[0][key] != null) {
                        if (patt.test(table[0][key].toString())) {
                            columns.push({ Name: key, DateField: true });
                        }
                        else {
                            columns.push({ Name: key, DateField: false });
                        }
                    }
                    else {
                        columns.push({ Name: key, DateField: false });
                    }
                }
            }
            //console.log(columns);
            var /** @type {?} */ filterRowCount = 0;
            for (var /** @type {?} */ j = 0; j < columns.length; j++) {
                var /** @type {?} */ tempTable = [];
                if (columns[j].DateField) {
                    tempTable = table.filter(function (a) { return _this.datetimeService.getDisplayDate(new Date(a[columns[j].Name])).toString().includes(keyword); });
                }
                else {
                    tempTable = table.filter(function (a) { return a[columns[j].Name].toString().toUpperCase().includes(keyword); });
                }
                filterRowCount += tempTable.length;
                if (tempTable.length > 0) {
                    table = tempTable;
                }
            }
            if (filterRowCount == 0) {
                table = [];
            }
        }
        return table;
    };
    return LightSearchPipe;
}());
LightSearchPipe.decorators = [
    { type: core.Pipe, args: [{
                name: 'lightSearch'
            },] },
];
/**
 * @nocollapse
 */
LightSearchPipe.ctorParameters = function () { return [
    { type: xontVenturaServices.DatetimeService, },
]; };

var XontVenturaServicesModule = (function () {
    function XontVenturaServicesModule() {
    }
    /**
     * @return {?}
     */
    XontVenturaServicesModule.forRoot = function () {
        return {
            ngModule: XontVenturaServicesModule,
            providers: [DatetimeService$1, CommonService]
        };
    };
    return XontVenturaServicesModule;
}());
XontVenturaServicesModule.decorators = [
    { type: core.NgModule, args: [{
                imports: [
                    common.CommonModule
                ],
                declarations: [
                    DisplayDatePipe, FormatMinusPipe, BlockOnPasteDirective,
                    LightSearchPipe //V3001
                ],
                exports: [
                    DisplayDatePipe, FormatMinusPipe, BlockOnPasteDirective
                ]
            },] },
];
/**
 * @nocollapse
 */
XontVenturaServicesModule.ctorParameters = function () { return []; };

exports.XontVenturaServicesModule = XontVenturaServicesModule;
exports.DatetimeService = DatetimeService$1;
exports.CommonService = CommonService;
exports.DisplayDatePipe = DisplayDatePipe;
exports.FormatMinusPipe = FormatMinusPipe;
exports.BlockOnPasteDirective = BlockOnPasteDirective;
exports.LightSearchPipe = LightSearchPipe;

Object.defineProperty(exports, '__esModule', { value: true });

})));
