// xont-ventura-validators/index.js
import { Attribute, Directive, ElementRef, EventEmitter, HostListener, Injectable, Input, NgModule, Output, Pipe, forwardRef } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NG_VALIDATORS, NG_VALUE_ACCESSOR } from '@angular/forms';
import { Http } from '@angular/http';
import { DatetimeService } from 'xont-ventura-services';

var ValidateRangeDirective = (function () {
    /**
     * @param {?} validateRange
     */
    function ValidateRangeDirective(validateRange) {
        this.validateRange = validateRange;
    }
    /**
     * @param {?} c
     * @return {?}
     */
    ValidateRangeDirective.prototype.validate = function (c) {
        if (c.value) {
            var /** @type {?} */ self_1 = Number(c.value.trim());
            var /** @type {?} */ array = this.validateRange.split(',');
            var /** @type {?} */ min = Number(array[0].trim());
            var /** @type {?} */ max = Number(array[1].trim());
            //console.log(self,1);
            if (isNaN(c.value.trim())) {
                //console.log(self);
                return null;
            }
            if (min <= self_1 && self_1 <= max) {
                return null;
            }
            return { validateRange: { valid: false } };
        }
    };
    return ValidateRangeDirective;
}());
ValidateRangeDirective.decorators = [
    { type: Directive, args: [{
                selector: '[validateRange][formControlName],[validateRange][formControl],[validateRange][ngModel]',
                providers: [
                    { provide: NG_VALIDATORS, useExisting: forwardRef(function () { return ValidateRangeDirective; }), multi: true }
                ]
            },] },
];
/**
 * @nocollapse
 */
ValidateRangeDirective.ctorParameters = function () { return [
    { type: undefined, decorators: [{ type: Attribute, args: ['validateRange',] },] },
]; };

var ValidateDateDirective = (function () {
    /**
     * @param {?} http
     */
    function ValidateDateDirective(http$$1) {
        this.http = http$$1;
        this.format = '';
        this.format = localStorage.getItem('ClientDateFormat');
        this.validator = this.validateDateFactory();
    }
    /**
     * @param {?} c
     * @return {?}
     */
    ValidateDateDirective.prototype.validate = function (c) {
        return this.validator(c);
    };
    /**
     * @return {?}
     */
    ValidateDateDirective.prototype.validateDateFactory = function () {
        var _this = this;
        return function (c) {
            var /** @type {?} */ text = String(c.value).trim();
            if (text == '') {
                return null;
            }
            else {
                var /** @type {?} */ year = '';
                var /** @type {?} */ month = '';
                var /** @type {?} */ date = '';
                var /** @type {?} */ array = [];
                if (_this.format == 'yyyy/mm/dd') {
                    array = text.split('/');
                    if (array.length == 3) {
                        year = array[0];
                        month = array[1];
                        date = array[2];
                        return _this.validateYearMonthDate(year, month, date);
                    }
                    else {
                        return { validateDate: { valid: false } };
                    }
                }
                else if (_this.format == 'yyyy/dd/mm') {
                    array = text.split('/');
                    if (array.length == 3) {
                        year = array[0];
                        month = array[2];
                        date = array[1];
                        return _this.validateYearMonthDate(year, month, date);
                    }
                    else {
                        return { validateDate: { valid: false } };
                    }
                }
                else if (_this.format == 'mm/yyyy/dd') {
                    array = text.split('/');
                    if (array.length == 3) {
                        year = array[1];
                        month = array[0];
                        date = array[2];
                        return _this.validateYearMonthDate(year, month, date);
                    }
                    else {
                        return { validateDate: { valid: false } };
                    }
                }
                else if (_this.format == 'mm/dd/yyyy') {
                    array = text.split('/');
                    if (array.length == 3) {
                        year = array[2];
                        month = array[0];
                        date = array[1];
                        return _this.validateYearMonthDate(year, month, date);
                    }
                    else {
                        return { validateDate: { valid: false } };
                    }
                }
                else if (_this.format == 'dd/yyyy/mm') {
                    array = text.split('/');
                    if (array.length == 3) {
                        year = array[1];
                        month = array[2];
                        date = array[0];
                        return _this.validateYearMonthDate(year, month, date);
                    }
                    else {
                        return { validateDate: { valid: false } };
                    }
                }
                else if (_this.format == 'dd/mm/yyyy') {
                    array = text.split('/');
                    if (array.length == 3) {
                        year = array[2];
                        month = array[1];
                        date = array[0];
                        return _this.validateYearMonthDate(year, month, date);
                    }
                    else {
                        return { validateDate: { valid: false } };
                    }
                }
                else if (_this.format == 'yyyy.mm.dd') {
                    array = text.split('.');
                    if (array.length == 3) {
                        year = array[0];
                        month = array[1];
                        date = array[2];
                        return _this.validateYearMonthDate(year, month, date);
                    }
                    else {
                        return { validateDate: { valid: false } };
                    }
                }
                else if (_this.format == 'yyyy.dd.mm') {
                    array = text.split('.');
                    if (array.length == 3) {
                        year = array[0];
                        month = array[2];
                        date = array[1];
                        return _this.validateYearMonthDate(year, month, date);
                    }
                    else {
                        return { validateDate: { valid: false } };
                    }
                }
                else if (_this.format == 'mm.yyyy.dd') {
                    array = text.split('.');
                    if (array.length == 3) {
                        year = array[1];
                        month = array[0];
                        date = array[2];
                        return _this.validateYearMonthDate(year, month, date);
                    }
                    else {
                        return { validateDate: { valid: false } };
                    }
                }
                else if (_this.format == 'mm.dd.yyyy') {
                    array = text.split('.');
                    if (array.length == 3) {
                        year = array[2];
                        month = array[0];
                        date = array[1];
                        return _this.validateYearMonthDate(year, month, date);
                    }
                    else {
                        return { validateDate: { valid: false } };
                    }
                }
                else if (_this.format == 'dd.yyyy.mm') {
                    array = text.split('.');
                    if (array.length == 3) {
                        year = array[1];
                        month = array[2];
                        date = array[0];
                        return _this.validateYearMonthDate(year, month, date);
                    }
                    else {
                        return { validateDate: { valid: false } };
                    }
                }
                else if (_this.format == 'dd.mm.yyyy') {
                    array = text.split('.');
                    if (array.length == 3) {
                        year = array[2];
                        month = array[1];
                        date = array[0];
                        return _this.validateYearMonthDate(year, month, date);
                    }
                    else {
                        return { validateDate: { valid: false } };
                    }
                }
                else if (_this.format == 'yyyy-mm-dd') {
                    array = text.split('-');
                    if (array.length == 3) {
                        year = array[0];
                        month = array[1];
                        date = array[2];
                        return _this.validateYearMonthDate(year, month, date);
                    }
                    else {
                        return { validateDate: { valid: false } };
                    }
                }
                else if (_this.format == 'yyyy-dd-mm') {
                    array = text.split('-');
                    if (array.length == 3) {
                        year = array[0];
                        month = array[2];
                        date = array[1];
                        return _this.validateYearMonthDate(year, month, date);
                    }
                    else {
                        return { validateDate: { valid: false } };
                    }
                }
                else if (_this.format == 'mm-yyyy-dd') {
                    array = text.split('-');
                    if (array.length == 3) {
                        year = array[1];
                        month = array[0];
                        date = array[2];
                        return _this.validateYearMonthDate(year, month, date);
                    }
                    else {
                        return { validateDate: { valid: false } };
                    }
                }
                else if (_this.format == 'mm-dd-yyyy') {
                    array = text.split('-');
                    if (array.length == 3) {
                        year = array[2];
                        month = array[0];
                        date = array[1];
                        return _this.validateYearMonthDate(year, month, date);
                    }
                    else {
                        return { validateDate: { valid: false } };
                    }
                }
                else if (_this.format == 'dd-yyyy-mm') {
                    array = text.split('-');
                    if (array.length == 3) {
                        year = array[1];
                        month = array[2];
                        date = array[0];
                        return _this.validateYearMonthDate(year, month, date);
                    }
                    else {
                        return { validateDate: { valid: false } };
                    }
                }
                else if (_this.format == 'dd-mm-yyyy') {
                    array = text.split('-');
                    if (array.length == 3) {
                        year = array[2];
                        month = array[1];
                        date = array[0];
                        return _this.validateYearMonthDate(year, month, date);
                    }
                    else {
                        return { validateDate: { valid: false } };
                    }
                }
                else {
                    alert('Invalid date format in "ventura.config.json"');
                }
            }
            return { validateDate: { valid: false } };
        };
    };
    /**
     * @param {?} year
     * @param {?} month
     * @param {?} date
     * @return {?}
     */
    ValidateDateDirective.prototype.validateYearMonthDate = function (year, month, date) {
        var /** @type {?} */ invalid = { validateDate: { valid: false } };
        if (year.length != year.trim().length) {
            return invalid;
        }
        if (!year.match(/^\d{4}$/)) {
            return invalid;
        }
        if (parseInt(year) < 1753) {
            return invalid;
        }
        if (month.length != month.trim().length) {
            return invalid;
        }
        if (!month.match(/^\d{2}$/)) {
            return invalid;
        }
        if (parseInt(month) < 1 || parseInt(month) > 12) {
            return invalid;
        }
        if (date.length != date.trim().length) {
            return invalid;
        }
        if (!date.match(/^\d{2}$/)) {
            return invalid;
        }
        var /** @type {?} */ maxDate = new Date(parseInt(year), parseInt(month), 0).getDate();
        if (parseInt(date) < 1 || parseInt(date) > maxDate) {
            return invalid;
        }
        return null;
    };
    return ValidateDateDirective;
}());
ValidateDateDirective.decorators = [
    { type: Directive, args: [{
                selector: '[validateDate][formControlName],[validateDate][formControl],[validateDate][ngModel]',
                providers: [
                    { provide: NG_VALIDATORS, useExisting: forwardRef(function () { return ValidateDateDirective; }), multi: true }
                ]
            },] },
];
/**
 * @nocollapse
 */
ValidateDateDirective.ctorParameters = function () { return [
    { type: Http, },
]; };

var ValidatorService = (function () {
    function ValidatorService() {
        this.noOfRoundableDigits = 11;
        this.noOfDecimalDigitsForValue = 2;
    }
    Object.defineProperty(ValidatorService.prototype, "noOfDecimalDigitsForQuantity", {
        /**
         * @return {?}
         */
        get: function () {
            if (localStorage.getItem("ControlData") != null) {
                var /** @type {?} */ data = JSON.parse(localStorage.getItem("ControlData"));
                return data.DecimalPlaces;
            }
            else {
                return 0;
            }
        },
        enumerable: true,
        configurable: true
    });
    
    /**
     * @param {?} evt
     * @param {?} noOfRoundableDigits
     * @param {?} noOfDecimalDigits
     * @param {?} withMinus
     * @return {?}
     */
    ValidatorService.prototype.numbersOnly = function (evt, noOfRoundableDigits, noOfDecimalDigits, withMinus) {
        var /** @type {?} */ iKeyCode = (evt.which) ? evt.which : evt.keyCode;
        var /** @type {?} */ text = evt.target.value.trim();
        var /** @type {?} */ originalText = evt.target.value;
        var /** @type {?} */ cursorStart = evt.target.selectionStart;
        var /** @type {?} */ cursorEnd = evt.target.selectionEnd;
        //V3003
        //if (originalText != "") {
        //    if (cursorStart == 0 && cursorEnd == 0) {
        //        return false;
        //    } else if (cursorStart != cursorEnd) {
        //        if (!(cursorStart == 0 && cursorEnd == originalText.length)) {
        //            return false;
        //        }
        //    } else if (!(cursorStart == originalText.length && cursorEnd == originalText.length)) {
        //        return false;
        //    }
        //}
        //V3003
        if (withMinus == true) {
            if (iKeyCode != 45 && iKeyCode != 46 && iKeyCode > 31 && (iKeyCode < 48 || iKeyCode > 57)) {
                return false;
            }
        }
        else {
            if (iKeyCode != 46 && iKeyCode > 31 && (iKeyCode < 48 || iKeyCode > 57)) {
                return false;
            }
        }
        //V3003
        //if (iKeyCode == 45) {
        //    if (text != "") {
        //        return false;//block - 
        //    }
        //}
        //V3003
        //V3003
        if (iKeyCode == 45) {
            if (text != "") {
                if (text.includes("-")) {
                    return false;
                }
                if (cursorStart != 0) {
                    return false;
                }
                else {
                    var /** @type {?} */ dotIndex = text.indexOf('.');
                    if (cursorEnd == dotIndex) {
                        return false;
                    }
                }
            }
        }
        else if (iKeyCode == 46) {
            if (text == "") {
                return false; //block . for first digit
            }
            else if (text.includes(".")) {
                return false; //block . if it already exists
            }
            else if (noOfDecimalDigits == 0) {
                return false;
            }
        }
        else {
            if (text == "") {
                if (noOfRoundableDigits <= 0) {
                    return false;
                }
            }
            else {
                var /** @type {?} */ textWithoutComma = text.replace(/,/g, "");
                textWithoutComma = textWithoutComma.replace(/-/g, "");
                if (cursorStart == cursorEnd) {
                    if (text.includes(".")) {
                        var /** @type {?} */ array = textWithoutComma.split('.');
                        var /** @type {?} */ dotIndex = text.indexOf('.');
                        if (cursorEnd > dotIndex) {
                            if (array[1].length >= noOfDecimalDigits) {
                                return false;
                            }
                        }
                        else {
                            if (array[0].length >= noOfRoundableDigits) {
                                return false;
                            }
                        }
                    }
                    else {
                        if (textWithoutComma.length >= noOfRoundableDigits) {
                            return false;
                        }
                    }
                }
                else {
                    if (text.includes(".")) {
                        var /** @type {?} */ dotIndex = text.indexOf('.');
                        if (cursorStart <= dotIndex && dotIndex < cursorEnd) {
                            var /** @type {?} */ array = textWithoutComma.split('.');
                            var /** @type {?} */ selectedDecimalCount = cursorEnd - (dotIndex + 1);
                            var /** @type {?} */ remainDecimalCount = array[1].length - selectedDecimalCount;
                            var /** @type {?} */ selectedRoundableCount = dotIndex - cursorStart;
                            var /** @type {?} */ remainRoundableCount = array[0].length - selectedRoundableCount;
                            if ((remainRoundableCount + remainDecimalCount) >= noOfRoundableDigits) {
                                return false;
                            }
                        }
                    }
                    if (text.includes("-")) {
                        if (cursorStart == 0 && cursorEnd == 1) {
                            var /** @type {?} */ array = textWithoutComma.split('.');
                            if (array[0].length >= noOfRoundableDigits) {
                                return false;
                            }
                        }
                    }
                }
            }
        }
        //V3003
        //V3003
        //else {
        //    if (noOfRoundableDigits == 0) {
        //        return false;
        //    }
        //    else if (noOfRoundableDigits > 0) {
        //        if (text.includes(".")) {
        //            var array = text.split('.');
        //            if (array[1].length >= noOfDecimalDigits && noOfDecimalDigits >= 0) {
        //                if (!(cursorStart == 0 && cursorEnd == originalText.length)) {
        //                    return false;
        //                }
        //            }
        //        } else {
        //            var textWithoutComma = text.replace(/,/g, "");
        //            textWithoutComma = textWithoutComma.replace(/-/g, "");
        //            if (textWithoutComma.length >= noOfRoundableDigits) {
        //                return false;
        //            }
        //        }
        //    }
        //    else {
        //        if (text.includes(".")) {
        //            var array = text.split('.');
        //            if (array[1].length >= noOfDecimalDigits && noOfDecimalDigits >= 0) {
        //                if (!(cursorStart == 0 && cursorEnd == originalText.length)) {
        //                    return false;
        //                }
        //            }
        //        }
        //    }
        //}
        //V3003
        return true;
    };
    /**
     * @param {?} val
     * @return {?}
     */
    ValidatorService.prototype.applyValueFormat = function (val) {
        var /** @type {?} */ text = val.trim().replace(/,/g, "");
        var /** @type {?} */ array = text.split('.');
        var /** @type {?} */ prefix = '';
        var /** @type {?} */ suffix = '';
        if (array.length > 1) {
            suffix = array[1];
        }
        prefix = array[0];
        if (array.length > 1) {
            prefix = prefix.replace(/\B(?=(\d{3})+(?!\d))/g, ",");
            return prefix + '.' + suffix;
        }
        else {
            prefix = prefix.replace(/\B(?=(\d{3})+(?!\d))/g, ",");
            return prefix;
        }
    };
    /**
     * @param {?} value
     * @param {?} noOfMinimumDecimalDigits
     * @return {?}
     */
    ValidatorService.prototype.completeDecimalPlaces = function (value, noOfMinimumDecimalDigits) {
        var /** @type {?} */ text = this.getNumberValue(value.trim(), noOfMinimumDecimalDigits).toFixed(noOfMinimumDecimalDigits);
        if (text.includes(".")) {
            var /** @type {?} */ array = text.split(".");
            return array[0].replace(/\B(?=(\d{3})+(?!\d))/g, ",") + "." + array[1];
        }
        else {
            return text.replace(/\B(?=(\d{3})+(?!\d))/g, ",");
        }
    };
    /**
     * @param {?} value
     * @param {?} noOfDecimalDigits
     * @return {?}
     */
    ValidatorService.prototype.getNumberValue = function (value, noOfDecimalDigits) {
        if (value != undefined || value != null) {
            if (value.toString().trim() == "") {
                return 0;
            }
            else {
                if (isNaN(parseFloat(value.toString().trim().replace(/,/g, "")))) {
                    return 0;
                }
                else {
                    return parseFloat(parseFloat(value.toString().trim().replace(/,/g, "")).toFixed(noOfDecimalDigits));
                }
            }
        }
        else {
            return 0;
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    ValidatorService.prototype.getFinalValueForStringNumber = function (value) {
        if (value != undefined || value != null) {
            if (value.toString().trim() == "") {
                return "";
            }
            else {
                if (isNaN(parseFloat(value.toString().trim()))) {
                    return "";
                }
                else {
                    return parseFloat(value.toString().trim()).toString();
                }
            }
        }
        else {
            return "";
        }
    };
    /**
     * @param {?} value
     * @param {?} pattern
     * @return {?}
     */
    ValidatorService.prototype.isValidPattern = function (value, pattern) {
        return pattern.test(value);
    };
    /**
     * @param {?} evt
     * @param {?} pattern
     * @return {?}
     */
    ValidatorService.prototype.allowCharactersByPattern = function (evt, pattern) {
        return pattern.test(this.getTextAfterKeyPress(evt));
    };
    /**
     * @param {?} evt
     * @return {?}
     */
    ValidatorService.prototype.getTextAfterKeyPress = function (evt) {
        var /** @type {?} */ part1 = "";
        var /** @type {?} */ part2 = "";
        var /** @type {?} */ currentText = evt.target.value;
        var /** @type {?} */ iKeyCode = (evt.which) ? evt.which : evt.keyCode;
        var /** @type {?} */ nextCharater = String.fromCharCode(iKeyCode);
        var /** @type {?} */ cursorStart = evt.selectionStart;
        var /** @type {?} */ cursorEnd = evt.selectionEnd;
        if (cursorStart == cursorEnd) {
            part1 = currentText.substring(0, cursorStart);
            part2 = currentText.substring(cursorStart, currentText.length);
        }
        else {
            part1 = currentText.substring(0, cursorStart);
            part2 = currentText.substring(cursorEnd, currentText.length);
        }
        return part1 + nextCharater + part2;
    };
    /**
     * @param {?} nativeElement
     * @return {?}
     */
    ValidatorService.prototype.indicateInvalidKeyPress = function (nativeElement) {
        $(nativeElement).css('outline-style', 'none');
        $(nativeElement).css('border-color', 'red');
        setTimeout(function () {
            $(nativeElement).css('outline-style', '');
            $(nativeElement).css('border-color', '');
        }, 500);
    };
    /**
     * @param {?} nativeElement
     * @return {?}
     */
    ValidatorService.prototype.indicateInvalidInput = function (nativeElement) {
        $(nativeElement).css('color', 'red');
        setTimeout(function () {
            $(nativeElement).css('color', '');
        }, 2000);
    };
    return ValidatorService;
}());
ValidatorService.decorators = [
    { type: Injectable },
];
/**
 * @nocollapse
 */
ValidatorService.ctorParameters = function () { return []; };

var XontValuePipe = (function () {
    /**
     * @param {?} validatorService
     */
    function XontValuePipe(validatorService) {
        this.validatorService = validatorService;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    XontValuePipe.prototype.transform = function (value) {
        if (value == undefined || value == null) {
            return (0).toFixed(this.validatorService.noOfDecimalDigitsForValue);
        }
        if (value.toString().trim() == "") {
            return (0).toFixed(this.validatorService.noOfDecimalDigitsForValue);
        }
        else {
            var /** @type {?} */ decimalValue = parseFloat(value.toString().trim()).toFixed(this.validatorService.noOfDecimalDigitsForValue);
            return this.validatorService.applyValueFormat(decimalValue);
        }
    };
    return XontValuePipe;
}());
XontValuePipe.decorators = [
    { type: Pipe, args: [{
                name: 'xont_value'
            },] },
];
/**
 * @nocollapse
 */
XontValuePipe.ctorParameters = function () { return [
    { type: ValidatorService, },
]; };

var CUSTOM_INPUT_CONTROL_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(function () { return XontValueDirective; }),
    multi: true
};
var XontValueDirective = (function () {
    /**
     * @param {?} elementRef
     * @param {?} valuePipe
     * @param {?} validatorService
     */
    function XontValueDirective(elementRef, valuePipe, validatorService) {
        this.elementRef = elementRef;
        this.valuePipe = valuePipe;
        this.validatorService = validatorService;
        this.ngModelChange = new EventEmitter();
        this.el = this.elementRef.nativeElement;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    XontValueDirective.prototype.writeValue = function (value) {
        if (value != undefined && value != null) {
            value = value.toString().replace(/,/g, "");
            if (this.el.value.toString().replace(/,/g, "") != value) {
                this.el.value = this.valuePipe.transform(value);
                this.ngModelChange.emit(this.validatorService.getNumberValue(value, this.validatorService.noOfDecimalDigitsForValue));
            }
        }
        else {
            //this.el.value =this.valuePipe.transform(value);
            //this.ngModelChange.emit(this.validatorService.getNumberValue(value,this.validatorService.noOfDecimalDigitsForValue));
        }
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    XontValueDirective.prototype.registerOnChange = function (fn) {
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    XontValueDirective.prototype.registerOnTouched = function (fn) {
    };
    /**
     * @param {?} value
     * @return {?}
     */
    XontValueDirective.prototype.onBlur = function (value) {
        this.el.value = this.validatorService.completeDecimalPlaces(value, this.validatorService.noOfDecimalDigitsForValue);
        this.ngModelChange.emit(this.validatorService.getNumberValue(value, this.validatorService.noOfDecimalDigitsForValue));
    };
    /**
     * @param {?} value
     * @return {?}
     */
    XontValueDirective.prototype.onFocus = function (value) {
        if (this.validatorService.getNumberValue(value, this.validatorService.noOfDecimalDigitsForValue) == 0) {
            this.el.value = "";
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    XontValueDirective.prototype.onKeyup = function (value) {
        this.el.value = this.validatorService.applyValueFormat(value);
        this.ngModelChange.emit(this.validatorService.getNumberValue(value, this.validatorService.noOfDecimalDigitsForValue));
    };
    /**
     * @param {?} evt
     * @return {?}
     */
    XontValueDirective.prototype.onKeypress = function (evt) {
        var /** @type {?} */ result = this.validatorService.numbersOnly(evt, this.validatorService.noOfRoundableDigits, this.validatorService.noOfDecimalDigitsForValue, false); //validate Price
        if (!result) {
            this.validatorService.indicateInvalidKeyPress(this.el);
        }
        return result;
    };
    /**
     * @return {?}
     */
    XontValueDirective.prototype.onDrop = function () {
        var _this = this;
        setTimeout(function () {
            if (isNaN(parseFloat(_this.el.value.replace(/,/g, "")))) {
                _this.validatorService.indicateInvalidInput(_this.el);
            }
        }, 10);
    };
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    XontValueDirective.prototype.setDisabledState = function (isDisabled) {
        this.elementRef.nativeElement.disabled = isDisabled;
    };
    return XontValueDirective;
}());
XontValueDirective.decorators = [
    { type: Directive, args: [{
                selector: "[xont_value]",
                providers: [CUSTOM_INPUT_CONTROL_VALUE_ACCESSOR]
            },] },
];
/**
 * @nocollapse
 */
XontValueDirective.ctorParameters = function () { return [
    { type: ElementRef, },
    { type: XontValuePipe, },
    { type: ValidatorService, },
]; };
XontValueDirective.propDecorators = {
    'ngModelChange': [{ type: Output },],
    'onBlur': [{ type: HostListener, args: ["blur", ["$event.target.value"],] },],
    'onFocus': [{ type: HostListener, args: ["focus", ["$event.target.value"],] },],
    'onKeyup': [{ type: HostListener, args: ["keyup", ["$event.target.value"],] },],
    'onKeypress': [{ type: HostListener, args: ["keypress", ["$event"],] },],
    'onDrop': [{ type: HostListener, args: ["drop",] },],
};

var XontQuantityPipe = (function () {
    /**
     * @param {?} validatorService
     */
    function XontQuantityPipe(validatorService) {
        this.validatorService = validatorService;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    XontQuantityPipe.prototype.transform = function (value) {
        if (value == undefined || value == null) {
            return (0).toFixed(this.validatorService.noOfDecimalDigitsForQuantity);
        }
        if (value.toString().trim() == "") {
            return (0).toFixed(this.validatorService.noOfDecimalDigitsForQuantity);
        }
        else {
            var /** @type {?} */ decimalValue = parseFloat(value.toString().trim()).toFixed(this.validatorService.noOfDecimalDigitsForQuantity);
            return this.validatorService.applyValueFormat(decimalValue);
        }
    };
    return XontQuantityPipe;
}());
XontQuantityPipe.decorators = [
    { type: Pipe, args: [{
                name: 'xont_quantity'
            },] },
];
/**
 * @nocollapse
 */
XontQuantityPipe.ctorParameters = function () { return [
    { type: ValidatorService, },
]; };

var CUSTOM_INPUT_CONTROL_QUANTITY_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(function () { return XontQuantityDirective; }),
    multi: true
};
var XontQuantityDirective = (function () {
    /**
     * @param {?} elementRef
     * @param {?} quantityPipe
     * @param {?} validatorService
     */
    function XontQuantityDirective(elementRef, quantityPipe, validatorService) {
        this.elementRef = elementRef;
        this.quantityPipe = quantityPipe;
        this.validatorService = validatorService;
        this.ngModelChange = new EventEmitter();
        this.el = this.elementRef.nativeElement;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    XontQuantityDirective.prototype.writeValue = function (value) {
        if (value != undefined && value != null) {
            value = value.toString().replace(/,/g, "");
            if (this.el.value.toString().replace(/,/g, "") != value) {
                this.el.value = this.quantityPipe.transform(value);
                this.ngModelChange.emit(this.validatorService.getNumberValue(value, this.validatorService.noOfDecimalDigitsForQuantity));
            }
        }
        else {
            //this.el.value =this.quantityPipe.transform(value);
            //this.ngModelChange.emit(this.validatorService.getNumberValue(value,this.validatorService.noOfDecimalDigitsForQuantity));
        }
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    XontQuantityDirective.prototype.registerOnChange = function (fn) {
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    XontQuantityDirective.prototype.registerOnTouched = function (fn) {
    };
    /**
     * @param {?} value
     * @return {?}
     */
    XontQuantityDirective.prototype.onBlur = function (value) {
        this.el.value = this.validatorService.completeDecimalPlaces(value, this.validatorService.noOfDecimalDigitsForQuantity);
        this.ngModelChange.emit(this.validatorService.getNumberValue(value, this.validatorService.noOfDecimalDigitsForQuantity));
    };
    /**
     * @param {?} value
     * @return {?}
     */
    XontQuantityDirective.prototype.onFocus = function (value) {
        if (this.validatorService.getNumberValue(value, this.validatorService.noOfDecimalDigitsForQuantity) == 0) {
            this.el.value = "";
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    XontQuantityDirective.prototype.onKeyup = function (value) {
        this.el.value = this.validatorService.applyValueFormat(value);
        this.ngModelChange.emit(this.validatorService.getNumberValue(value, this.validatorService.noOfDecimalDigitsForQuantity));
    };
    /**
     * @param {?} evt
     * @return {?}
     */
    XontQuantityDirective.prototype.onKeypress = function (evt) {
        var /** @type {?} */ result = this.validatorService.numbersOnly(evt, this.validatorService.noOfRoundableDigits, this.validatorService.noOfDecimalDigitsForQuantity, false);
        if (!result) {
            this.validatorService.indicateInvalidKeyPress(this.el);
        }
        return result;
    };
    /**
     * @return {?}
     */
    XontQuantityDirective.prototype.onDrop = function () {
        var _this = this;
        setTimeout(function () {
            if (isNaN(parseFloat(_this.el.value.replace(/,/g, "")))) {
                _this.validatorService.indicateInvalidInput(_this.el);
            }
        }, 10);
    };
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    XontQuantityDirective.prototype.setDisabledState = function (isDisabled) {
        this.elementRef.nativeElement.disabled = isDisabled;
    };
    return XontQuantityDirective;
}());
XontQuantityDirective.decorators = [
    { type: Directive, args: [{ selector: "[xont_quantity]", providers: [CUSTOM_INPUT_CONTROL_QUANTITY_ACCESSOR] },] },
];
/**
 * @nocollapse
 */
XontQuantityDirective.ctorParameters = function () { return [
    { type: ElementRef, },
    { type: XontQuantityPipe, },
    { type: ValidatorService, },
]; };
XontQuantityDirective.propDecorators = {
    'ngModelChange': [{ type: Output },],
    'onBlur': [{ type: HostListener, args: ["blur", ["$event.target.value"],] },],
    'onFocus': [{ type: HostListener, args: ["focus", ["$event.target.value"],] },],
    'onKeyup': [{ type: HostListener, args: ["keyup", ["$event.target.value"],] },],
    'onKeypress': [{ type: HostListener, args: ["keypress", ["$event"],] },],
    'onDrop': [{ type: HostListener, args: ["drop",] },],
};

var XontCodeDirective = (function () {
    /**
     * @param {?} elementRef
     * @param {?} validatorService
     */
    function XontCodeDirective(elementRef, validatorService) {
        this.elementRef = elementRef;
        this.validatorService = validatorService;
        this.ngModelChange = new EventEmitter();
        this.codeRegx = new RegExp("^[A-Za-z0-9_-]*$");
        this.el = this.elementRef.nativeElement;
    }
    /**
     * @param {?} evt
     * @return {?}
     */
    XontCodeDirective.prototype.onKeypress = function (evt) {
        var /** @type {?} */ iKeyCode = (evt.which) ? evt.which : evt.keyCode;
        var /** @type {?} */ charater = String.fromCharCode(iKeyCode);
        var /** @type {?} */ result = this.validatorService.isValidPattern(charater, this.codeRegx);
        if (!result) {
            this.validatorService.indicateInvalidKeyPress(this.el);
        }
        return result;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    XontCodeDirective.prototype.onBlur = function (value) {
        if (!this.codeRegx.test(value)) {
            this.el.value = "";
            this.ngModelChange.emit("");
        }
    };
    /**
     * @return {?}
     */
    XontCodeDirective.prototype.onPaste = function () {
        var _this = this;
        setTimeout(function () {
            var /** @type {?} */ result = _this.validatorService.isValidPattern(_this.el.value, _this.codeRegx);
            if (!result) {
                _this.validatorService.indicateInvalidInput(_this.el);
            }
        }, 10);
    };
    /**
     * @return {?}
     */
    XontCodeDirective.prototype.onDrop = function () {
        var _this = this;
        setTimeout(function () {
            var /** @type {?} */ result = _this.validatorService.isValidPattern(_this.el.value, _this.codeRegx);
            if (!result) {
                _this.validatorService.indicateInvalidInput(_this.el);
            }
        }, 10);
    };
    return XontCodeDirective;
}());
XontCodeDirective.decorators = [
    { type: Directive, args: [{
                selector: "[xont_code]"
            },] },
];
/**
 * @nocollapse
 */
XontCodeDirective.ctorParameters = function () { return [
    { type: ElementRef, },
    { type: ValidatorService, },
]; };
XontCodeDirective.propDecorators = {
    'ngModelChange': [{ type: Output },],
    'onKeypress': [{ type: HostListener, args: ["keypress", ["$event"],] },],
    'onBlur': [{ type: HostListener, args: ["blur", ["$event.target.value"],] },],
    'onPaste': [{ type: HostListener, args: ["paste",] },],
    'onDrop': [{ type: HostListener, args: ["drop",] },],
};

var XontDescriptionDirective = (function () {
    /**
     * @param {?} elementRef
     * @param {?} validatorService
     */
    function XontDescriptionDirective(elementRef, validatorService) {
        this.elementRef = elementRef;
        this.validatorService = validatorService;
        this.ngModelChange = new EventEmitter();
        this.descRegx = new RegExp("^[^&#\";']*$");
        this.el = this.elementRef.nativeElement;
    }
    /**
     * @param {?} evt
     * @return {?}
     */
    XontDescriptionDirective.prototype.onKeypress = function (evt) {
        var /** @type {?} */ iKeyCode = (evt.which) ? evt.which : evt.keyCode;
        var /** @type {?} */ charater = String.fromCharCode(iKeyCode);
        var /** @type {?} */ result = this.validatorService.isValidPattern(charater, this.descRegx);
        if (!result) {
            this.validatorService.indicateInvalidKeyPress(this.el);
        }
        return result;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    XontDescriptionDirective.prototype.onBlur = function (value) {
        if (!this.descRegx.test(value)) {
            this.el.value = "";
            this.ngModelChange.emit("");
        }
    };
    /**
     * @return {?}
     */
    XontDescriptionDirective.prototype.onPaste = function () {
        var _this = this;
        setTimeout(function () {
            var /** @type {?} */ result = _this.validatorService.isValidPattern(_this.el.value, _this.descRegx);
            if (!result) {
                _this.validatorService.indicateInvalidInput(_this.el);
            }
        }, 10);
    };
    /**
     * @return {?}
     */
    XontDescriptionDirective.prototype.onDrop = function () {
        var _this = this;
        setTimeout(function () {
            var /** @type {?} */ result = _this.validatorService.isValidPattern(_this.el.value, _this.descRegx);
            if (!result) {
                _this.validatorService.indicateInvalidInput(_this.el);
            }
        }, 10);
    };
    return XontDescriptionDirective;
}());
XontDescriptionDirective.decorators = [
    { type: Directive, args: [{
                selector: "[xont_description]"
            },] },
];
/**
 * @nocollapse
 */
XontDescriptionDirective.ctorParameters = function () { return [
    { type: ElementRef, },
    { type: ValidatorService, },
]; };
XontDescriptionDirective.propDecorators = {
    'ngModelChange': [{ type: Output },],
    'onKeypress': [{ type: HostListener, args: ["keypress", ["$event"],] },],
    'onBlur': [{ type: HostListener, args: ["blur", ["$event.target.value"],] },],
    'onPaste': [{ type: HostListener, args: ["paste",] },],
    'onDrop': [{ type: HostListener, args: ["drop",] },],
};

var XontNumberDirective = (function () {
    /**
     * @param {?} elementRef
     * @param {?} validatorService
     */
    function XontNumberDirective(elementRef, validatorService) {
        this.elementRef = elementRef;
        this.validatorService = validatorService;
        this.ngModelChange = new EventEmitter();
        this.integerRegx = new RegExp("^[0-9]*$");
        this.el = this.elementRef.nativeElement;
    }
    /**
     * @param {?} evt
     * @return {?}
     */
    XontNumberDirective.prototype.onKeypress = function (evt) {
        var /** @type {?} */ iKeyCode = (evt.which) ? evt.which : evt.keyCode;
        var /** @type {?} */ charater = String.fromCharCode(iKeyCode);
        var /** @type {?} */ result = this.validatorService.isValidPattern(charater, this.integerRegx);
        if (!result) {
            this.validatorService.indicateInvalidKeyPress(this.el);
        }
        return result;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    XontNumberDirective.prototype.onBlur = function (value) {
        if (!this.integerRegx.test(value)) {
            this.el.value = "";
            this.ngModelChange.emit("");
        }
    };
    /**
     * @return {?}
     */
    XontNumberDirective.prototype.onPaste = function () {
        var _this = this;
        setTimeout(function () {
            var /** @type {?} */ result = _this.validatorService.isValidPattern(_this.el.value, _this.integerRegx);
            if (!result) {
                _this.validatorService.indicateInvalidInput(_this.el);
            }
        }, 10);
    };
    /**
     * @return {?}
     */
    XontNumberDirective.prototype.onDrop = function () {
        var _this = this;
        setTimeout(function () {
            var /** @type {?} */ result = _this.validatorService.isValidPattern(_this.el.value, _this.integerRegx);
            if (!result) {
                _this.validatorService.indicateInvalidInput(_this.el);
            }
        }, 10);
    };
    return XontNumberDirective;
}());
XontNumberDirective.decorators = [
    { type: Directive, args: [{
                selector: "[xont_number]"
            },] },
];
/**
 * @nocollapse
 */
XontNumberDirective.ctorParameters = function () { return [
    { type: ElementRef, },
    { type: ValidatorService, },
]; };
XontNumberDirective.propDecorators = {
    'ngModelChange': [{ type: Output },],
    'onKeypress': [{ type: HostListener, args: ["keypress", ["$event"],] },],
    'onBlur': [{ type: HostListener, args: ["blur", ["$event.target.value"],] },],
    'onPaste': [{ type: HostListener, args: ["paste",] },],
    'onDrop': [{ type: HostListener, args: ["drop",] },],
};

var CUSTOM_INPUT_CONTROL_DECIMAL_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(function () { return XontDecimalDirective; }),
    multi: true
};
var XontDecimalDirective = (function () {
    /**
     * @param {?} elementRef
     * @param {?} validatorService
     */
    function XontDecimalDirective(elementRef, validatorService) {
        this.elementRef = elementRef;
        this.validatorService = validatorService;
        this.ngModelChange = new EventEmitter();
        this.el = this.elementRef.nativeElement;
    }
    Object.defineProperty(XontDecimalDirective.prototype, "roudableDigits", {
        /**
         * @return {?}
         */
        get: function () {
            var /** @type {?} */ result = 0;
            if (this.inputParams.trim() != "") {
                var /** @type {?} */ array = this.inputParams.trim().split(",");
                if (array[0])
                    result = parseInt(array[0]);
            }
            return result;
        },
        enumerable: true,
        configurable: true
    });
    
    Object.defineProperty(XontDecimalDirective.prototype, "decimalDigits", {
        /**
         * @return {?}
         */
        get: function () {
            var /** @type {?} */ result = 0;
            if (this.inputParams.trim() != "") {
                var /** @type {?} */ array = this.inputParams.trim().split(",");
                if (array[1])
                    result = parseInt(array[1]);
            }
            return result;
        },
        enumerable: true,
        configurable: true
    });
    
    Object.defineProperty(XontDecimalDirective.prototype, "allowMinus", {
        /**
         * @return {?}
         */
        get: function () {
            var /** @type {?} */ result = false;
            if (this.inputParams.trim() != "") {
                var /** @type {?} */ array = this.inputParams.trim().split(",");
                if (array[2]) {
                    if (array[2] == "true") {
                        result = true;
                    }
                }
            }
            return result;
        },
        enumerable: true,
        configurable: true
    });
    
    /**
     * @param {?} value
     * @return {?}
     */
    XontDecimalDirective.prototype.writeValue = function (value) {
        if (value != undefined && value != null) {
            if (value.toString().trim() != this.el.value.trim()) {
                var /** @type {?} */ finalVal = this.validatorService.getNumberValue(value, this.decimalDigits);
                this.el.value = finalVal.toFixed(this.decimalDigits);
                this.ngModelChange.emit(finalVal);
            }
        }
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    XontDecimalDirective.prototype.registerOnChange = function (fn) {
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    XontDecimalDirective.prototype.registerOnTouched = function (fn) {
    };
    /**
     * @param {?} evt
     * @return {?}
     */
    XontDecimalDirective.prototype.onKeypress = function (evt) {
        var /** @type {?} */ result = this.validatorService.numbersOnly(evt, this.roudableDigits, this.decimalDigits, this.allowMinus);
        if (!result) {
            this.validatorService.indicateInvalidKeyPress(this.el);
        }
        return result;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    XontDecimalDirective.prototype.onKeyup = function (value) {
        this.ngModelChange.emit(this.validatorService.getNumberValue(value, this.decimalDigits));
    };
    /**
     * @param {?} value
     * @return {?}
     */
    XontDecimalDirective.prototype.onBlur = function (value) {
        var /** @type {?} */ finalVal = this.validatorService.getNumberValue(value, this.decimalDigits);
        this.el.value = finalVal.toFixed(this.decimalDigits);
        this.ngModelChange.emit(finalVal);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    XontDecimalDirective.prototype.onFocus = function (value) {
        if (this.validatorService.getNumberValue(value, this.decimalDigits) == 0) {
            this.el.value = "";
        }
    };
    /**
     * @return {?}
     */
    XontDecimalDirective.prototype.onPaste = function () {
        var _this = this;
        setTimeout(function () {
            if (isNaN(parseFloat(_this.el.value))) {
                _this.validatorService.indicateInvalidInput(_this.el);
            }
        }, 10);
    };
    /**
     * @return {?}
     */
    XontDecimalDirective.prototype.onDrop = function () {
        var _this = this;
        setTimeout(function () {
            if (isNaN(parseFloat(_this.el.value))) {
                _this.validatorService.indicateInvalidInput(_this.el);
            }
        }, 10);
    };
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    XontDecimalDirective.prototype.setDisabledState = function (isDisabled) {
        this.elementRef.nativeElement.disabled = isDisabled;
    };
    return XontDecimalDirective;
}());
XontDecimalDirective.decorators = [
    { type: Directive, args: [{
                selector: "[xont_decimal]",
                providers: [CUSTOM_INPUT_CONTROL_DECIMAL_ACCESSOR]
            },] },
];
/**
 * @nocollapse
 */
XontDecimalDirective.ctorParameters = function () { return [
    { type: ElementRef, },
    { type: ValidatorService, },
]; };
XontDecimalDirective.propDecorators = {
    'ngModelChange': [{ type: Output },],
    'inputParams': [{ type: Input, args: ['xont_decimal',] },],
    'onKeypress': [{ type: HostListener, args: ["keypress", ["$event"],] },],
    'onKeyup': [{ type: HostListener, args: ["keyup", ["$event.target.value"],] },],
    'onBlur': [{ type: HostListener, args: ["blur", ["$event.target.value"],] },],
    'onFocus': [{ type: HostListener, args: ["focus", ["$event.target.value"],] },],
    'onPaste': [{ type: HostListener, args: ["paste",] },],
    'onDrop': [{ type: HostListener, args: ["drop",] },],
};

var XontDatePipe = (function () {
    /**
     * @param {?} datetimeService
     */
    function XontDatePipe(datetimeService) {
        this.datetimeService = datetimeService;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    XontDatePipe.prototype.transform = function (value) {
        var /** @type {?} */ result = '';
        if (!value) {
            return result;
        }
        var /** @type {?} */ text = value.toString().trim();
        if (value instanceof Date) {
            if (value.toString() != 'Invalid Date') {
                result = this.datetimeService.getDisplayDate(value);
            }
        }
        else if (text != '') {
            var /** @type {?} */ d = new Date(text);
            if (d.toString() != 'Invalid Date') {
                result = this.datetimeService.getDisplayDate(d);
            }
        }
        return result;
    };
    return XontDatePipe;
}());
XontDatePipe.decorators = [
    { type: Pipe, args: [{
                name: 'xont_date'
            },] },
];
/**
 * @nocollapse
 */
XontDatePipe.ctorParameters = function () { return [
    { type: DatetimeService, },
]; };

var XontDateTimePipe = (function () {
    /**
     * @param {?} datetimeService
     */
    function XontDateTimePipe(datetimeService) {
        this.datetimeService = datetimeService;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    XontDateTimePipe.prototype.transform = function (value) {
        var /** @type {?} */ result = '';
        if (!value) {
            return result;
        }
        var /** @type {?} */ text = value.toString().trim();
        if (value instanceof Date) {
            if (value.toString() != 'Invalid Date') {
                result = this.datetimeService.getDisplayDate(value) + ' ' + this.datetimeService.get12HourTime(value);
            }
        }
        else if (text != '') {
            var /** @type {?} */ d = new Date(text);
            if (d.toString() != 'Invalid Date') {
                result = this.datetimeService.getDisplayDate(d) + ' ' + this.datetimeService.get12HourTime(d);
            }
        }
        return result;
    };
    return XontDateTimePipe;
}());
XontDateTimePipe.decorators = [
    { type: Pipe, args: [{
                name: 'xont_datetime'
            },] },
];
/**
 * @nocollapse
 */
XontDateTimePipe.ctorParameters = function () { return [
    { type: DatetimeService, },
]; };

var XontTimePipe = (function () {
    /**
     * @param {?} datetimeService
     */
    function XontTimePipe(datetimeService) {
        this.datetimeService = datetimeService;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    XontTimePipe.prototype.transform = function (value) {
        var /** @type {?} */ result = '';
        if (!value) {
            return result;
        }
        var /** @type {?} */ text = value.toString().trim();
        if (value instanceof Date) {
            if (value.toString() != 'Invalid Date') {
                result = this.datetimeService.get12HourTime(value);
            }
        }
        else if (text != '') {
            var /** @type {?} */ d = new Date(text);
            if (d.toString() != 'Invalid Date') {
                result = this.datetimeService.get12HourTime(d);
            }
        }
        return result;
    };
    return XontTimePipe;
}());
XontTimePipe.decorators = [
    { type: Pipe, args: [{
                name: 'xont_time'
            },] },
];
/**
 * @nocollapse
 */
XontTimePipe.ctorParameters = function () { return [
    { type: DatetimeService, },
]; };

var XontVenturaValidatorsModule = (function () {
    function XontVenturaValidatorsModule() {
    }
    /**
     * @return {?}
     */
    XontVenturaValidatorsModule.forRoot = function () {
        return {
            ngModule: XontVenturaValidatorsModule,
            providers: [ValidatorService]
        };
    };
    return XontVenturaValidatorsModule;
}());
XontVenturaValidatorsModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    ValidateRangeDirective,
                    ValidateDateDirective,
                    XontValueDirective,
                    XontValuePipe,
                    XontQuantityDirective,
                    XontQuantityPipe,
                    XontCodeDirective,
                    XontDescriptionDirective,
                    XontNumberDirective,
                    XontDecimalDirective,
                    XontDatePipe,
                    XontDateTimePipe,
                    XontTimePipe
                ],
                exports: [
                    ValidateRangeDirective,
                    ValidateDateDirective,
                    XontValueDirective,
                    XontValuePipe,
                    XontQuantityDirective,
                    XontQuantityPipe,
                    XontCodeDirective,
                    XontDescriptionDirective,
                    XontNumberDirective,
                    XontDecimalDirective,
                    XontDatePipe,
                    XontDateTimePipe,
                    XontTimePipe
                ],
                providers: [DatetimeService, XontQuantityPipe, XontValuePipe, XontDatePipe, ValidatorService, XontDateTimePipe, XontTimePipe],
            },] },
];
/**
 * @nocollapse
 */
XontVenturaValidatorsModule.ctorParameters = function () { return []; };

export { XontVenturaValidatorsModule, ValidateRangeDirective, ValidateDateDirective, CUSTOM_INPUT_CONTROL_VALUE_ACCESSOR, XontValueDirective, XontValuePipe, CUSTOM_INPUT_CONTROL_QUANTITY_ACCESSOR, XontQuantityDirective, XontQuantityPipe, ValidatorService, XontCodeDirective, XontDescriptionDirective, XontNumberDirective, CUSTOM_INPUT_CONTROL_DECIMAL_ACCESSOR, XontDecimalDirective, XontDatePipe, XontDateTimePipe, XontTimePipe };
