// xont-ventura-buttons/index.js
import { Component, Input, Output, EventEmitter, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var VenturaButtonComponent = (function () {
    function VenturaButtonComponent() {
        this.id = '';
        this.type = '';
        this.btnCaption = '';
        this.confMsg = 'This will discard modified data.';
        this.confContinueMsg = 'Do you wish to continue ?';
        this.confOkBtnCaption = 'OK';
        this.confCancelBtnCaption = 'Cancel';
        this.modifyCheck = true;
        this.modified = true;
        this.onClick = new EventEmitter();
        this.popUpVisible = false;
    }
    /**
     * @return {?}
     */
    VenturaButtonComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.type.trim().toLowerCase() == 'cancel' && !this.btnCaption) {
            this.btnCaption = 'Cancel';
        }
    };
    /**
     * @return {?}
     */
    VenturaButtonComponent.prototype.btnTemplateButton_Click = /**
     * @return {?}
     */
    function () {
        if (this.type.trim().toLowerCase() == 'cancel') {
            if (!this.modifyCheck) {
                this.onClick.emit();
            }
            else {
                if (this.modified) {
                    this.popUpVisible = true;
                }
                else {
                    this.onClick.emit();
                }
            }
        }
        else {
            this.onClick.emit();
        }
    };
    /**
     * @return {?}
     */
    VenturaButtonComponent.prototype.OK_Click = /**
     * @return {?}
     */
    function () {
        this.popUpVisible = false;
        this.onClick.emit();
    };
    /**
     * @return {?}
     */
    VenturaButtonComponent.prototype.Cancel_Click = /**
     * @return {?}
     */
    function () {
        this.popUpVisible = false;
    };
    VenturaButtonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ventura-button',
                    template: "<button type=\"button\" name=\"{{name}}\" class=\"MainButtonStyle\" id=\"{{id}}\" \n        (click)=\"btnTemplateButton_Click()\" class=\"MainButtonStyle\">\n        {{btnCaption}}\n    </button>\n    <div class=\"PopUpMessage\" *ngIf=\"popUpVisible\">\n        <div class=\"divInner\">\n            <div class=\"divCaptions\">\n                <span class=\"Captionstyle\">{{confMsg}}</span>\n                <div><span class=\"Captionstyle\"> {{confContinueMsg}}</span></div>\n            </div>\n            <div>\n                <button type=\"button\" id=\"{{id}} + '_OK'\" class=\"MainButtonStyle\" (click)=\"OK_Click()\">{{confOkBtnCaption}}</button>\n                <button type=\"button\" id=\"{{id}} + '_Cancel'\" class=\"MainButtonStyle\" (click)=\"Cancel_Click()\">{{confCancelBtnCaption}}</button>\n            </div>\n        </div>\n    </div>",
                    styles: [
                        '.PopUpMessage { width: 100%; height: 100%; position: fixed; top: 0; background: #96969669; z- index: 20000;}',
                        '.divInner {position: absolute;  left: 40%; top: 40%; background: white; border: #4c87bf 1px solid; border-radius:12px; padding: 15px 20px 10px 20px;}',
                        '.divCaptions {padding-bottom:12px;}'
                    ]
                },] },
    ];
    //private methods end
    /** @nocollapse */
    VenturaButtonComponent.propDecorators = {
        "id": [{ type: Input },],
        "type": [{ type: Input },],
        "btnCaption": [{ type: Input },],
        "confMsg": [{ type: Input },],
        "confContinueMsg": [{ type: Input },],
        "confOkBtnCaption": [{ type: Input },],
        "confCancelBtnCaption": [{ type: Input },],
        "modifyCheck": [{ type: Input },],
        "modified": [{ type: Input },],
        "onClick": [{ type: Output },],
    };
    return VenturaButtonComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var XontVenturaButtonModule = (function () {
    function XontVenturaButtonModule() {
    }
    /**
     * @return {?}
     */
    XontVenturaButtonModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: XontVenturaButtonModule,
        };
    };
    XontVenturaButtonModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [
                        VenturaButtonComponent,
                    ],
                    exports: [
                        VenturaButtonComponent,
                    ]
                },] },
    ];
    return XontVenturaButtonModule;
}());

export { XontVenturaButtonModule, VenturaButtonComponent };
