// xont-ventura-buttons/xont-ventura-buttons.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/common'], factory) :
    (global = global || self, factory(global['xont-ventura-buttons'] = {}, global.core, global.common));
}(this, function (exports, core, common) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var VenturaButtonComponent = (function () {
        function VenturaButtonComponent() {
            this.id = '';
            this.type = '';
            this.btnCaption = '';
            this.confMsg = 'This will discard modified data.';
            this.confContinueMsg = 'Do you wish to continue ?';
            this.confOkBtnCaption = 'OK';
            this.confCancelBtnCaption = 'Cancel';
            this.modifyCheck = true;
            this.modified = true;
            this.onClick = new core.EventEmitter();
            this.popUpVisible = false;
        }
        /**
         * @return {?}
         */
        VenturaButtonComponent.prototype.ngOnInit = /**
         * @return {?}
         */
        function () {
            if (this.type.trim().toLowerCase() == 'cancel' && !this.btnCaption) {
                this.btnCaption = 'Cancel';
            }
        };
        /**
         * @return {?}
         */
        VenturaButtonComponent.prototype.btnTemplateButton_Click = /**
         * @return {?}
         */
        function () {
            if (this.type.trim().toLowerCase() == 'cancel') {
                if (!this.modifyCheck) {
                    this.onClick.emit();
                }
                else {
                    if (this.modified) {
                        this.popUpVisible = true;
                    }
                    else {
                        this.onClick.emit();
                    }
                }
            }
            else {
                this.onClick.emit();
            }
        };
        /**
         * @return {?}
         */
        VenturaButtonComponent.prototype.OK_Click = /**
         * @return {?}
         */
        function () {
            this.popUpVisible = false;
            this.onClick.emit();
        };
        /**
         * @return {?}
         */
        VenturaButtonComponent.prototype.Cancel_Click = /**
         * @return {?}
         */
        function () {
            this.popUpVisible = false;
        };
        VenturaButtonComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'ventura-button',
                        template: "<button type=\"button\" name=\"{{name}}\" class=\"MainButtonStyle\" id=\"{{id}}\" \n        (click)=\"btnTemplateButton_Click()\" class=\"MainButtonStyle\">\n        {{btnCaption}}\n    </button>\n    <div class=\"PopUpMessage\" *ngIf=\"popUpVisible\">\n        <div class=\"divInner\">\n            <div class=\"divCaptions\">\n                <span class=\"Captionstyle\">{{confMsg}}</span>\n                <div><span class=\"Captionstyle\"> {{confContinueMsg}}</span></div>\n            </div>\n            <div>\n                <button type=\"button\" id=\"{{id}} + '_OK'\" class=\"MainButtonStyle\" (click)=\"OK_Click()\">{{confOkBtnCaption}}</button>\n                <button type=\"button\" id=\"{{id}} + '_Cancel'\" class=\"MainButtonStyle\" (click)=\"Cancel_Click()\">{{confCancelBtnCaption}}</button>\n            </div>\n        </div>\n    </div>",
                        styles: [
                            '.PopUpMessage { width: 100%; height: 100%; position: fixed; top: 0; background: #96969669; z- index: 20000;}',
                            '.divInner {position: absolute;  left: 40%; top: 40%; background: white; border: #4c87bf 1px solid; border-radius:12px; padding: 15px 20px 10px 20px;}',
                            '.divCaptions {padding-bottom:12px;}'
                        ]
                    },] },
        ];
        //private methods end
        /** @nocollapse */
        VenturaButtonComponent.propDecorators = {
            "id": [{ type: core.Input },],
            "type": [{ type: core.Input },],
            "btnCaption": [{ type: core.Input },],
            "confMsg": [{ type: core.Input },],
            "confContinueMsg": [{ type: core.Input },],
            "confOkBtnCaption": [{ type: core.Input },],
            "confCancelBtnCaption": [{ type: core.Input },],
            "modifyCheck": [{ type: core.Input },],
            "modified": [{ type: core.Input },],
            "onClick": [{ type: core.Output },],
        };
        return VenturaButtonComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var XontVenturaButtonModule = (function () {
        function XontVenturaButtonModule() {
        }
        /**
         * @return {?}
         */
        XontVenturaButtonModule.forRoot = /**
         * @return {?}
         */
        function () {
            return {
                ngModule: XontVenturaButtonModule,
            };
        };
        XontVenturaButtonModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule
                        ],
                        declarations: [
                            VenturaButtonComponent,
                        ],
                        exports: [
                            VenturaButtonComponent,
                        ]
                    },] },
        ];
        return XontVenturaButtonModule;
    }());

    exports.XontVenturaButtonModule = XontVenturaButtonModule;
    exports.VenturaButtonComponent = VenturaButtonComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
