// simple-data-slider/index.js
import { Component, EventEmitter, Input, NgModule, Output, ViewChild, forwardRef } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DataTableModule } from 'angular2-datatable';
import { FormsModule, NG_VALUE_ACCESSOR } from '@angular/forms';
import { filter } from 'lodash';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var SIMPLE_DATA_SLIDER_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(function () { return SimpleDataSliderComponent; }),
    multi: true
};
var SimpleDataSliderComponent = (function () {
    function SimpleDataSliderComponent() {
        this._originalDataSet = null;
        this._dataSet = null;
        this._minWidth = 500;
        this._pagerLeft = 250;
        this._codeCleared = false;
        this._descCleared = false;
        this._dataSliderStatus = 'none';
        this._stillDataLoading = false;
        this._prevCode = '';
        this._prevDesc = '';
        this._pageSize = 12;
        this._message = '';
        this.ngModelChange = new EventEmitter();
        this.onChanged = new EventEmitter();
        this.onDataBind = new EventEmitter();
        this.valid = true;
        this.mandatoryValid = true;
        this.valueValid = true;
    }
    /**
     * @param {?} code
     * @param {?} desc
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.setValues = /**
     * @param {?} code
     * @param {?} desc
     * @return {?}
     */
    function (code, desc) {
        this.ngModel[this.modelProps[0]] = code;
        this.ngModel[this.modelProps[1]] = desc;
        this.SetValidations(true, !code);
    };
    /**
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.ngModelChange.emit(this.ngModel);
        this._minWidth = this.inputWidths[0] + this.inputWidths[1] + 15;
        if (!this.modelProps)
            this.modelProps = this.fields;
        this._prevCode = this.ngModel[this.modelProps[0]].toString().trim();
        this._prevDesc = this.ngModel[this.modelProps[1]].toString().trim();
        this.SetValidations(true, !this._prevCode);
    };
    /**
     * @param {?} item
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.ItemSelected = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        var /** @type {?} */ selectedCode = item[this.fields[0]].toString().trim();
        var /** @type {?} */ selectedDesc = item[this.fields[1]].toString().trim();
        this.ngModel[this.modelProps[0]] = selectedCode;
        this.ngModel[this.modelProps[1]] = selectedDesc;
        this.SetValidations(true, false);
        if (selectedCode != this._prevCode || selectedDesc != this._prevDesc) {
            this.onChanged.emit(this.ngModel);
            this._prevCode = selectedCode;
            this._prevDesc = selectedDesc;
        }
        this.ResetProps();
    };
    /**
     * @param {?} e
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.pagerClickCapture = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        event.preventDefault();
        event.stopPropagation();
    };
    /**
     * @param {?} e
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.CodeMousedIn = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (this._dataSliderStatus == 'none') {
            this._dataSliderStatus = 'block';
            this.PopulateDataSet();
        }
    };
    /**
     * @param {?} e
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.DescMousedIn = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (this._dataSliderStatus == 'none') {
            this._dataSliderStatus = 'block';
            this.PopulateDataSet();
        }
    };
    /**
     * @param {?} e
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.CodeKeyDown = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (e.keyCode != 9 && this._dataSet == null) {
            e.preventDefault();
            e.stopPropagation();
            return;
        }
        if (e.keyCode == 8 && this._codeCleared && !this.ngModel[this.modelProps[0]]) {
            this.descInput.nativeElement.focus();
            e.preventDefault();
            this._codeCleared = false;
        }
    };
    /**
     * @param {?} e
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.DescKeyDown = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (e.keyCode != 9 && this._dataSet == null) {
            e.preventDefault();
            e.stopPropagation();
            return;
        }
        if (e.keyCode == 8 && this._descCleared && !this.ngModel[this.modelProps[1]]) {
            this.codeInput.nativeElement.focus();
            e.preventDefault();
            this._descCleared = false;
        }
        else if (e.keyCode == 9 && this._dataSliderStatus == 'block') {
            var /** @type {?} */ typedCode = this.ngModel[this.modelProps[0]].toString().trim();
            var /** @type {?} */ typedDesc = this.ngModel[this.modelProps[1]].toString().trim();
            if (this.validationOutput) {
                if (typedCode || typedDesc) {
                    this.FilterSource();
                    if (this._dataSet && this._dataSet.length > 0) {
                        var /** @type {?} */ item = this._dataSet[0];
                        this.ngModel[this.modelProps[0]] = item[this.fields[0]].toString().trim();
                        this.ngModel[this.modelProps[1]] = item[this.fields[1]].toString().trim();
                        this.SetValidations(true, false);
                    }
                    else {
                        if (this.matchExact) {
                            this.ngModel[this.modelProps[0]] = '';
                            this.ngModel[this.modelProps[1]] = '';
                            this.SetValidations(true, true);
                        }
                        else {
                            this.SetValidations(false, !typedCode);
                        }
                    }
                }
                else {
                    this.SetValidations(true, true);
                }
            }
            else
                this.SetValidations(true, !typedCode);
            if (typedCode != this._prevCode || typedDesc != this._prevDesc) {
                this.onChanged.emit(this.ngModel);
                this._prevCode = typedCode;
                this._prevDesc = typedDesc;
            }
            this.ResetProps();
        }
    };
    /**
     * @param {?} isValueValid
     * @param {?} isEmpty
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.SetValidations = /**
     * @param {?} isValueValid
     * @param {?} isEmpty
     * @return {?}
     */
    function (isValueValid, isEmpty) {
        if (this.mandatory && isEmpty)
            this.mandatoryValid = false;
        else
            this.mandatoryValid = true;
        if (isValueValid) {
            if (this.validationOutput)
                this.ngModel[this.validationOutput] = true;
            this.valueValid = true;
        }
        else {
            if (this.validationOutput)
                this.ngModel[this.validationOutput] = false;
            this.valueValid = false;
        }
        if (!this.mandatoryValid) {
            this._message = "*";
            this.valid = false;
        }
        else if (!this.valueValid) {
            this._message = this.errorMessage;
            this.valid = false;
        }
        else {
            this._message = '';
            this.valid = true;
        }
    };
    /**
     * @param {?} e
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.CodeKeyUp = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (e.keyCode == 8 && !this.ngModel[this.modelProps[0]]) {
            this._codeCleared = true;
        }
        else
            this._codeCleared = false;
        //if not tab apply filter showing data if not shown already
        if (e.keyCode != 9) {
            if (this._dataSliderStatus != 'block')
                this._dataSliderStatus = 'block';
            if (this._originalDataSet == null && !this._stillDataLoading)
                this.PopulateDataSet();
            this.FilterSource();
        }
    };
    /**
     * @param {?} e
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.DescKeyUp = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (e.keyCode == 8 && !this.ngModel[this.modelProps[1]]) {
            this._descCleared = true;
        }
        else
            this._descCleared = false;
        if (e.keyCode != 9) {
            if (this._dataSliderStatus != 'block')
                this._dataSliderStatus = 'block';
            if (this._originalDataSet == null && !this._stillDataLoading)
                this.PopulateDataSet();
            this.FilterSource();
        }
    };
    /**
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.ClearModel = /**
     * @return {?}
     */
    function () {
        this.ngModel[this.modelProps[0]] = '';
        this.ngModel[this.modelProps[1]] = '';
        this.FilterSource();
        this.codeInput.nativeElement.focus();
        this.SetValidations(true, true);
    };
    //Misleading the interface start
    /**
     * @param {?} value
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
    };
    /**
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.AddEvent = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ self = this;
        /**
         * @param {?} e
         * @return {?}
         */
        function handleClick(e) {
            setTimeout(function () {
                if (!self.element.nativeElement.contains(e.target)) {
                    if (self._dataSliderStatus == 'block') {
                        var /** @type {?} */ typedCode = self.ngModel[self.modelProps[0]].toString().trim();
                        var /** @type {?} */ typedDesc = self.ngModel[self.modelProps[1]].toString().trim();
                        if (self.validationOutput) {
                            if (typedCode || typedDesc) {
                                self.FilterSource();
                                if (self._dataSet && self._dataSet.length > 0) {
                                    var /** @type {?} */ item = self._dataSet[0];
                                    self.ngModel[self.modelProps[0]] = item[self.fields[0]].toString().trim();
                                    self.ngModel[self.modelProps[1]] = item[self.fields[1]].toString().trim();
                                    self.SetValidations(true, false);
                                }
                                else {
                                    if (self.matchExact) {
                                        self.ngModel[self.modelProps[0]] = '';
                                        self.ngModel[self.modelProps[1]] = '';
                                        self.SetValidations(true, true);
                                    }
                                    else {
                                        self.SetValidations(false, !typedCode);
                                    }
                                }
                            }
                            else {
                                self.SetValidations(true, true);
                            }
                        }
                        else
                            self.SetValidations(true, !typedCode);
                        if (typedCode != self._prevCode || typedDesc != self._prevDesc) {
                            self.onChanged.emit(self.ngModel);
                            self._prevCode = typedCode;
                            self._prevDesc = typedDesc;
                        }
                    }
                    self.ResetProps();
                    window.removeEventListener("click", handleClick);
                }
            }, 0);
        }
        window.removeEventListener("click", handleClick);
        window.addEventListener("click", handleClick);
    };
    /**
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.PopulateDataSet = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._stillDataLoading = true;
        this.onDataBind.emit();
        var /** @type {?} */ subsription = this.dataSourceObservable.subscribe(function (data) {
            _this._dataSet = data;
            _this._originalDataSet = data;
            _this._pagerLeft = (data.length / _this._pageSize) > 4 ? 193 : 220;
        }, function (error) { console.error('this is from data slider', error); }, function () { _this._stillDataLoading = false; });
        this.AddEvent();
    };
    /**
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.FilterSource = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ FirstfilterField = this.fields[0];
        var /** @type {?} */ FirstfilterValue = this.ngModel[this.modelProps[0]].toString().toUpperCase();
        var /** @type {?} */ SecondfilterField = this.fields[1];
        var /** @type {?} */ SecondfilterValue = this.ngModel[this.modelProps[1]].toString().toUpperCase();
        if (this._originalDataSet != null) {
            this._dataSet = filter(this._originalDataSet, function (row) {
                if (FirstfilterValue && SecondfilterValue) {
                    return (row[FirstfilterField].toString().toUpperCase().indexOf(FirstfilterValue) > -1)
                        && (row[SecondfilterField].toString().toUpperCase().indexOf(SecondfilterValue) > -1);
                }
                else if (FirstfilterValue)
                    return (row[FirstfilterField].toString().toUpperCase().indexOf(FirstfilterValue) > -1);
                else if (SecondfilterValue)
                    return (row[SecondfilterField].toString().toUpperCase().indexOf(SecondfilterValue) > -1);
                else
                    return true;
            });
            this.scrollDiv.nativeElement.scrollTop = 0;
        }
    };
    /**
     * @return {?}
     */
    SimpleDataSliderComponent.prototype.ResetProps = /**
     * @return {?}
     */
    function () {
        this._codeCleared = false;
        this._descCleared = false;
        this._dataSliderStatus = 'none';
        this._dataSet = null;
        this._originalDataSet = null;
    };
    SimpleDataSliderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'simple-data-slider',
                    styles: ['.Textboxstyle  { padding-left:2px;}',
                        ".dataSlider {\n            position: absolute;\n            min-height:260px;\n            border-radius: 0 7px 7px 7px;\n            background-color:white;\n            margin-top: 3px;\n            margin-bottom:15px;\n            z-index:1000;\n            border: 1.5px #5399d6 solid;\n            padding-left: 5px;\n            color: #01539D;\n            font-family: Trebuchet MS;\n            background-color:ghostwhite;\n         }",
                        '.dtRow { cursor:pointer; }',
                        '.dtRow > td {padding-right:10px;}',
                        '.dtRow > td:hover {color:red; transform: scale(1.015);}',
                        '.pmtPaginator { position: absolute; bottom: 6px;}',
                        '.sliderLoader {height: 80px; margin: auto; display: block; margin-top: 80px;}',
                        '.clickButton {padding-left:6.5px;font-size:16px;padding-right:6.5px;cursor:pointer;display:inline-block;}',
                        '.scrollDiv { overflow: auto; overflow-x: hidden; padding-left:10px; margin:5px 5px 5px 0;}',
                        '.footerDiv {height:26px;padding-top:35px;}',
                        '.closeI {top:-24px;right:22px;position: absolute;font-size:18px;cursor:pointer;}'
                    ],
                    template: "<div #element>\n                    <input #codeInput [(ngModel)]=\"ngModel[modelProps[0]]\" [maxlength]=\"maxLengths[0]\" [style.width]=\"inputWidths[0] + 'px'\" \n                        (mousedown)=\"CodeMousedIn($event)\" (keydown)=\"CodeKeyDown($event)\" (keyup)=\"CodeKeyUp($event)\" class=\"Textboxstyle\"/>\n                    <div (click)=\"CodeMousedIn($event)\" class=\"fa fa-angle-double-down clickButton\" aria-hidden=\"true\"></div>\n                    <input #descInput [(ngModel)]=\"ngModel[modelProps[1]]\" [maxlength]=\"maxLengths[1]\" [style.width]=\"inputWidths[1] + 'px'\"\n                        (mousedown)=\"DescMousedIn($event)\" (keydown)=\"DescKeyDown($event)\" (keyup)=\"DescKeyUp($event)\" class=\"Textboxstyle\" style=\"margin-right:25px;\"/>\n                    <small *ngIf=\"_message\" class=\"Errormessagetextstyle\">{{_message}}</small>\n                    <div class=\"dataSlider\" [style.display]=\"_dataSliderStatus\" [style.minWidth]=\"_minWidth + 60 + 'px'\"> \n                        <div *ngIf=\"_dataSet == null\">\n                            <img class=\"sliderLoader\" src='../App_Themes_V3/Blue/images/load_pmt.gif'/>\n                        </div>\n                        <div #scrollDiv [style.height]=\"_dataSet?.length > _pageSize ? '222px' : '245px'\" class=\"scrollDiv\" *ngIf= \"!_stillDataLoading\">\n                            <table [mfRowsOnPage]=\"_pageSize\" [style.minWidth]=\"_minWidth + 'px'\" [mfData]=\"_dataSet\" #mf=\"mfDataTable\">\n                            <tbody>\n                                <tr *ngFor=\"let item of mf.data\" class=\"dtRow\" (click)=\"ItemSelected(item)\" >\n                                    <td [style.minWidth]=\"_minWidth + 'px'\">{{item[fields[0]].toString().trim()}} - {{item[fields[1]].toString().trim()}}</td>\n                                </tr>\n                            </tbody>\n                             <tfoot>\n                              <div class=\"footerDiv\" *ngIf=\"_dataSet?.length > _pageSize || _dataSet?.length == 0\">\n                                <div class=\"pmtPaginator\" (click)=\"pagerClickCapture($event)\" *ngIf=\"_dataSet?.length > _pageSize\" [style.left]=\"_pagerLeft + 20 + 'px'\">\n                                    <mfBootstrapPaginator></mfBootstrapPaginator>\n                                </div>\n                                <div class=\"pmtPaginator\" [style.left]=\"(_minWidth/2 - 60) + 'px'\" *ngIf= \"_dataSet?.length == 0\">\n                                    No Data Found to Display\n                                </div>\n                              </div>\n\n                            </tfoot>\n                            </table>\n                        <div>\n                    </div>\n                <i class=\"fa fa-times-circle closeI\" title=\"clear code/description\" (click)=\"ClearModel()\"></i>\n                </div>",
                    providers: [SIMPLE_DATA_SLIDER_VALUE_ACCESSOR]
                },] },
    ];
    //Private Methods End
    /** @nocollapse */
    SimpleDataSliderComponent.propDecorators = {
        "maxLengths": [{ type: Input },],
        "inputWidths": [{ type: Input },],
        "fields": [{ type: Input },],
        "modelProps": [{ type: Input },],
        "validationOutput": [{ type: Input },],
        "errorMessage": [{ type: Input },],
        "mandatory": [{ type: Input },],
        "matchExact": [{ type: Input },],
        "ngModel": [{ type: Input },],
        "ngModelChange": [{ type: Output },],
        "onChanged": [{ type: Output },],
        "onDataBind": [{ type: Output },],
        "element": [{ type: ViewChild, args: ['element',] },],
        "codeInput": [{ type: ViewChild, args: ['codeInput',] },],
        "descInput": [{ type: ViewChild, args: ['descInput',] },],
        "scrollDiv": [{ type: ViewChild, args: ['scrollDiv',] },],
    };
    return SimpleDataSliderComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var SimpleDataSliderModule = (function () {
    function SimpleDataSliderModule() {
    }
    /**
     * @return {?}
     */
    SimpleDataSliderModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: SimpleDataSliderModule,
        };
    };
    SimpleDataSliderModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule, DataTableModule, FormsModule
                    ],
                    declarations: [
                        SimpleDataSliderComponent
                    ],
                    exports: [
                        SimpleDataSliderComponent
                    ]
                },] },
    ];
    return SimpleDataSliderModule;
}());

export { SimpleDataSliderModule, SIMPLE_DATA_SLIDER_VALUE_ACCESSOR, SimpleDataSliderComponent };
