// xont-ventura-list-prompt/index.js
import { Injectable, forwardRef, Component, Input, Output, ViewChild, ViewChildren, EventEmitter, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DataTable, DataTableModule } from 'angular2-datatable';
import { NG_VALUE_ACCESSOR, FormsModule } from '@angular/forms';
import { CommonService } from 'xont-ventura-services';
import { Headers, Http } from '@angular/http';
import { Observable } from 'rxjs/Rx';
import { union, filter, orderBy } from 'lodash';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var XontVenturaListPromptService = (function () {
    function XontVenturaListPromptService(http, commonService) {
        this.http = http;
        this.commonService = commonService;
    }
    /**
     * @return {?}
     */
    XontVenturaListPromptService.prototype.GetApiPrefix = /**
     * @return {?}
     */
    function () {
        return this.commonService.getAPIPrefix();
    };
    /**
     * @param {?} masterDefValueArgs
     * @return {?}
     */
    XontVenturaListPromptService.prototype.GetMasterDefinitionValues = /**
     * @param {?} masterDefValueArgs
     * @return {?}
     */
    function (masterDefValueArgs) {
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.GetApiPrefix() + "/api/Prompt/GetMasterValues", masterDefValueArgs, { headers: headers })
            .map(function (response) { return response.json(); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} predefinedPromptArgs
     * @return {?}
     */
    XontVenturaListPromptService.prototype.GetPredefinedPromptData = /**
     * @param {?} predefinedPromptArgs
     * @return {?}
     */
    function (predefinedPromptArgs) {
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.GetApiPrefix() + "/api/Prompt/GetPredefinedPrompt", { listFromID: predefinedPromptArgs.ListFromID, externalParameters: predefinedPromptArgs.ExternalParameters }, { headers: headers })
            .map(function (response) { return response.json(); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} listPromptProductArgs
     * @return {?}
     */
    XontVenturaListPromptService.prototype.GetListPromptProductData = /**
     * @param {?} listPromptProductArgs
     * @return {?}
     */
    function (listPromptProductArgs) {
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.GetApiPrefix() + "/api/ListPrompt/GetListPromptProductData", listPromptProductArgs, { headers: headers })
            .map(function (response) { return response.json(); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} promptTerritoryArgs
     * @return {?}
     */
    XontVenturaListPromptService.prototype.GetListPromptTerritoryData = /**
     * @param {?} promptTerritoryArgs
     * @return {?}
     */
    function (promptTerritoryArgs) {
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        //return this.http.post(this.GetApiPrefix() + "/api/Prompt/GetMasterValues", promptTerritoryArgs, { headers: headers })V3008
        return this.http.post(this.GetApiPrefix() + "/api/ListPrompt/GetListPromptTerritoryData", promptTerritoryArgs, { headers: headers }) //V3008Added
            .map(function (response) { return response.json(); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} listPromptExecutiveArgs
     * @return {?}
     */
    XontVenturaListPromptService.prototype.GetListPromptExecutiveData = /**
     * @param {?} listPromptExecutiveArgs
     * @return {?}
     */
    function (listPromptExecutiveArgs) {
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.GetApiPrefix() + "/api/ListPrompt/GetListPromptExecutiveData", listPromptExecutiveArgs, { headers: headers })
            .map(function (response) { return response.json(); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} listPromptRetailerArgs
     * @return {?}
     */
    XontVenturaListPromptService.prototype.GetListPromptRetailerData = /**
     * @param {?} listPromptRetailerArgs
     * @return {?}
     */
    function (listPromptRetailerArgs) {
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.GetApiPrefix() + "/api/ListPrompt/GetListPromptRetailerData", listPromptRetailerArgs, { headers: headers })
            .map(function (response) { return response.json(); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    XontVenturaListPromptService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    XontVenturaListPromptService.ctorParameters = function () { return [
        { type: Http, },
        { type: CommonService, },
    ]; };
    return XontVenturaListPromptService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
// right before adding inbuilt prompts and right after adding strict keyword
var /** @type {?} */ XONT_VENTURA_LIST_PROMPT_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(function () { return ListPromptComponent; }),
    multi: true
};
var ListPromptComponent = (function () {
    function ListPromptComponent(listPromptService) {
        this.listPromptService = listPromptService;
        this._originalDataSet = null;
        this._dataSet = null;
        this._minWidth = 500;
        this._dataSliderStatus = 'none';
        this._stillDataLoading = false;
        this._pageSize = 10;
        this._lastEmptiedInputIndex = -1;
        this._currentCursorInputIndex = -1;
        this.inputWidths = [120];
        this.headers = null;
        this.strict = false;
        this.disabled = false;
        this.type = '';
        this.status = 'All';
        this.masterGroup = null;
        //for product prompt when used with a product classificatin selector.
        this.clsGeo = null;
        this.clsRetailer = null;
        this.clsProduct = null;
        this.clsExecutive = null;
        this.territory = null;
        this.ngModelChange = new EventEmitter();
        this.onChanged = new EventEmitter();
        this.onDataBind = new EventEmitter();
        this.dataSourceObservable = null;
        this.filter = '';
    }
    /**
     * @return {?}
     */
    ListPromptComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.ngModelChange.emit(this.ngModel);
        this._minWidth = this.inputWidths.reduce(function (a, b) { return a + b; }, 0) + 15;
        if (!this.gridFields || this.gridFields.length == 0) {
            this.gridFields = this.fields;
        }
        if (!this.gridHeaders || this.gridHeaders.length == 0) {
            this.gridHeaders = this.headers;
        }
        if (!this.dataFields || this.dataFields.length == 0) {
            this.dataFields = this.fields;
        }
        if (!this.modelProps)
            this.modelProps = this.dataFields;
    };
    Object.defineProperty(ListPromptComponent.prototype, "valid", {
        //getters start
        get: /**
         * @return {?}
         */
        function () {
            if (this._dataSliderStatus == 'block')
                return false;
            else if (!this.mandatory)
                return true;
            else
                return !this.isEmpty;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListPromptComponent.prototype, "isEmpty", {
        get: /**
         * @return {?}
         */
        function () {
            for (var /** @type {?} */ inputIndex = 0; inputIndex < this.inputWidths.length; inputIndex++) {
                if (this.ngModel[this.modelProps[inputIndex]])
                    return false;
            }
            return true;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListPromptComponent.prototype, "isCurrentFieldEmpty", {
        get: /**
         * @return {?}
         */
        function () {
            return !this.ngModel[this.modelProps[this._currentCursorInputIndex]];
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ListPromptComponent.prototype.loadTopOne = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.getObservableReady();
        if (this.dataSourceObservable != null) {
            return Observable.create(function (observer) {
                _this.dataSourceObservable.subscribe(function (data) {
                    if (data && data.length > 0) {
                        var /** @type {?} */ keys = Object.keys(data[0]);
                        if (!_this.dataFields || _this.dataFields.length == 0)
                            _this.dataFields = [keys[0], keys[1]];
                        _this.SetModel(data[0]);
                    }
                    else {
                        _this.SetModel(null);
                    }
                }, function (error) { console.error('this is from list prompt, error top one loading', error); }, function () {
                    _this.dataSourceObservable = null;
                    observer.next();
                    observer.complete();
                });
            });
        }
        else {
            console.error("this is from list prompt - Observable data source hasn't been set. did you handle onDataBind of the list prompt?");
        }
    };
    /**
     * @return {?}
     */
    ListPromptComponent.prototype.loadWhenOnlyOne = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.getObservableReady();
        if (this.dataSourceObservable != null) {
            return Observable.create(function (observer) {
                _this.dataSourceObservable.subscribe(function (data) {
                    if (data && data.length > 0 && data.length == 1) {
                        var /** @type {?} */ keys = Object.keys(data[0]);
                        if (!_this.dataFields || _this.dataFields.length == 0)
                            _this.dataFields = [keys[0], keys[1]];
                        _this.SetModel(data[0]);
                    }
                    else {
                        _this.SetModel(null);
                    }
                }, function (error) { console.error('this is from list prompt, error load when only one loading', error); }, function () {
                    _this.dataSourceObservable = null;
                    observer.next();
                    observer.complete();
                });
            });
        }
        else {
            console.error("this is from list prompt - Observable data source hasn't been set. did you handle onDataBind of the list prompt?");
        }
    };
    /**
     * @return {?}
     */
    ListPromptComponent.prototype.loadDefaults = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.getObservableReady();
        if (this.dataSourceObservable != null) {
            return Observable.create(function (observer) {
                _this.dataSourceObservable.subscribe(function (data) {
                    if (data && data.length > 0) {
                        var /** @type {?} */ keys = Object.keys(data[0]);
                        if (!_this.dataFields || _this.dataFields.length == 0)
                            _this.dataFields = [keys[0], keys[1]];
                        _this.SetModel(data[0]);
                    }
                    else {
                        _this.SetModel(null);
                    }
                }, function (error) { console.error('this is from list prompt, error in default loading', error); }, function () {
                    _this.dataSourceObservable = null;
                    observer.next();
                    observer.complete();
                });
            });
        }
        else {
            console.error("this is from list prompt - Observable data source hasn't been set. did you handle onDataBind of the list prompt?");
        }
    };
    /**
     * @return {?}
     */
    ListPromptComponent.prototype.focus = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ self = this;
        /**
         * @return {?}
         */
        function setFocus() {
            if (self.inputElements && self.inputElements.toArray()[0]) {
                self.inputElements.toArray()[0].nativeElement.focus();
            }
            else {
                setTimeout(setFocus, 100);
            }
        }
        setFocus();
    };
    //public methods end
    //Events of the View start
    /**
     * @param {?} e
     * @return {?}
     */
    ListPromptComponent.prototype.pagerClickCapture = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        event.preventDefault();
        event.stopPropagation();
    };
    /**
     * @param {?} e
     * @return {?}
     */
    ListPromptComponent.prototype.ClickedOn = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (e.target.attributes['disabled']) {
            return;
        }
        if (this._dataSliderStatus == 'none') {
            this._dataSliderStatus = 'block';
            this.PopulateDataSet();
        }
        this.inputElements.toArray()[0].nativeElement.focus();
        this._currentCursorInputIndex = 0;
    };
    /**
     * @param {?} e
     * @param {?} inputIndex
     * @return {?}
     */
    ListPromptComponent.prototype.InputMousedIn = /**
     * @param {?} e
     * @param {?} inputIndex
     * @return {?}
     */
    function (e, inputIndex) {
        this._currentCursorInputIndex = inputIndex;
        if (this._dataSliderStatus == 'block') {
            this.FilterSource();
        }
        if (this._dataSliderStatus == 'none') {
            this._dataSliderStatus = 'block';
            this.PopulateDataSet();
        }
    };
    /**
     * @param {?} e
     * @param {?} isLast
     * @param {?} inputIndex
     * @return {?}
     */
    ListPromptComponent.prototype.InputKeyDown = /**
     * @param {?} e
     * @param {?} isLast
     * @param {?} inputIndex
     * @return {?}
     */
    function (e, isLast, inputIndex) {
        if (e.keyCode != 9)
            this._currentCursorInputIndex = inputIndex;
        if (e.keyCode == 32 && this._dataSet == null) {
            e.preventDefault();
            e.stopPropagation();
            return;
        }
        else if (isLast && e.keyCode == 9 && this._dataSliderStatus == 'block') {
            var /** @type {?} */ item = null;
            if (this.strict) {
                if (!this.isCurrentFieldEmpty) {
                    this.FilterSource();
                    item = (this._dataSet && this._dataSet.length > 0) ? this._dataSet[0] : null;
                }
                this.SetModel(item);
            }
            this.onChanged.emit(item);
            this.ResetProps();
        }
    };
    /**
     * @param {?} e
     * @param {?} inputIndex
     * @return {?}
     */
    ListPromptComponent.prototype.InputKeyUp = /**
     * @param {?} e
     * @param {?} inputIndex
     * @return {?}
     */
    function (e, inputIndex) {
        //if not tab apply filter showing data if not shown already
        if (e.keyCode != 9) {
            if (this._dataSliderStatus != 'block')
                this._dataSliderStatus = 'block';
            if (this._originalDataSet == null && !this._stillDataLoading)
                this.PopulateDataSet();
            this.FilterSource();
        }
        if (e.keyCode == 8 && e.target.value == '') {
            if (this._lastEmptiedInputIndex < 0)
                this._lastEmptiedInputIndex = inputIndex;
            else {
                var /** @type {?} */ focusIndex = this._lastEmptiedInputIndex == 0 ? this.inputWidths.length - 1 : inputIndex - 1;
                this.inputElements.toArray()[focusIndex].nativeElement.focus();
                this._lastEmptiedInputIndex = -1;
                this._currentCursorInputIndex = focusIndex;
                this.FilterSource();
            }
        }
    };
    /**
     * @param {?} item
     * @return {?}
     */
    ListPromptComponent.prototype.ItemSelected = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        this.SetModel(item);
        this.onChanged.emit(item);
        this.ResetProps();
    };
    /**
     * @return {?}
     */
    ListPromptComponent.prototype.ClearModel = /**
     * @return {?}
     */
    function () {
        this.SetModel(null);
        this.FilterSource();
        this.inputElements.toArray()[0].nativeElement.focus();
    };
    //events of the View end
    //Misleading the interface start
    /**
     * @param {?} value
     * @return {?}
     */
    ListPromptComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    ListPromptComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    ListPromptComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
    };
    /**
     * @param {?} columns
     * @param {?} prefixWithAlias
     * @return {?}
     */
    ListPromptComponent.prototype.getCommaSepratedList = /**
     * @param {?} columns
     * @param {?} prefixWithAlias
     * @return {?}
     */
    function (columns, prefixWithAlias) {
        var /** @type {?} */ selectQueryColumns = '';
        columns.forEach(function (col) {
            selectQueryColumns += prefixWithAlias + col + ',';
        });
        selectQueryColumns = selectQueryColumns.slice(0, -1);
        return selectQueryColumns;
    };
    /**
     * @return {?}
     */
    ListPromptComponent.prototype.getObservableReady = /**
     * @return {?}
     */
    function () {
        this.onDataBind.emit();
        if (this.type) {
            var /** @type {?} */ typeSwitch = this.type.toLowerCase();
            switch (typeSwitch) {
                case 'product': {
                    var /** @type {?} */ selectQueryColumnsArray = union(this.dataFields, this.gridFields).filter(function (col) { return (col != 'ProductCode' && col != 'Description' && col != 'Status'); });
                    var /** @type {?} */ selectQueryColumns = this.getCommaSepratedList(selectQueryColumnsArray, 'p.');
                    var /** @type {?} */ apiArgs = { strict: this.strict, clsProduct: this.ngModel[this.clsProduct], status: this.status, filter: this.filter, selectQueryColumns: selectQueryColumns };
                    this.dataSourceObservable = this.listPromptService.GetListPromptProductData(apiArgs);
                    if (!this.dataFields || this.dataFields.length == 0)
                        this.dataFields = ['ProductCode', 'Description'];
                    break;
                }
                case 'territory': {
                    //V3008Remove start
                    //let territoryStatus = 'All';
                    //if (this.status) {
                    //    if (this.status.toLowerCase() == 'active')
                    //        territoryStatus = 'Active';
                    //    else if (this.status.toLowerCase() == 'inactive')
                    //        territoryStatus = 'Inactive';
                    //}
                    //let apiArgs = { selector: [{ "MasterGroup": "TETY", "HierarchyRequired": "0" }], activeStatus: territoryStatus, selectedIndex: 0 };
                    //V3008Remove end
                    //V3008Adding start
                    var /** @type {?} */ selectQueryColumnsArray = union(this.dataFields, this.gridFields).filter(function (col) { return (col != 'MasterGroupValue' && col != 'MasterGroupValueDescription'); });
                    var /** @type {?} */ selectQueryColumns = selectQueryColumnsArray.join();
                    var /** @type {?} */ apiArgs = { strict: this.strict, status: this.status, filter: this.filter, selectQueryColumns: selectQueryColumns };
                    //V3008Adding end
                    this.dataSourceObservable = this.listPromptService.GetListPromptTerritoryData(apiArgs);
                    if (!this.dataFields || this.dataFields.length == 0)
                        this.dataFields = ['MasterGroupValue', 'MasterGroupValueDescription'];
                    break;
                }
                case 'executive': {
                    var /** @type {?} */ selectQueryColumnsArray = union(this.dataFields, this.gridFields).filter(function (col) { return (col != 'ExecutiveCode' && col != 'ExecutiveName' && col != 'Status'); });
                    var /** @type {?} */ selectQueryColumns = this.getCommaSepratedList(selectQueryColumnsArray, 'e.');
                    var /** @type {?} */ apiArgs = { strict: this.strict, clsGeo: this.ngModel[this.clsGeo], clsExecutive: this.ngModel[this.clsExecutive], territory: this.ngModel[this.territory], status: this.status, filter: this.filter, selectQueryColumns: selectQueryColumns };
                    this.dataSourceObservable = this.listPromptService.GetListPromptExecutiveData(apiArgs);
                    if (!this.dataFields || this.dataFields.length == 0)
                        this.dataFields = ['ExecutiveCode', 'ExecutiveName'];
                    break;
                }
                case 'retailer': {
                    var /** @type {?} */ selectQueryColumnsArray = union(this.dataFields, this.gridFields).filter(function (col) { return (col != 'RetailerCode' && col != 'RetailerName' && col != 'Status'); });
                    var /** @type {?} */ selectQueryColumns = this.getCommaSepratedList(selectQueryColumnsArray, 'r.');
                    var /** @type {?} */ apiArgs = { strict: this.strict, clsGeo: this.ngModel[this.clsGeo], clsRetailer: this.ngModel[this.clsRetailer], territory: this.ngModel[this.territory], status: this.status, filter: this.filter, selectQueryColumns: selectQueryColumns };
                    this.dataSourceObservable = this.listPromptService.GetListPromptRetailerData(apiArgs);
                    if (!this.dataFields || this.dataFields.length == 0)
                        this.dataFields = ['RetailerCode', 'RetailerName'];
                    break;
                }
                default:
                    break;
            }
        }
        //else {
        //    this.onDataBind.emit();
        //}
    };
    /**
     * @return {?}
     */
    ListPromptComponent.prototype.FilterSource = /**
     * @return {?}
     */
    function () {
        if (this._originalDataSet != null) {
            var /** @type {?} */ FilterColumn_1 = this.dataFields[this._currentCursorInputIndex];
            var /** @type {?} */ FilterValue_1 = this.ngModel[this.modelProps[this._currentCursorInputIndex]];
            if (FilterValue_1)
                FilterValue_1 = FilterValue_1.toString().toUpperCase();
            this._dataSet = filter(this._originalDataSet, function (row) {
                if (FilterColumn_1 && FilterValue_1) {
                    return row[FilterColumn_1].toString().toUpperCase().indexOf(FilterValue_1) > -1;
                }
                else
                    return true;
            });
            this.scrollDiv.nativeElement.scrollTop = 0;
        }
    };
    /**
     * @return {?}
     */
    ListPromptComponent.prototype.ResetProps = /**
     * @return {?}
     */
    function () {
        this._dataSliderStatus = 'none';
        this._dataSet = null;
        this._originalDataSet = null;
        this._lastEmptiedInputIndex = -1;
        this._currentCursorInputIndex = -1;
    };
    /**
     * @param {?} item
     * @return {?}
     */
    ListPromptComponent.prototype.SetModel = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        if (item) {
            for (var /** @type {?} */ fieldIndex = 0; fieldIndex < this.modelProps.length; fieldIndex++) {
                var /** @type {?} */ cellValue = item[this.dataFields[fieldIndex]];
                if (cellValue) {
                    cellValue = cellValue.toString().trim();
                }
                this.ngModel[this.modelProps[fieldIndex]] = cellValue;
            }
        }
        else {
            for (var /** @type {?} */ fieldIndex = 0; fieldIndex < this.modelProps.length; fieldIndex++) {
                this.ngModel[this.modelProps[fieldIndex]] = '';
            }
        }
    };
    /**
     * @return {?}
     */
    ListPromptComponent.prototype.AddEvent = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ self = this;
        /**
         * @param {?} e
         * @return {?}
         */
        function handleClick(e) {
            setTimeout(function () {
                if (!self.element.nativeElement.contains(e.target)) {
                    if (self._dataSliderStatus == 'block') {
                        var /** @type {?} */ item = null;
                        if (self.strict) {
                            if (!self.isCurrentFieldEmpty) {
                                self.FilterSource();
                                item = (self._dataSet && self._dataSet.length > 0) ? self._dataSet[0] : null;
                            }
                            self.SetModel(item);
                        }
                        self.onChanged.emit(item);
                    }
                    self.ResetProps();
                    window.removeEventListener("click", handleClick);
                }
            }, 0);
        }
        window.removeEventListener("click", handleClick);
        window.addEventListener("click", handleClick);
    };
    /**
     * @return {?}
     */
    ListPromptComponent.prototype.PopulateDataSet = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var /** @type {?} */ masterControlData = JSON.parse(localStorage.getItem('PROMPT_MasterControlData'));
        if (masterControlData)
            this._pageSize = masterControlData.AllowPaging === "1" ? masterControlData.PageSize : masterControlData.ExtendedPageSize;
        this._stillDataLoading = true;
        this.onDataBind.emit();
        //prompt data is provided by the developer himself.
        if (this.dataSourceObservable) {
            var /** @type {?} */ subsription = this.dataSourceObservable.subscribe(function (data) {
                if (data && data.length > 0) {
                    var /** @type {?} */ keys = Object.keys(data[0]);
                    if (!_this.dataFields || _this.dataFields.length == 0)
                        _this.dataFields = [keys[0], keys[1]];
                    if (!_this.gridFields || _this.gridFields.length == 0)
                        _this.gridFields = _this.dataFields;
                }
                _this._dataSet = data;
                _this._originalDataSet = data;
            }, function (error) { console.error('this is from list prompt - error in developer given observable', error); }, function () { _this._stillDataLoading = false; });
        }
        else if (this.masterGroup) {
            var /** @type {?} */ groupStatus = 'All';
            if (this.status) {
                if (this.status.toLowerCase() == 'active')
                    groupStatus = 'Active';
                else if (this.status.toLowerCase() == 'inactive')
                    groupStatus = 'Inactive';
            }
            var /** @type {?} */ apiArgs = { selector: [{ "MasterGroup": this.masterGroup, "HierarchyRequired": "0" }], activeStatus: groupStatus, selectedIndex: 0 };
            this.listPromptService.GetMasterDefinitionValues(apiArgs).subscribe(function (data) {
                _this._dataSet = data;
                _this._originalDataSet = data;
                if (data && data.length > 0) {
                    if (!_this.dataFields || _this.dataFields.length == 0)
                        _this.dataFields = ['MasterGroupValue', 'MasterGroupValueDescription'];
                    if (!_this.gridFields || _this.gridFields.length == 0)
                        _this.gridFields = _this.dataFields;
                    if (!_this.gridHeaders || _this.gridHeaders.length == 0) {
                        _this.gridHeaders = ['Code', 'Description'];
                    }
                }
            }, function (error) { console.error('this is from list prompt- loading inbuilt MasterGroup prompt', error); }, function () { _this._stillDataLoading = false; });
        }
        else if (this.predefinedPromptArgs && this.predefinedPromptArgs.length > 0) {
            var /** @type {?} */ apiArgs = { ListFromID: this.predefinedPromptArgs[0], ExternalParameters: this.predefinedPromptArgs[1] || '' };
            this.listPromptService.GetPredefinedPromptData(apiArgs)
                .subscribe(function (data) {
                if (data && data.length > 0) {
                    var /** @type {?} */ keys = Object.keys(data[0]);
                    if (!_this.dataFields || _this.dataFields.length == 0)
                        _this.dataFields = [keys[0], keys[1]];
                    if (!_this.gridFields || _this.gridFields.length == 0)
                        _this.gridFields = _this.dataFields;
                }
                _this._dataSet = data;
                _this._originalDataSet = data;
            }, function (error) { console.error('this is from list prompt - loading from prompt predefind prompt args', error); }, function () { _this._stillDataLoading = false; });
        }
        else if (this.type) {
            var /** @type {?} */ typeSwitch = this.type.toLowerCase();
            switch (typeSwitch) {
                case 'product': {
                    var /** @type {?} */ selectQueryColumnsArray = union(this.dataFields, this.gridFields).filter(function (col) { return (col != 'ProductCode' && col != 'Description' && col != 'Status'); });
                    var /** @type {?} */ selectQueryColumns = this.getCommaSepratedList(selectQueryColumnsArray, 'p.');
                    var /** @type {?} */ apiArgs_1 = { strict: this.strict, clsProduct: this.ngModel[this.clsProduct], status: this.status, filter: this.filter, selectQueryColumns: selectQueryColumns };
                    this.listPromptService.GetListPromptProductData(apiArgs_1).subscribe(function (data) {
                        _this._dataSet = data;
                        _this._originalDataSet = data;
                        if (data && data.length > 0) {
                            var /** @type {?} */ keys = Object.keys(data[0]);
                            if (!_this.dataFields || _this.dataFields.length == 0)
                                _this.dataFields = ['ProductCode', 'Description'];
                            if (!_this.gridFields || _this.gridFields.length == 0)
                                _this.gridFields = _this.dataFields;
                            if (!_this.gridHeaders || _this.gridHeaders.length == 0) {
                                _this.gridHeaders = ['Product Code', 'Description'];
                            }
                        }
                    }, function (error) { console.error('this is from list prompt - loading inbuilt prompt product', error, apiArgs_1); }, function () { _this._stillDataLoading = false; });
                    break;
                }
                case 'territory': {
                    //V3008Remove start
                    //let territoryStatus = 'All';
                    //if (this.status) {
                    //    if (this.status.toLowerCase() == 'active')
                    //        territoryStatus = 'Active';
                    //    else if (this.status.toLowerCase() == 'inactive')
                    //        territoryStatus = 'Inactive';
                    //}
                    //let apiArgs = { selector: [{ "MasterGroup": "TETY", "HierarchyRequired": "0" }], activeStatus: territoryStatus, selectedIndex: 0 };
                    //V3008Remove end
                    //V3008Adding start
                    var /** @type {?} */ selectQueryColumnsArray = union(this.dataFields, this.gridFields).filter(function (col) { return (col != 'MasterGroupValue' && col != 'MasterGroupValueDescription'); });
                    var /** @type {?} */ selectQueryColumns = selectQueryColumnsArray.join();
                    var /** @type {?} */ apiArgs = { strict: this.strict, status: this.status, filter: this.filter, selectQueryColumns: selectQueryColumns };
                    //V3008Adding end
                    this.listPromptService.GetListPromptTerritoryData(apiArgs).subscribe(function (data) {
                        _this._dataSet = data;
                        _this._originalDataSet = data;
                        if (data && data.length > 0) {
                            if (!_this.dataFields || _this.dataFields.length == 0)
                                _this.dataFields = ['MasterGroupValue', 'MasterGroupValueDescription'];
                            if (!_this.gridFields || _this.gridFields.length == 0)
                                _this.gridFields = _this.dataFields;
                            if (!_this.gridHeaders || _this.gridHeaders.length == 0) {
                                _this.gridHeaders = ['Territory Code', 'Description'];
                            }
                        }
                    }, function (error) { console.error('this is from list prompt- loading inbuilt prompt territory', error); }, function () { _this._stillDataLoading = false; });
                    break;
                }
                case 'executive': {
                    var /** @type {?} */ selectQueryColumnsArray = union(this.dataFields, this.gridFields).filter(function (col) { return (col != 'ExecutiveCode' && col != 'ExecutiveName' && col != 'Status'); });
                    var /** @type {?} */ selectQueryColumns = this.getCommaSepratedList(selectQueryColumnsArray, 'e.');
                    var /** @type {?} */ apiArgs_2 = { strict: this.strict, clsGeo: this.ngModel[this.clsGeo], clsExecutive: this.ngModel[this.clsExecutive], territory: this.ngModel[this.territory], status: this.status, filter: this.filter, selectQueryColumns: selectQueryColumns };
                    this.listPromptService.GetListPromptExecutiveData(apiArgs_2).subscribe(function (data) {
                        _this._dataSet = data;
                        _this._originalDataSet = data;
                        if (data && data.length > 0) {
                            var /** @type {?} */ keys = Object.keys(data[0]);
                            if (!_this.dataFields || _this.dataFields.length == 0)
                                _this.dataFields = ['ExecutiveCode', 'ExecutiveName'];
                            if (!_this.gridFields || _this.gridFields.length == 0)
                                _this.gridFields = _this.dataFields;
                            if (!_this.gridHeaders || _this.gridHeaders.length == 0) {
                                _this.gridHeaders = ['Executive Code', 'Executive Name'];
                            }
                        }
                    }, function (error) { console.error('this is from list prompt - loading inbuilt prompt executive ', error, apiArgs_2); }, function () { _this._stillDataLoading = false; });
                    break;
                }
                case 'retailer': {
                    var /** @type {?} */ selectQueryColumnsArray = union(this.dataFields, this.gridFields).filter(function (col) { return (col != 'RetailerCode' && col != 'RetailerName' && col != 'Status'); });
                    var /** @type {?} */ selectQueryColumns = this.getCommaSepratedList(selectQueryColumnsArray, 'r.');
                    var /** @type {?} */ apiArgs_3 = { strict: this.strict, clsGeo: this.ngModel[this.clsGeo], clsRetailer: this.ngModel[this.clsRetailer], territory: this.ngModel[this.territory], status: this.status, filter: this.filter, selectQueryColumns: selectQueryColumns };
                    this.listPromptService.GetListPromptRetailerData(apiArgs_3).subscribe(function (data) {
                        _this._dataSet = data;
                        _this._originalDataSet = data;
                        if (data && data.length > 0) {
                            var /** @type {?} */ keys = Object.keys(data[0]);
                            if (!_this.dataFields || _this.dataFields.length == 0)
                                _this.dataFields = ['RetailerCode', 'RetailerName'];
                            if (!_this.gridFields || _this.gridFields.length == 0)
                                _this.gridFields = _this.dataFields;
                            if (!_this.gridHeaders || _this.gridHeaders.length == 0) {
                                _this.gridHeaders = ['Retailer Code', 'Retailer Name'];
                            }
                        }
                    }, function (error) { console.error('this is from list prompt - loading inbuilt prompt retailer', error, apiArgs_3); }, function () { _this._stillDataLoading = false; });
                    break;
                }
                default:
                    break;
            }
        }
        this.AddEvent();
    };
    ListPromptComponent.decorators = [
        { type: Component, args: [{
                    selector: 'xont-ventura-list-prompt',
                    styles: [
                        '.Textboxstyle  { padding-left:2px;}',
                        ".dataSlider {\n            position: absolute;\n            min-height:320px;\n            border-radius: 0 7px 7px 7px;\n            background-color:white;\n            margin-top: 3px;\n            margin-bottom:15px;\n            z-index:1000;\n            border: 1.5px #5399d6 solid;\n            padding-left: 5px;\n            color: #01539D;\n            font-family: Trebuchet MS;\n            background-color:ghostwhite;\n         }",
                        '.dtRow { cursor:pointer; }',
                        '.dtRow > td {padding-right:10px;}',
                        '.dtRow:hover {color:red;}',
                        '.pmtPaginator { position:absolute; bottom:4px; border:2px solid #006699; border-radius:6px;}',
                        '.sliderLoader {height: 80px; margin: auto; display: block; margin-top: 80px;}',
                        '.clickButton {padding-left:5px;font-size:16px;padding-right:5px;cursor:pointer;display:inline-block;}',
                        '.scrollDiv { overflow: auto; overflow-x: hidden; padding-left:10px; margin:8px 5px 5px 0;}',
                        '.closeI {font-size:18px;cursor:pointer;color:#006699;vertical-align:middle;}',
                        '.Textboxstyle[disabled] {background-color: rgba(170, 170, 170, 0.19);}',
                        '.disabledInput {background-color: rgba(170, 170, 170, 0.19);}',
                        '.clickButtonDisabled {padding-left:5px;font-size:16px;padding-right:5px;display:inline-block;}',
                        '.tableStyles tbody td { padding: 2.5px 0 2.5px 3.5px;border: 1px solid #AAAAAA;background:white;font-size:small;font-family:Trebuchet MS;font-size:13px;}',
                        '.headerTable {margin-left:10px;margin-top:4px; background-color: #006699;color: #ffffff;}',
                        '.tableStyles th {padding: 2.5px 0 2.5px 3.5px;font-family:Trebuchet MS; font-size:13px;margin-left:10px;margin-top:4px; background-color: #006699;color: #ffffff;}',
                        '.tableStyles th:hover {text-decoration:underline;cursor:pointer;}',
                        '.sortIcon {padding-left:7px;}',
                        '.pmtPaginator > span {padding-left:5px;padding-right:5px;}',
                        '.lmptErrormessagetextstyle {position:absolute;font-size:20px;margin-left:3px;font-weight:500;color:#e50505;font-family:Trebuchet MS;top:-14px;left:-2.5px;}',
                        '.MandatoryCurrent {top:1px;left:17px;}'
                    ],
                    template: "<div #element>\n                <ng-template ngFor let-input [ngForOf]=\"inputWidths\" let-inputIndex=\"index\"\n                        let-isFirst=\"first\" let-isLast=\"last\">\n                    <input #inputElement autocomplete=\"off\"\n                        [(ngModel)]=\"ngModel[modelProps[inputIndex]]\"\n                        [disabled]=\"disabled\" [maxlength]=\"maxLengths[inputIndex]\"\n                        [class.disabledInput]=\"disabled\" \n                        [style.width]=\"inputWidths[inputIndex] + 'px'\" \n                        (mousedown)=\"InputMousedIn($event, inputIndex)\"\n                        (keyup)=\"InputKeyUp($event, inputIndex)\" \n                        (keydown)=\"InputKeyDown($event, isLast, inputIndex)\"\n                        class=\"Textboxstyle\"/>\n                    <span *ngIf=\"isFirst\">\n                        <div (click)=\"ClickedOn($event)\" *ngIf=\"!disabled\" class=\"fa fa-angle-double-down clickButton\" aria-hidden=\"true\"></div>\n                        <div class=\"fa fa-angle-double-down clickButtonDisabled\" *ngIf=\"disabled\" aria-hidden=\"true\"></div>\n                    </span>\n                    <span *ngIf=\"isLast\" style=\"position:relative;display:inline-block;\">\n                        <i class=\"fa fa-times-circle closeI\" *ngIf=\"_dataSliderStatus == 'block'\" title=\"clear code/description\" (click)=\"ClearModel()\"></i>\n                        <small *ngIf=\"mandatory && !valid\" [class.MandatoryCurrent]=\"_currentCursorInputIndex >= 0\" class=\"lmptErrormessagetextstyle\">*</small>\n                    </span>\n                </ng-template>\n                    <div class=\"dataSlider\" [style.display]=\"_dataSliderStatus\" [style.minWidth]=\"_minWidth + 60 + 'px'\"> \n                        <div *ngIf=\"_dataSet == null\">\n                            <img class=\"sliderLoader\" src='../App_Themes_V3/Blue/images/load_pmt.gif'/>\n                        </div>\n\n                        <div #scrollDiv [style.height]=\"_dataSet?.length > _pageSize ? '270px' : '275px'\" class=\"scrollDiv\" *ngIf= \"!_stillDataLoading\">\n                            <table [mfRowsOnPage]=\"_pageSize\" [style.minWidth]=\"_minWidth + 20 + 'px'\" [mfData]=\"_dataSet\" #mf=\"mfDataTable\" \n                        class=\"tableStyles\">\n                            \n                            <thead *ngIf=\"gridHeaders?.length > 0 && gridFields?.length > 0\">\n                                <tr>\n                                    <th *ngFor=\"let header of gridHeaders let gridHeaderIndex = index\">\n                                        <mfDefaultSorter [by]=\"gridFields[gridHeaderIndex]\">{{header}}</mfDefaultSorter>\n                                    </th>\n                                </tr>\n                            </thead>\n\n                            <tbody>\n                                <tr *ngFor=\"let item of mf.data\" class=\"dtRow\" (click)=\"ItemSelected(item)\" >\n                                    <td *ngFor=\"let field of gridFields let fieldIndex = index\">{{item[field].toString().trim()}}</td>\n                                </tr>\n                            </tbody>\n                             <tfoot>\n                              <div *ngIf=\"_dataSet?.length > _pageSize || _dataSet?.length == 0\">\n                                <div class=\"pmtPaginator\" (click)=\"pagerClickCapture($event)\" *ngIf=\"_dataSet?.length > _pageSize\">\n                                    <mfBootstrapPaginator></mfBootstrapPaginator>\n                                </div>\n                                <div class=\"pmtPaginator\" *ngIf= \"_dataSet?.length == 0\">\n                                    <span>No Data Found to Display</span>\n                                </div>\n                              </div>\n\n                            </tfoot>\n                            </table>\n                        <div>\n                    </div>\n                </div>",
                    providers: [XONT_VENTURA_LIST_PROMPT_VALUE_ACCESSOR]
                },] },
    ];
    //Private Methods End
    /** @nocollapse */
    ListPromptComponent.ctorParameters = function () { return [
        { type: XontVenturaListPromptService, },
    ]; };
    ListPromptComponent.propDecorators = {
        "maxLengths": [{ type: Input },],
        "inputWidths": [{ type: Input },],
        "fields": [{ type: Input },],
        "dataFields": [{ type: Input },],
        "gridFields": [{ type: Input },],
        "gridHeaders": [{ type: Input },],
        "modelProps": [{ type: Input },],
        "headers": [{ type: Input },],
        "mandatory": [{ type: Input },],
        "strict": [{ type: Input },],
        "disabled": [{ type: Input },],
        "ngModel": [{ type: Input },],
        "type": [{ type: Input },],
        "status": [{ type: Input },],
        "predefinedPromptArgs": [{ type: Input },],
        "masterGroup": [{ type: Input },],
        "clsGeo": [{ type: Input },],
        "clsRetailer": [{ type: Input },],
        "clsProduct": [{ type: Input },],
        "clsExecutive": [{ type: Input },],
        "territory": [{ type: Input },],
        "ngModelChange": [{ type: Output },],
        "onChanged": [{ type: Output },],
        "onDataBind": [{ type: Output },],
        "element": [{ type: ViewChild, args: ['element',] },],
        "scrollDiv": [{ type: ViewChild, args: ['scrollDiv',] },],
        "inputElements": [{ type: ViewChildren, args: ['inputElement',] },],
    };
    return ListPromptComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
// right before adding inbuilt prompts and right after adding strict keyword
var /** @type {?} */ XONT_VENTURA_LISTPROMPT_MULTIPLE_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(function () { return ListPromptMultipleComponent; }),
    multi: true
};
var ListPromptMultipleComponent = (function () {
    function ListPromptMultipleComponent(listPromptService) {
        this.listPromptService = listPromptService;
        this._originalDataSet = null;
        this._dataSet = null;
        this._minWidth = 500;
        this._dataSliderStatus = 'none';
        this._stillDataLoading = false;
        this._pageSize = 10;
        this._lastEmptiedInputIndex = -1;
        this._currentCursorInputIndex = -1;
        this._activePage = 1;
        this._originalPageSize = 10;
        this._textBoxesHeight = '40px';
        this.isAllChecked = false;
        this.isRangeApplied = false;
        this.selectMode = '';
        this.selectedRows = [];
        this.isSelectedOnlyMode = false;
        this._searchTypeClass = 'fa fa-indent';
        this._currentSortBy = '';
        this._currentSortOrder = 'asc';
        this.selectedModeProp = '';
        this.selectedRowsProp = '';
        this.uniqueColumn = '';
        this.inputWidths = [120];
        this.disabled = false;
        this.type = '';
        this.status = 'All';
        this.masterGroup = null;
        //for product prompt when used with a product classificatin selector.
        this.clsGeo = null;
        this.clsRetailer = null;
        this.clsProduct = null;
        this.clsExecutive = null;
        this.territory = null;
        this.ngModelChange = new EventEmitter();
        this.onChanged = new EventEmitter();
        this.onDataBind = new EventEmitter();
        this.dataSourceObservable = null;
        this.filter = '';
    }
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.ngModelChange.emit(this.ngModel);
        this._minWidth = this.inputWidths.reduce(function (a, b) { return a + b; }, 0) + 15;
        if (!this.modelProps)
            this.modelProps = this.dataFields;
        if (localStorage.getItem('MultiplePromptSearchTypeClass')) {
            var /** @type {?} */ lastUsedSearchTypeClass = JSON.parse(localStorage.getItem('MultiplePromptSearchTypeClass'));
            if (lastUsedSearchTypeClass)
                this._searchTypeClass = lastUsedSearchTypeClass;
        }
    };
    /**
     * @param {?} orderBy
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.sortByChanged = /**
     * @param {?} orderBy
     * @return {?}
     */
    function (orderBy$$1) {
        this._currentSortBy = orderBy$$1;
    };
    /**
     * @param {?} orderType
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.sortOrderChanged = /**
     * @param {?} orderType
     * @return {?}
     */
    function (orderType) {
        this._currentSortOrder = orderType;
    };
    /**
     * @param {?} item
     * @param {?} field
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.giveMeTheCellValue = /**
     * @param {?} item
     * @param {?} field
     * @return {?}
     */
    function (item, field) {
        var /** @type {?} */ FilterColumn = this.dataFields[this._currentCursorInputIndex];
        var /** @type {?} */ FilterValue = this.ngModel[this.modelProps[this._currentCursorInputIndex]];
        var /** @type {?} */ FilterValueUpper = '';
        if (this._searchTypeClass == "fa fa-indent" && FilterValue && field == FilterColumn) {
            FilterValueUpper = FilterValue.toString().toUpperCase();
            var /** @type {?} */ regEx = new RegExp("(" + FilterValue + ")", 'gi');
            return item[FilterColumn].replace(regEx, '<mark>$1</mark>').toString().trim();
        }
        else {
            return item[field].toString().trim();
        }
    };
    Object.defineProperty(ListPromptMultipleComponent.prototype, "valid", {
        //getters start
        get: /**
         * @return {?}
         */
        function () {
            if (this._dataSliderStatus == 'block')
                return false;
            else if (!this.mandatory)
                return true;
            else
                return !this.isEmpty;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListPromptMultipleComponent.prototype, "isEmpty", {
        get: /**
         * @return {?}
         */
        function () {
            for (var /** @type {?} */ inputIndex = 0; inputIndex < this.inputWidths.length; inputIndex++) {
                if (this.ngModel[this.modelProps[inputIndex]])
                    return false;
            }
            return true;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListPromptMultipleComponent.prototype, "isCurrentFieldEmpty", {
        get: /**
         * @return {?}
         */
        function () {
            return !this.ngModel[this.modelProps[this._currentCursorInputIndex]];
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.focus = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ self = this;
        /**
         * @return {?}
         */
        function setFocus() {
            if (self.inputElements && self.inputElements.toArray()[0]) {
                self.inputElements.toArray()[0].nativeElement.focus();
            }
            else {
                setTimeout(setFocus, 100);
            }
        }
        setFocus();
    };
    //public methods end
    //multiple prompt
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.rowCountClick = /**
     * @return {?}
     */
    function () {
        this.isSelectedOnlyMode = !this.isSelectedOnlyMode;
        if (this.isSelectedOnlyMode) {
            this._dataSet = this.selectedRows;
            this._pageSize = 50000;
        }
        else {
            this._dataSet = this._originalDataSet;
            this._pageSize = this._originalPageSize;
        }
    };
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.rowCountMouseEnter = /**
     * @return {?}
     */
    function () {
        if (this.selectedRows.length <= 5000) {
            this._dataSet = this.selectedRows;
            this._pageSize = 50000;
        }
    };
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.rowCountMouseLeave = /**
     * @return {?}
     */
    function () {
        if (this.selectedRows.length <= 5000 && !this.isSelectedOnlyMode) {
            this._dataSet = this._originalDataSet;
            this._pageSize = this._originalPageSize;
        }
    };
    /**
     * @param {?} e
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.pagerClickCapture = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        event.preventDefault();
        event.stopPropagation();
    };
    //multiple prompt
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.CheckAll = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.isAllChecked = !this.isAllChecked;
        if (this.isAllChecked) {
            this.selectedRows = [];
            this._dataSet.forEach(function (item) {
                item.selected = true;
                _this.selectedRows.push(item);
            });
        }
        else {
            this._dataSet.forEach(function (item) { return item.selected = false; });
            this.selectedRows = [];
        }
    };
    //multiple prompt
    /**
     * @param {?} event
     * @param {?} item
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.CheckChanged = /**
     * @param {?} event
     * @param {?} item
     * @return {?}
     */
    function (event, item) {
        var _this = this;
        if (!this.isRangeApplied) {
            item.selected = !!!item.selected;
            if (item.selected) {
                this.selectedRows.push(item);
            }
            else {
                this.isAllChecked = false;
                var /** @type {?} */ itemIndex = this.selectedRows.findIndex(function (selected) { return selected[_this.dataFields[0]].trim() == item[_this.dataFields[0]].trim(); });
                this.selectedRows.splice(itemIndex, 1);
            }
        }
        event.stopPropagation();
    };
    //multiple prompt
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.ClearSearchBoxes_Click = /**
     * @return {?}
     */
    function () {
        this.ClearModel();
        this.FilterSource();
        this.inputElements.toArray()[0].nativeElement.focus();
    };
    //multiple prompt
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.ApplyRange = /**
     * @return {?}
     */
    function () {
        if (this._currentSortBy == this.uniqueColumn && this._currentSortOrder == "asc") {
            this.isRangeApplied = !this.isRangeApplied;
            this.ClearModel();
            this.FilterSource();
            this.markRange();
        }
        else {
            var /** @type {?} */ notPermittedDiv_1 = document.createElement('div');
            notPermittedDiv_1.id = "mlptApplyRangeNotPermittedDiv";
            notPermittedDiv_1.innerText = 'Must be ordered by unique column in ascending.';
            notPermittedDiv_1.style.position = 'absolute';
            notPermittedDiv_1.style.bottom = '12%';
            notPermittedDiv_1.style.right = '2%';
            notPermittedDiv_1.style.padding = '13px';
            notPermittedDiv_1.style.backgroundColor = 'white';
            notPermittedDiv_1.style.border = '#7DB9E8 2px solid';
            notPermittedDiv_1.style.backgroundColor = 'white';
            notPermittedDiv_1.style.borderRadius = '20px';
            this.divRangeArea.nativeElement.appendChild(notPermittedDiv_1);
            var /** @type {?} */ opacity_1 = 1;
            var /** @type {?} */ fadeEffect = setInterval(function () {
                if (opacity_1 > 0) {
                    opacity_1 -= 0.08;
                    notPermittedDiv_1.style.opacity = opacity_1.toString();
                }
                else {
                    clearInterval(fadeEffect);
                    notPermittedDiv_1.parentElement.removeChild(notPermittedDiv_1);
                }
            }, 200);
        }
    };
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.btnOK_click = /**
     * @return {?}
     */
    function () {
        this.SetModel();
        this.ResetProps();
        var /** @type {?} */ checkedMarkedRows = this.selectedRows.filter(function (item) { return item.selected; });
        var /** @type {?} */ eventArgs = { Data: checkedMarkedRows, Type: this.selectMode };
        this.onChanged.emit(eventArgs);
    };
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.btnCancel_click = /**
     * @return {?}
     */
    function () {
        this.ClearModel();
        if (this.selectedModeProp) {
            this.ngModel[this.selectedModeProp] = '';
        }
        if (this.selectedRowsProp) {
            this.ngModel[this.selectedRowsProp] = null;
        }
        this.ResetProps();
        this.isRangeApplied = false;
        this.isAllChecked = false;
        this.selectedRows = [];
        var /** @type {?} */ eventArgs = { Data: null, Type: '' };
        this.onChanged.emit(eventArgs);
    };
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.reset = /**
     * @return {?}
     */
    function () {
        this.btnCancel_click();
    };
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.iSearchType_Click = /**
     * @return {?}
     */
    function () {
        this._searchTypeClass = this._searchTypeClass == "fa fa-filter" ? "fa fa-indent" : "fa fa-filter";
        localStorage.setItem('MultiplePromptSearchTypeClass', JSON.stringify(this._searchTypeClass));
        this.focus();
        this._dataSet = this._originalDataSet;
        this.FilterSource();
    };
    //multiple prompt specific end
    /**
     * @param {?} e
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.ClickedOn = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        //V3009Adding start
        if (e.target.attributes['disabled']) {
            return;
        }
        //V3009Adding end
        if (this._dataSliderStatus == 'none') {
            this.initiateUI();
        }
        this.inputElements.toArray()[0].nativeElement.focus();
        this._currentCursorInputIndex = 0;
    };
    /**
     * @param {?} e
     * @param {?} inputIndex
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.InputMousedIn = /**
     * @param {?} e
     * @param {?} inputIndex
     * @return {?}
     */
    function (e, inputIndex) {
        this._currentCursorInputIndex = inputIndex;
        if (this._dataSliderStatus == 'block') {
            this.FilterSource();
        }
        if (this._dataSliderStatus == 'none') {
            this.initiateUI();
        }
    };
    /**
     * @param {?} e
     * @param {?} isLast
     * @param {?} inputIndex
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.InputKeyDown = /**
     * @param {?} e
     * @param {?} isLast
     * @param {?} inputIndex
     * @return {?}
     */
    function (e, isLast, inputIndex) {
        if (e.keyCode != 9)
            this._currentCursorInputIndex = inputIndex;
        if (e.keyCode == 32 && this._dataSet == null) {
            e.preventDefault();
            e.stopPropagation();
            return;
        }
    };
    /**
     * @param {?} e
     * @param {?} inputIndex
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.InputKeyUp = /**
     * @param {?} e
     * @param {?} inputIndex
     * @return {?}
     */
    function (e, inputIndex) {
        //if not tab apply filter showing data if not shown already
        if (e.keyCode != 9) {
            if (this._dataSliderStatus != 'block') {
                this._dataSliderStatus = 'block';
                this._textBoxesHeight = '24px';
            }
            if (this._originalDataSet == null && !this._stillDataLoading)
                this.PopulateDataSet();
            this.FilterSource();
        }
        if (e.keyCode == 8 && e.target.value == '') {
            if (this._lastEmptiedInputIndex < 0)
                this._lastEmptiedInputIndex = inputIndex;
            else {
                var /** @type {?} */ focusIndex = this._lastEmptiedInputIndex == 0 ? this.inputWidths.length - 1 : inputIndex - 1;
                if (this.inputElements.toArray()[focusIndex] && this.inputElements.toArray()[focusIndex].nativeElement) {
                    this.inputElements.toArray()[focusIndex].nativeElement.focus();
                    this._lastEmptiedInputIndex = -1;
                    this._currentCursorInputIndex = focusIndex;
                }
                this.FilterSource();
            }
        }
    };
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.ClearModel = /**
     * @return {?}
     */
    function () {
        for (var /** @type {?} */ fieldIndex = 0; fieldIndex < this.modelProps.length; fieldIndex++) {
            this.ngModel[this.modelProps[fieldIndex]] = '';
        }
    };
    //events of the View end
    //Misleading the interface start
    /**
     * @param {?} value
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
    };
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.initiateUI = /**
     * @return {?}
     */
    function () {
        this._dataSliderStatus = 'block';
        this._textBoxesHeight = '24px';
        this._currentSortOrder = 'asc';
        if (this.uniqueColumn)
            this._currentSortBy = this.uniqueColumn;
        else if (this.dataFields && this.dataFields[0])
            this._currentSortBy = this.dataFields[0];
        var /** @type {?} */ lastUsedSearchTypeClass = JSON.parse(localStorage.getItem('MultiplePromptSearchTypeClass'));
        if (lastUsedSearchTypeClass)
            this._searchTypeClass = lastUsedSearchTypeClass;
        this.PopulateDataSet();
    };
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.markRange = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var /** @type {?} */ selectedTwo = this._dataSet.filter(function (item) { return item.selected; });
        var /** @type {?} */ firstIndex = this._dataSet.findIndex(function (selected) { return selected[_this.dataFields[0]].trim() == selectedTwo[0][_this.dataFields[0]].trim(); });
        var /** @type {?} */ lastIndex = this._dataSet.findIndex(function (selected) { return selected[_this.dataFields[0]].trim() == selectedTwo[1][_this.dataFields[0]].trim(); });
        if (this.isRangeApplied) {
            var /** @type {?} */ selectedRows = [];
            for (var /** @type {?} */ index = firstIndex; index <= lastIndex; index++) {
                var /** @type {?} */ currentItem = this._dataSet[index];
                currentItem.InRange = true;
                selectedRows.push(currentItem);
            }
            this.selectedRows = selectedRows;
        }
        else {
            for (var /** @type {?} */ index = firstIndex; index <= lastIndex; index++) {
                var /** @type {?} */ currentItem = this._dataSet[index];
                currentItem.InRange = false;
            }
            this.selectedRows = selectedTwo;
        }
    };
    /**
     * @param {?} columns
     * @param {?} prefixWithAlias
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.getCommaSepratedList = /**
     * @param {?} columns
     * @param {?} prefixWithAlias
     * @return {?}
     */
    function (columns, prefixWithAlias) {
        var /** @type {?} */ selectQueryColumns = '';
        columns.forEach(function (col) {
            selectQueryColumns += prefixWithAlias + col + ',';
        });
        selectQueryColumns = selectQueryColumns.slice(0, -1);
        return selectQueryColumns;
    };
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.FilterSource = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this._originalDataSet != null) {
            var /** @type {?} */ FilterColumn_1 = this.dataFields[this._currentCursorInputIndex];
            var /** @type {?} */ FilterValue_1 = this.ngModel[this.modelProps[this._currentCursorInputIndex]];
            var /** @type {?} */ FilterValueUpper_1 = '';
            if (FilterValue_1)
                FilterValueUpper_1 = FilterValue_1.toString().toUpperCase();
            if (this._searchTypeClass == "fa fa-indent") {
                var /** @type {?} */ firstFoundIndex = 1;
                //in selected only mode selectedRows should be filtered
                if (this.isSelectedOnlyMode) {
                    firstFoundIndex = this.selectedRows.findIndex(function (item) { return item[FilterColumn_1].toString().toUpperCase().indexOf(FilterValueUpper_1) > -1; });
                }
                else {
                    var /** @type {?} */ orderedDataSet = null;
                    if (this._currentSortBy && this._currentSortOrder)
                        orderedDataSet = orderBy(this._dataSet, [function (i) { return i[_this._currentSortBy].toLowerCase(); }], [this._currentSortOrder]);
                    else
                        orderedDataSet = this._dataSet;
                    firstFoundIndex = orderedDataSet.findIndex(function (item) { return item[FilterColumn_1].toString().toUpperCase().indexOf(FilterValueUpper_1) > -1; });
                }
                var /** @type {?} */ firstFoundIndexPageNumb = Math.floor(firstFoundIndex / this._pageSize) + 1;
                this._activePage = firstFoundIndexPageNumb;
                if (this._activePage > 0) {
                    this.gridTable.setPage(this._activePage, this._pageSize);
                }
                //setting scroll position
                var /** @type {?} */ scrollingIndex = firstFoundIndex % this._pageSize;
                var /** @type {?} */ scrollingPx = 23 * scrollingIndex;
                this.scrollDiv.nativeElement.scrollTop = scrollingPx;
            }
            else {
                if (this.isSelectedOnlyMode) {
                    this._dataSet = filter(this.selectedRows, function (row) {
                        if (FilterColumn_1 && FilterValue_1) {
                            return row[FilterColumn_1].toString().toUpperCase().indexOf(FilterValueUpper_1) > -1;
                        }
                        else
                            return true;
                    });
                }
                else {
                    this._dataSet = filter(this._originalDataSet, function (row) {
                        if (FilterColumn_1 && FilterValue_1) {
                            return row[FilterColumn_1].toString().toUpperCase().indexOf(FilterValueUpper_1) > -1;
                        }
                        else
                            return true;
                    });
                }
                this.scrollDiv.nativeElement.scrollTop = 0;
            }
        }
    };
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.ResetProps = /**
     * @return {?}
     */
    function () {
        this._dataSliderStatus = 'none';
        this._textBoxesHeight = '40px';
        this._dataSet = null;
        this._originalDataSet = null;
        this._lastEmptiedInputIndex = -1;
        this._currentCursorInputIndex = -1;
        this.isSelectedOnlyMode = false;
    };
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.SetModel = /**
     * @return {?}
     */
    function () {
        //setting the select mode
        if (this.isRangeApplied) {
            this.selectMode = 'range';
        }
        else if (this.selectedRows.length > 0) {
            if (this.selectedRows.length == 1) {
                this.selectMode = 'single';
            }
            else {
                this.selectMode = 'multiple';
            }
        }
        var /** @type {?} */ selectedCheckedRows = this.selectedRows.filter(function (item) { return item.selected; });
        var _loop_1 = function (fieldIndex) {
            var /** @type {?} */ fieldName = this_1.dataFields[fieldIndex];
            var /** @type {?} */ textValue = '';
            selectedCheckedRows.forEach(function (item) {
                textValue += item[fieldName].trim() + ', ';
            });
            textValue = textValue.replace(/,\s*$/, "");
            this_1.ngModel[this_1.modelProps[fieldIndex]] = textValue;
        };
        var this_1 = this;
        for (var /** @type {?} */ fieldIndex = 0; fieldIndex < this.modelProps.length; fieldIndex++) {
            _loop_1(fieldIndex);
        }
        if (this.selectedModeProp)
            this.ngModel[this.selectedModeProp] = this.selectMode;
        if (this.selectedRowsProp)
            this.ngModel[this.selectedRowsProp] = selectedCheckedRows;
    };
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.PopulateDataSet = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var /** @type {?} */ masterControlData = JSON.parse(localStorage.getItem('PROMPT_MasterControlData'));
        if (masterControlData) {
            this._pageSize = masterControlData.AllowPaging === "1" ? masterControlData.PageSize : masterControlData.ExtendedPageSize;
        }
        this._originalPageSize = this._pageSize;
        this._stillDataLoading = true;
        this.onDataBind.emit();
        //prompt data is provided by the developer himself.
        if (this.dataSourceObservable) {
            var /** @type {?} */ subsription = this.dataSourceObservable.subscribe(function (data) {
                if (data && data.length > 0) {
                    var /** @type {?} */ keys = Object.keys(data[0]);
                    if (!_this.dataFields || _this.dataFields.length == 0)
                        _this.dataFields = [keys[0], keys[1]];
                    if (!_this.gridFields || _this.gridFields.length == 0)
                        _this.gridFields = _this.dataFields;
                }
                _this._dataSet = data;
                _this._originalDataSet = data;
                _this.MethodAfterDataLoaded();
            }, function (error) { console.error('this is from list prompt - error in developer given observable', error); }, function () { _this._stillDataLoading = false; });
        }
        else if (this.masterGroup) {
            var /** @type {?} */ groupStatus = 'All';
            if (this.status) {
                if (this.status.toLowerCase() == 'active')
                    groupStatus = 'Active';
                else if (this.status.toLowerCase() == 'inactive')
                    groupStatus = 'Inactive';
            }
            var /** @type {?} */ apiArgs = { selector: [{ "MasterGroup": this.masterGroup, "HierarchyRequired": "0" }], activeStatus: groupStatus, selectedIndex: 0 };
            this.listPromptService.GetMasterDefinitionValues(apiArgs).subscribe(function (data) {
                _this._dataSet = data;
                _this._originalDataSet = data;
                if (data && data.length > 0) {
                    if (!_this.dataFields || _this.dataFields.length == 0)
                        _this.dataFields = ['MasterGroupValue', 'MasterGroupValueDescription'];
                    if (!_this.gridFields || _this.gridFields.length == 0)
                        _this.gridFields = _this.dataFields;
                    if (!_this.gridHeaders || _this.gridHeaders.length == 0) {
                        _this.gridHeaders = ['Code', 'Description'];
                    }
                }
                _this.MethodAfterDataLoaded();
            }, function (error) { console.error('this is from list prompt- loading inbuilt MasterGroup prompt', error); }, function () { _this._stillDataLoading = false; });
        }
        else if (this.predefinedPromptArgs && this.predefinedPromptArgs.length > 0) {
            var /** @type {?} */ apiArgs = { ListFromID: this.predefinedPromptArgs[0], ExternalParameters: this.predefinedPromptArgs[1] || '' };
            this.listPromptService.GetPredefinedPromptData(apiArgs)
                .subscribe(function (data) {
                if (data && data.length > 0) {
                    var /** @type {?} */ keys = Object.keys(data[0]);
                    if (!_this.dataFields || _this.dataFields.length == 0)
                        _this.dataFields = [keys[0], keys[1]];
                    if (!_this.gridFields || _this.gridFields.length == 0)
                        _this.gridFields = _this.dataFields;
                }
                _this._dataSet = data;
                _this._originalDataSet = data;
                _this.MethodAfterDataLoaded();
            }, function (error) { console.error('this is from list prompt - loading from prompt predefind prompt args', error); }, function () { _this._stillDataLoading = false; });
        }
        else if (this.type) {
            var /** @type {?} */ typeSwitch = this.type.toLowerCase();
            switch (typeSwitch) {
                case 'product': {
                    var /** @type {?} */ selectQueryColumnsArray = union(this.dataFields, this.gridFields).filter(function (col) { return (col != 'ProductCode' && col != 'Description' && col != 'Status'); });
                    var /** @type {?} */ selectQueryColumns = this.getCommaSepratedList(selectQueryColumnsArray, 'p.');
                    var /** @type {?} */ apiArgs_1 = { strict: true, clsProduct: this.ngModel[this.clsProduct], status: this.status, filter: this.filter, selectQueryColumns: selectQueryColumns };
                    this.listPromptService.GetListPromptProductData(apiArgs_1).subscribe(function (data) {
                        _this._dataSet = data;
                        _this._originalDataSet = data;
                        _this.MethodAfterDataLoaded();
                        if (data && data.length > 0) {
                            var /** @type {?} */ keys = Object.keys(data[0]);
                            if (!_this.dataFields || _this.dataFields.length == 0)
                                _this.dataFields = ['ProductCode', 'Description'];
                            if (!_this.gridFields || _this.gridFields.length == 0)
                                _this.gridFields = _this.dataFields;
                            if (!_this.gridHeaders || _this.gridHeaders.length == 0) {
                                _this.gridHeaders = ['Product Code', 'Description'];
                            }
                        }
                        _this.MethodAfterDataLoaded();
                    }, function (error) { console.error('this is from list prompt - loading inbuilt prompt product', error, apiArgs_1); }, function () { _this._stillDataLoading = false; });
                    break;
                }
                case 'territory': {
                    //V3009Remove start
                    //let territoryStatus = 'All';
                    //if (this.status) {
                    //    if (this.status.toLowerCase() == 'active')
                    //        territoryStatus = 'Active';
                    //    else if (this.status.toLowerCase() == 'inactive')
                    //        territoryStatus = 'Inactive';
                    //}
                    //let apiArgs = { selector: [{ "MasterGroup": "TETY", "HierarchyRequired": "0" }], activeStatus: territoryStatus, selectedIndex: 0 };
                    //V3009Remove end
                    //V3009Adding start
                    var /** @type {?} */ selectQueryColumnsArray = union(this.dataFields, this.gridFields).filter(function (col) { return (col != 'MasterGroupValue' && col != 'MasterGroupValueDescription'); });
                    var /** @type {?} */ selectQueryColumns = selectQueryColumnsArray.join();
                    var /** @type {?} */ apiArgs = { strict: false, status: this.status, filter: this.filter, selectQueryColumns: selectQueryColumns };
                    //V3009Adding end
                    this.listPromptService.GetListPromptTerritoryData(apiArgs).subscribe(function (data) {
                        _this._dataSet = data;
                        _this._originalDataSet = data;
                        if (data && data.length > 0) {
                            if (!_this.dataFields || _this.dataFields.length == 0)
                                _this.dataFields = ['MasterGroupValue', 'MasterGroupValueDescription'];
                            if (!_this.gridFields || _this.gridFields.length == 0)
                                _this.gridFields = _this.dataFields;
                            if (!_this.gridHeaders || _this.gridHeaders.length == 0) {
                                _this.gridHeaders = ['Territory Code', 'Description'];
                            }
                        }
                        _this.MethodAfterDataLoaded();
                    }, function (error) { console.error('this is from list prompt- loading inbuilt prompt territory', error); }, function () { _this._stillDataLoading = false; });
                    break;
                }
                case 'executive': {
                    var /** @type {?} */ selectQueryColumnsArray = union(this.dataFields, this.gridFields).filter(function (col) { return (col != 'ExecutiveCode' && col != 'ExecutiveName' && col != 'Status'); });
                    var /** @type {?} */ selectQueryColumns = this.getCommaSepratedList(selectQueryColumnsArray, 'e.');
                    var /** @type {?} */ apiArgs_2 = { strict: true, clsGeo: this.ngModel[this.clsGeo], clsExecutive: this.ngModel[this.clsExecutive], territory: this.ngModel[this.territory], status: this.status, filter: this.filter, selectQueryColumns: selectQueryColumns };
                    this.listPromptService.GetListPromptExecutiveData(apiArgs_2).subscribe(function (data) {
                        _this._dataSet = data;
                        _this._originalDataSet = data;
                        if (data && data.length > 0) {
                            var /** @type {?} */ keys = Object.keys(data[0]);
                            if (!_this.dataFields || _this.dataFields.length == 0)
                                _this.dataFields = ['ExecutiveCode', 'ExecutiveName'];
                            if (!_this.gridFields || _this.gridFields.length == 0)
                                _this.gridFields = _this.dataFields;
                            if (!_this.gridHeaders || _this.gridHeaders.length == 0) {
                                _this.gridHeaders = ['Executive Code', 'Executive Name'];
                            }
                        }
                        _this.MethodAfterDataLoaded();
                    }, function (error) { console.error('this is from list prompt - loading inbuilt prompt executive ', error, apiArgs_2); }, function () { _this._stillDataLoading = false; });
                    break;
                }
                case 'retailer': {
                    var /** @type {?} */ selectQueryColumnsArray = union(this.dataFields, this.gridFields).filter(function (col) { return (col != 'RetailerCode' && col != 'RetailerName' && col != 'Status'); });
                    var /** @type {?} */ selectQueryColumns = this.getCommaSepratedList(selectQueryColumnsArray, 'r.');
                    var /** @type {?} */ apiArgs_3 = { strict: true, clsGeo: this.ngModel[this.clsGeo], clsRetailer: this.ngModel[this.clsRetailer], territory: this.ngModel[this.territory], status: this.status, filter: this.filter, selectQueryColumns: selectQueryColumns };
                    this.listPromptService.GetListPromptRetailerData(apiArgs_3).subscribe(function (data) {
                        _this._dataSet = data;
                        _this._originalDataSet = data;
                        if (data && data.length > 0) {
                            var /** @type {?} */ keys = Object.keys(data[0]);
                            if (!_this.dataFields || _this.dataFields.length == 0)
                                _this.dataFields = ['RetailerCode', 'RetailerName'];
                            if (!_this.gridFields || _this.gridFields.length == 0)
                                _this.gridFields = _this.dataFields;
                            if (!_this.gridHeaders || _this.gridHeaders.length == 0) {
                                _this.gridHeaders = ['Retailer Code', 'Retailer Name'];
                            }
                        }
                        _this.MethodAfterDataLoaded();
                    }, function (error) { console.error('this is from list prompt - loading inbuilt prompt retailer', error, apiArgs_3); }, function () { _this._stillDataLoading = false; });
                    break;
                }
                default:
                    break;
            }
        }
    };
    /**
     * @param {?} rowsToSelect
     * @param {?} mode
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.SetPromptSelections = /**
     * @param {?} rowsToSelect
     * @param {?} mode
     * @return {?}
     */
    function (rowsToSelect, mode) {
        var _this = this;
        if (mode && mode.trim() == 'range') {
            this.isRangeApplied = true;
        }
        if (rowsToSelect && rowsToSelect.length > 0) {
            var /** @type {?} */ selectedRows_1 = [];
            rowsToSelect.forEach(function (selected, index) {
                var /** @type {?} */ selectedItem = _this._dataSet.find(function (item) { return item[_this.uniqueColumn].trim() == selected[_this.uniqueColumn].trim(); });
                if (selectedItem) {
                    if (_this.isRangeApplied) {
                        selectedItem.InRange = true;
                        if (index == 0 || index == rowsToSelect.length - 1) {
                            selectedItem.selected = true;
                        }
                    }
                    else {
                        selectedItem.selected = true;
                    }
                    selectedRows_1.push(selectedItem);
                }
            });
            this.selectedRows = selectedRows_1;
        }
    };
    /**
     * @return {?}
     */
    ListPromptMultipleComponent.prototype.MethodAfterDataLoaded = /**
     * @return {?}
     */
    function () {
        if (!this.uniqueColumn)
            this.uniqueColumn = this.dataFields[0];
        this._currentSortBy = this.uniqueColumn;
        this.SetPromptSelections(this.selectedRows, ''); //populating already selected rows start
        this.ClearModel();
    };
    ListPromptMultipleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'xont-ventura-listPrompt-multiple',
                    styles: [
                        '.Textboxstyle {padding-left:2px;}',
                        '#element {position:relative;height:40px;}',
                        '.chkAllTh , .chkTd {position:relative;width:32px;cursor:pointer;}',
                        '.chkAllTh > input, .chkTd > input {position:absolute;left:10px;top:1px;cursor:pointer;}',
                        '#divApplyRange {position:absolute;left:-70px;cursor:pointer;bottom:29px;}',
                        '#divApplyRange i {font-size:20px;margin-left:4px;}',
                        '#divApplyRange > span {display:inline-block;position:absolute;top:-6px;border:2px solid;padding:5px 10px 2px 7px;border-radius:16px;}',
                        '#divOptions {position:absolute;left:350px;bottom:2px;}',
                        '#divButtonSet {position:absolute;left:320px;bottom:2px;width:168px;}',
                        '#divRowCount {position:absolute;bottom:2px;background-color:white;cursor:pointer;padding:5px 11px 4px 12px;border-radius:22px;border:solid 2px;}',
                        '.RowSelected td {background-color: rgb(154, 222, 224) !important;}',
                        '.dataSlider mark {padding:0 !important;background-color:yellow !important;}',
                        '.MainButtonStyle {width:80px;height:20px;}',
                        '.AstrixWhenInvoked {position:absolute;top:-6px;}',
                        '.AstrixWhenInvoked > i {position:absolute;top:6px;right:-20px;}',
                        '.AstrixWhenInvoked > small {position:absolute;right:-26px;}',
                        ".dataSlider {\n            position:absolute;min-height:350px;top:-6px;border-radius:0 7px 7px 7px;\n            margin-bottom:15px;border:1.5px #5399d6 solid;padding-left:5px;\n            color:#01539D;padding-top:29px;margin-top:0px;margin-left:-7px;\n            font-family:Trebuchet MS;background-color:ghostwhite;\n         }",
                        '.dtRow { cursor:pointer; }',
                        '.dtRow > td {padding-right:10px;}',
                        '.dtRow:hover {color:red;}',
                        '.pmtPaginator { position:absolute; bottom:4px; border:2px solid #006699; border-radius:6px;}',
                        '.sliderLoader {height: 80px; margin: auto; display: block; margin-top: 80px;}',
                        '.clickButton {padding-left:5px;font-size:16px;padding-right:5px;cursor:pointer;display:inline-block;vertical-align:top;margin-top:5px;}',
                        '.scrollDiv {overflow:auto;overflow-x: hidden;margin:8px 5px 5px 0;}',
                        '.closeI {font-size:18px;cursor:pointer;color:#006699;vertical-align:top;margin-top:4px;}',
                        '.disabledInput {background-color: rgba(170, 170, 170, 0.19);}',
                        '.Textboxstyle[disabled] {background-color: rgba(170, 170, 170, 0.19);}',
                        '.clickButtonDisabled {padding-left:5px;font-size:16px;padding-right:5px;display:inline-block;vertical-align:top;margin-top:5px;}',
                        '.tableStyles tbody td { padding: 2.5px 0 2.5px 3.5px;border: 1px solid #AAAAAA;background:white;font-size:small;font-family:Trebuchet MS;font-size:13px;}',
                        '.headerTable {margin-left:10px;margin-top:4px; background-color: #006699;color: #ffffff;}',
                        '.tableStyles th {padding: 2.5px 0 2.5px 3.5px;font-family:Trebuchet MS; font-size:13px;margin-left:10px;margin-top:4px; background-color: #006699;color: #ffffff;}',
                        '.tableStyles th:hover {text-decoration:underline;cursor:pointer;}',
                        '.sortIcon {padding-left:7px;}',
                        '.pmtPaginator > span {padding-left:5px;padding-right:5px;}',
                        '.zIndexElement { z-index:1;}',
                        '.zIndexTextBoxes { z-index:1;}',
                        '#divUnderlay:hover {background:rgba(150, 150, 150, 0.4);}',
                        '.lpmtMandatoryValidatorStyle {color:#e50505;font-family:Trebuchet MS;font-size:20px;font-style:normal;}',
                        '.fa-indent {color: transparent; text-shadow: 0 0 0 #006699;float: left; margin-top: -2px; font-size: 16px; display: inline-block; border-radius: 16px; box-shadow: 0px 0px 2px #069;;padding: 5px; padding-top: 6px;margin-right: 5px;}',
                        '.fa-filter {color: transparent; text-shadow: 0 0 0 #006699;float: left; margin-top: -2px; font-size: 17px; display: inline-block; border-radius: 16px; box-shadow: 0px 0px 2px #069;;padding: 5px; padding-top: 6px;margin-right: 5px;padding-right: 6.35px; padding-left:6.35px;}',
                        '.sortIcon {padding-left:7px;}',
                        '.mlptApplyRangeNotPermittedDivClass { position: absolute; bottom: 12%; right: 2%; padding: 13px; background-color: white; border: #7DB9E8 2px solid; border-radius: 20px;}'
                    ],
                    template: "<div #element id='controllerElement' style='position:relative;' [class.zIndexElement]=\"_dataSliderStatus == 'block'\">\n               <div id='divUnderlay' style=\"position:fixed;height:100%;width:100%;top:0;left:0;\" *ngIf=\"_dataSliderStatus == 'block'\"></div>\n               <div style='position:relative;height:40px;' [class.zIndexTextBoxes]=\"_dataSliderStatus == 'block'\">\n                <ng-template ngFor let-input [ngForOf]=\"inputWidths\" let-inputIndex=\"index\"\n                        let-isFirst=\"first\" let-isLast=\"last\">\n                    <textarea #inputElement\n                        [(ngModel)]=\"ngModel[modelProps[inputIndex]]\"\n                        [disabled]=\"disabled\" \n                        [maxlength]=\"maxLengths[inputIndex]\"\n                        [class.disabledInput]=\"disabled\"\n                        [style.height]=\"_textBoxesHeight\"\n                        [style.width]=\"inputWidths[inputIndex] + 'px'\"\n                        (mousedown)=\"InputMousedIn($event, inputIndex)\"\n                        (keyup)=\"InputKeyUp($event, inputIndex)\" \n                        (keydown)=\"InputKeyDown($event, isLast, inputIndex)\"\n                        class=\"Textboxstyle\" style='resize:none;'></textarea>\n                    <span *ngIf=\"isFirst\">\n                        <span style=\"cursor:pointer;\" contenteditable=\"false\"><i [class]=\"_searchTypeClass\" contenteditable=\"false\" id=\"iSearchType\" *ngIf=\"_dataSliderStatus == 'block'\" title=\"filter/focus\" (click)=\"iSearchType_Click()\"></i></span>\n                        <div (click)=\"ClickedOn($event)\" *ngIf=\"!disabled\" class=\"fa fa-angle-double-down clickButton\" aria-hidden=\"true\"></div>\n                        <div class=\"fa fa-angle-double-down clickButtonDisabled\" *ngIf=\"disabled\" aria-hidden=\"true\"></div>\n                    </span>\n                    <span *ngIf=\"isLast\">\n                        <i class=\"fa fa-times-circle closeI\" *ngIf=\"_dataSliderStatus == 'block'\" title=\"clear code/description\" (click)=\"ClearSearchBoxes_Click()\"></i>\n                        <small *ngIf=\"mandatory && !valid\" class=\"lpmtMandatoryValidatorStyle\">*</small>\n                    </span>\n                </ng-template>\n                </div>\n\n                    <div class=\"dataSlider\" [style.display]=\"_dataSliderStatus\" [style.minWidth]=\"_minWidth < 500 ? '500px' :(_minWidth + 105 + 'px')\"> \n                        <div *ngIf=\"_dataSet == null\">\n                            <img class=\"sliderLoader\" src='../App_Themes_V3/Blue/images/load_pmt.gif'/>\n                        </div>\n\n                        <div #scrollDiv style='height:270px;' class=\"scrollDiv\" *ngIf= \"!_stillDataLoading\">\n                            <table #gridTable (mfSortByChange)=\"sortByChanged($event)\" (mfSortOrderChange) = \"sortOrderChanged($event)\" [mfRowsOnPage]=\"_pageSize\" [style.minWidth]=\"_minWidth < 480 ? '480px' : (_minWidth + 79 + 'px')\" [mfData]=\"_dataSet\" #mf=\"mfDataTable\" \n                        class=\"tableStyles\">\n                            \n                            <thead *ngIf=\"gridHeaders?.length > 0 && gridFields?.length > 0\">\n                                <tr>\n                                    <th class='chkAllTh'>\n                                        <input type='checkbox' [disabled]=\"isRangeApplied || isSelectedOnlyMode\" [checked]=\"isAllChecked\" id='chkAll' (click)=\"CheckAll()\"/>\n                                    </th>\n                                    <th *ngFor=\"let header of gridHeaders let gridHeaderIndex = index\">\n                                        <mfDefaultSorter [by]=\"gridFields[gridHeaderIndex]\">{{header}}</mfDefaultSorter>\n                                    </th>\n                                </tr>\n                            </thead>\n\n                            <tbody>\n                                <tr *ngFor=\"let item of mf.data\" class=\"dtRow\" (click)=\"CheckChanged($event,item)\" [class.RowSelected]=\"item.InRange || item.selected || isAllChecked\" >\n                                    <td class='chkTd'>\n                                        <input type='checkbox' [disabled]=\"isRangeApplied\" [checked]=\"item.selected\" (click)=\"CheckChanged($event,item)\"/>\n                                    </td>\n                                    <td  *ngFor=\"let field of gridFields let fieldIndex = index\" [innerHtml]=\"giveMeTheCellValue(item, field)\"></td>\n                                </tr>\n                            </tbody>\n                             <tfoot>\n                              <div *ngIf=\"_dataSet?.length > _pageSize || _dataSet?.length == 0\">\n                                <div class=\"pmtPaginator\" (click)=\"pagerClickCapture($event)\" *ngIf=\"_dataSet?.length > _pageSize\">\n                                    <mfBootstrapPaginator></mfBootstrapPaginator>\n                                </div>\n                                <div class=\"pmtPaginator\" *ngIf= \"_dataSet?.length == 0\">\n                                    <span *ngIf= \"_originalDataSet?.length == 0\">No Data Found </span>\n                                    <span *ngIf= \"_originalDataSet?.length != 0 && selectedRows?.length == 0\">No Selected Rows Found </span>\n                                </div>\n                             </div>\n                             <div #divRangeArea>\n                                <div id=\"divOptions\">\n                                    <div id=\"divApplyRange\" title='Apply Range' *ngIf=\"(selectedRows?.length == 2 || isRangeApplied) && !isSelectedOnlyMode\" (click)=\"ApplyRange()\">\n                                        <span [style.backgroundColor]=\"isRangeApplied ? '#9adee0' : 'white'\"> <i class=\"fa fa-arrows-v\" aria-hidden=\"true\"></i></span>\n                                    </div>\n                                    <div id='divRowCount' [style.left]=\"-110-5*(selectedRows.length.toString().length) +'px'\" [style.backgroundColor]=\"isSelectedOnlyMode ? '#9adee0' : 'white'\" title='Currently Selected' (click)=\"rowCountClick()\" (mouseenter)=\"rowCountMouseEnter()\" (mouseleave)=\"rowCountMouseLeave()\">\n                                        <span id='spanRowCount'>{{selectedRows?.length}}</span>\n                                    </div>\n                                </div>\n                              </div>\n                            </tfoot>\n                            </table>\n                        <div>\n                    </div>\n                    </div>\n                    <div id='divButtonSet'>\n                        <button (click)=\"btnOK_click()\" [disabled]=\"selectedRows.length == 0 || _dataSliderStatus == 'none'\" class=\"MainButtonStyle\">OK</button>\n                        <button (click)=\"btnCancel_click()\" class=\"MainButtonStyle\">Cancel</button>                    \n                    </div>\n                </div>\n\n",
                    providers: [XONT_VENTURA_LISTPROMPT_MULTIPLE_VALUE_ACCESSOR]
                },] },
    ];
    //Private Methods End
    /** @nocollapse */
    ListPromptMultipleComponent.ctorParameters = function () { return [
        { type: XontVenturaListPromptService, },
    ]; };
    ListPromptMultipleComponent.propDecorators = {
        "selectedModeProp": [{ type: Input },],
        "selectedRowsProp": [{ type: Input },],
        "uniqueColumn": [{ type: Input },],
        "maxLengths": [{ type: Input },],
        "inputWidths": [{ type: Input },],
        "dataFields": [{ type: Input },],
        "gridFields": [{ type: Input },],
        "gridHeaders": [{ type: Input },],
        "modelProps": [{ type: Input },],
        "mandatory": [{ type: Input },],
        "disabled": [{ type: Input },],
        "ngModel": [{ type: Input },],
        "type": [{ type: Input },],
        "status": [{ type: Input },],
        "predefinedPromptArgs": [{ type: Input },],
        "masterGroup": [{ type: Input },],
        "clsGeo": [{ type: Input },],
        "clsRetailer": [{ type: Input },],
        "clsProduct": [{ type: Input },],
        "clsExecutive": [{ type: Input },],
        "territory": [{ type: Input },],
        "ngModelChange": [{ type: Output },],
        "onChanged": [{ type: Output },],
        "onDataBind": [{ type: Output },],
        "element": [{ type: ViewChild, args: ['element',] },],
        "scrollDiv": [{ type: ViewChild, args: ['scrollDiv',] },],
        "divRangeArea": [{ type: ViewChild, args: ['divRangeArea',] },],
        "inputElements": [{ type: ViewChildren, args: ['inputElement',] },],
        "gridTable": [{ type: ViewChild, args: [DataTable,] },],
    };
    return ListPromptMultipleComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var XontVenturaListPromptModule = (function () {
    function XontVenturaListPromptModule() {
    }
    /**
     * @return {?}
     */
    XontVenturaListPromptModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: XontVenturaListPromptModule,
        };
    };
    XontVenturaListPromptModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule, DataTableModule, FormsModule
                    ],
                    declarations: [
                        ListPromptComponent, ListPromptMultipleComponent
                    ],
                    exports: [
                        ListPromptComponent, ListPromptMultipleComponent
                    ],
                    providers: [
                        XontVenturaListPromptService,
                        CommonService
                    ]
                },] },
    ];
    return XontVenturaListPromptModule;
}());

export { XontVenturaListPromptModule, XONT_VENTURA_LIST_PROMPT_VALUE_ACCESSOR, ListPromptComponent };
