// xont-ventura-batchselector/index.js
import { Injectable, Component, Input, Output, ViewChild, EventEmitter, Pipe, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Headers, Http, HttpModule } from '@angular/http';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { BusyModule } from 'angular2-busy';
import { DataTableModule } from 'angular2-datatable';
import { FormsModule } from '@angular/forms';
import { XontVenturaMessagePromptModule } from 'xont-ventura-message-prompt';
import { XontVenturaDatepickerModule } from 'xont-ventura-datepicker';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/catch';
import { Observable } from 'rxjs/Rx';
import { Subject } from 'rxjs/Subject';
import { CommonService, DatetimeService } from 'xont-ventura-services';
import { XontVenturaValidatorsModule } from 'xont-ventura-validators';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var ProductBatchSerial = (function () {
    function ProductBatchSerial() {
        this.BatchDate = null;
        this.BatchSerial = "";
        this.BatchSerialPrice = 0;
        this.BatchTransactionCost = 0;
        this.ExpiryDate = null;
        this.LocationCode = "";
        this.LotNumber = "";
        this.MovementType = "";
        this.ProductCode = "";
        this.Quantity = 0;
        this.TerritoryCode = "";
        this.WarehouseCode = "";
    }
    return ProductBatchSerial;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var BatchSelectorService = (function () {
    function BatchSelectorService(http, commonService, datetimeService) {
        // if (localStorage.getItem("ControlData") !=null) {
        //     let controlData:any=JSON.parse(localStorage.getItem("ControlData"));
        //     this.decimalPlaces=controlData.DecimalPlaces;
        // }
        this.http = http;
        this.commonService = commonService;
        this.datetimeService = datetimeService;
        this.componentMethodCallSource = new Subject();
        // Observable string streams
        this.componentMethodCalled$ = this.componentMethodCallSource.asObservable();
        this.taskCode = '';
    }
    /**
     * @return {?}
     */
    BatchSelectorService.prototype.getAPIPrefix = /**
     * @return {?}
     */
    function () {
        return this.commonService.getAPIPrefix();
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} checkExpiryDate
     * @return {?}
     */
    BatchSelectorService.prototype.AutoSelect1 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} checkExpiryDate
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, checkExpiryDate) {
        var _this = this;
        this.taskCode = taskCode;
        var /** @type {?} */ data = {
            Type: '1',
            CheckExpiryDate: checkExpiryDate,
            TerritoryCode: territoryCode,
            WarehouseCode: warehouseCode,
            LocationCode: locationCode,
            ProductCode: productCode,
            Qty: requiredQty,
            UOM: uom
        };
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.getAPIPrefix() + "/api/STTRN02/GetBatchSerials", data, { headers: headers })
            .map(function (response) { return _this.AutoSelectBatches(response.json(), requiredQty); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} checkExpiryDate
     * @return {?}
     */
    BatchSelectorService.prototype.AutoSelect2 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} checkExpiryDate
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, checkExpiryDate) {
        var _this = this;
        this.taskCode = taskCode;
        var /** @type {?} */ data = {
            Type: '2',
            CheckExpiryDate: checkExpiryDate,
            TerritoryCode: territoryCode,
            WarehouseCode: warehouseCode,
            LocationCode: locationCode,
            ProductCode: productCode,
            Qty: requiredQty,
            UOM: uom,
            UOM1: uom1,
            ConversionFactor: conversionFactor,
            ReservedBatchSerials: reservedBatchSerials,
        };
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.getAPIPrefix() + "/api/STTRN02/GetBatchSerials", data, { headers: headers })
            .map(function (response) { return _this.AutoSelectBatches(response.json(), requiredQty); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} autoSelect
     * @param {?} checkExpiryDate
     * @return {?}
     */
    BatchSelectorService.prototype.AutoSelect3 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} autoSelect
     * @param {?} checkExpiryDate
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, autoSelect, checkExpiryDate) {
        var _this = this;
        this.taskCode = taskCode;
        var /** @type {?} */ data = {
            Type: '3',
            CheckExpiryDate: checkExpiryDate,
            TerritoryCode: territoryCode,
            WarehouseCode: warehouseCode,
            LocationCode: locationCode,
            ProductCode: productCode,
            Qty: requiredQty,
            UOM: uom,
            UOM1: uom1,
            ConversionFactor: conversionFactor,
            AutoSelect: autoSelect,
            ReservedBatchSerials: reservedBatchSerials
        };
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.getAPIPrefix() + "/api/STTRN02/GetBatchSerials", data, { headers: headers })
            .map(function (response) { return _this.AutoSelectBatches(response.json(), requiredQty); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @return {?}
     */
    BatchSelectorService.prototype.AutoSelect4 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, selectedBatchSerials) {
        var _this = this;
        this.taskCode = taskCode;
        var /** @type {?} */ data = {
            Type: '4',
            TerritoryCode: territoryCode,
            WarehouseCode: warehouseCode,
            LocationCode: locationCode,
            ProductCode: productCode,
            Qty: requiredQty,
            UOM: uom,
            UOM1: uom1,
            ConversionFactor: conversionFactor,
            ReservedBatchSerials: reservedBatchSerials,
            SelectedBatchSerials: selectedBatchSerials,
            AutoSelect: true
        };
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.getAPIPrefix() + "/api/STTRN02/GetBatchSerials", data, { headers: headers })
            .map(function (response) { return _this.AutoSelectBatches(response.json(), requiredQty); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @param {?} addAllocated
     * @return {?}
     */
    BatchSelectorService.prototype.AutoSelect5 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @param {?} addAllocated
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, selectedBatchSerials, addAllocated) {
        var _this = this;
        this.taskCode = taskCode;
        var /** @type {?} */ data = {
            Type: '5',
            TerritoryCode: territoryCode,
            WarehouseCode: warehouseCode,
            LocationCode: locationCode,
            ProductCode: productCode,
            Qty: requiredQty,
            UOM: uom,
            UOM1: uom1,
            ConversionFactor: conversionFactor,
            AddAllocated: addAllocated,
            ReservedBatchSerials: reservedBatchSerials,
            SelectedBatchSerials: selectedBatchSerials,
            AutoSelect: true
        };
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.getAPIPrefix() + "/api/STTRN02/GetBatchSerials", data, { headers: headers })
            .map(function (response) { return _this.AutoSelectBatches(response.json(), requiredQty); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} expireDate
     * @return {?}
     */
    BatchSelectorService.prototype.AutoSelectExpireBatch = /**
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} expireDate
     * @return {?}
     */
    function (territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, expireDate) {
        var _this = this;
        var /** @type {?} */ data = {
            Type: '6',
            TerritoryCode: territoryCode,
            WarehouseCode: warehouseCode,
            LocationCode: locationCode,
            ProductCode: productCode,
            Qty: requiredQty,
            UOM: uom,
            ExpireDate: expireDate
        };
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.getAPIPrefix() + "/api/STTRN02/GetBatchSerials", data, { headers: headers })
            .map(function (response) { return _this.AutoSelectBatches(response.json(), requiredQty); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} checkExpiryDate
     * @return {?}
     */
    BatchSelectorService.prototype.GetBatchSerials1 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} checkExpiryDate
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, checkExpiryDate) {
        this.taskCode = taskCode;
        var /** @type {?} */ data = {
            Type: '1',
            CheckExpiryDate: checkExpiryDate,
            TerritoryCode: territoryCode,
            WarehouseCode: warehouseCode,
            LocationCode: locationCode,
            ProductCode: productCode,
            Qty: requiredQty,
            UOM: uom
        };
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.getAPIPrefix() + "/api/STTRN02/GetBatchSerials", data, { headers: headers })
            .map(function (response) { return response.json(); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} checkExpiryDate
     * @return {?}
     */
    BatchSelectorService.prototype.GetBatchSerials2 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} checkExpiryDate
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, checkExpiryDate) {
        this.taskCode = taskCode;
        var /** @type {?} */ data = {
            Type: '2',
            CheckExpiryDate: checkExpiryDate,
            TerritoryCode: territoryCode,
            WarehouseCode: warehouseCode,
            LocationCode: locationCode,
            ProductCode: productCode,
            Qty: requiredQty,
            UOM: uom,
            UOM1: uom1,
            ConversionFactor: conversionFactor,
            ReservedBatchSerials: reservedBatchSerials,
        };
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.getAPIPrefix() + "/api/STTRN02/GetBatchSerials", data, { headers: headers })
            .map(function (response) { return response.json(); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} autoSelect
     * @param {?} checkExpiryDate
     * @return {?}
     */
    BatchSelectorService.prototype.GetBatchSerials3 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} autoSelect
     * @param {?} checkExpiryDate
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, autoSelect, checkExpiryDate) {
        this.taskCode = taskCode;
        var /** @type {?} */ data = {
            Type: '3',
            CheckExpiryDate: checkExpiryDate,
            TerritoryCode: territoryCode,
            WarehouseCode: warehouseCode,
            LocationCode: locationCode,
            ProductCode: productCode,
            Qty: requiredQty,
            UOM: uom,
            UOM1: uom1,
            ConversionFactor: conversionFactor,
            AutoSelect: autoSelect,
            ReservedBatchSerials: reservedBatchSerials
        };
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.getAPIPrefix() + "/api/STTRN02/GetBatchSerials", data, { headers: headers })
            .map(function (response) { return response.json(); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @param {?} checkExpiryDate
     * @return {?}
     */
    BatchSelectorService.prototype.GetBatchSerials4 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @param {?} checkExpiryDate
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, selectedBatchSerials, checkExpiryDate) {
        this.taskCode = taskCode;
        var /** @type {?} */ data = {
            Type: '4',
            CheckExpiryDate: checkExpiryDate,
            TerritoryCode: territoryCode,
            WarehouseCode: warehouseCode,
            LocationCode: locationCode,
            ProductCode: productCode,
            Qty: requiredQty,
            UOM: uom,
            UOM1: uom1,
            ConversionFactor: conversionFactor,
            ReservedBatchSerials: reservedBatchSerials,
            SelectedBatchSerials: selectedBatchSerials
        };
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.getAPIPrefix() + "/api/STTRN02/GetBatchSerials", data, { headers: headers })
            .map(function (response) { return response.json(); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @param {?} addAllocated
     * @return {?}
     */
    BatchSelectorService.prototype.GetBatchSerials5 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @param {?} addAllocated
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, selectedBatchSerials, addAllocated) {
        this.taskCode = taskCode;
        var /** @type {?} */ data = {
            Type: '5',
            TerritoryCode: territoryCode,
            WarehouseCode: warehouseCode,
            LocationCode: locationCode,
            ProductCode: productCode,
            Qty: requiredQty,
            UOM: uom,
            UOM1: uom1,
            ConversionFactor: conversionFactor,
            AddAllocated: addAllocated,
            ReservedBatchSerials: reservedBatchSerials,
            SelectedBatchSerials: selectedBatchSerials
        };
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.getAPIPrefix() + "/api/STTRN02/GetBatchSerials", data, { headers: headers })
            .map(function (response) { return response.json(); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} expireDate
     * @return {?}
     */
    BatchSelectorService.prototype.GetBatchSerials6 = /**
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} expireDate
     * @return {?}
     */
    function (territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, expireDate) {
        var /** @type {?} */ data = {
            Type: '6',
            TerritoryCode: territoryCode,
            WarehouseCode: warehouseCode,
            LocationCode: locationCode,
            ProductCode: productCode,
            Qty: requiredQty,
            UOM: uom,
            ExpireDate: expireDate
        };
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.getAPIPrefix() + "/api/STTRN02/GetBatchSerials", data, { headers: headers })
            .map(function (response) { return response.json(); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} expireDate
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @return {?}
     */
    BatchSelectorService.prototype.GetBatchSerials7 = /**
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} expireDate
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @return {?}
     */
    function (territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, expireDate, uom1, conversionFactor, reservedBatchSerials, selectedBatchSerials) {
        var /** @type {?} */ data = {
            Type: '7',
            TerritoryCode: territoryCode,
            WarehouseCode: warehouseCode,
            LocationCode: locationCode,
            ProductCode: productCode,
            Qty: requiredQty,
            UOM: uom,
            ExpireDate: expireDate,
            UOM1: uom1,
            ConversionFactor: conversionFactor,
            ReservedBatchSerials: reservedBatchSerials,
            SelectedBatchSerials: selectedBatchSerials
        };
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.getAPIPrefix() + "/api/STTRN02/GetBatchSerials", data, { headers: headers })
            .map(function (response) { return response.json(); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @param {?} checkExpiryDate
     * @param {?} addAllocated
     * @return {?}
     */
    BatchSelectorService.prototype.GetBatchSerials8 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @param {?} checkExpiryDate
     * @param {?} addAllocated
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, selectedBatchSerials, checkExpiryDate, addAllocated) {
        this.taskCode = taskCode;
        var /** @type {?} */ data = {
            Type: '8',
            CheckExpiryDate: checkExpiryDate,
            TerritoryCode: territoryCode,
            WarehouseCode: warehouseCode,
            LocationCode: locationCode,
            ProductCode: productCode,
            Qty: requiredQty,
            UOM: uom,
            UOM1: uom1,
            ConversionFactor: conversionFactor,
            AddAllocated: addAllocated,
            ReservedBatchSerials: reservedBatchSerials,
            SelectedBatchSerials: selectedBatchSerials
        };
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.getAPIPrefix() + "/api/STTRN02/GetBatchSerials", data, { headers: headers })
            .map(function (response) { return response.json(); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @param {?} checkExpiryDate
     * @param {?} addAllocated
     * @return {?}
     */
    BatchSelectorService.prototype.AutoSelect8 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @param {?} checkExpiryDate
     * @param {?} addAllocated
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, selectedBatchSerials, checkExpiryDate, addAllocated) {
        var _this = this;
        this.taskCode = taskCode;
        var /** @type {?} */ data = {
            Type: '8',
            CheckExpiryDate: checkExpiryDate,
            TerritoryCode: territoryCode,
            WarehouseCode: warehouseCode,
            LocationCode: locationCode,
            ProductCode: productCode,
            Qty: requiredQty,
            UOM: uom,
            UOM1: uom1,
            ConversionFactor: conversionFactor,
            AddAllocated: addAllocated,
            ReservedBatchSerials: reservedBatchSerials,
            SelectedBatchSerials: selectedBatchSerials,
            AutoSelect: true
        };
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.getAPIPrefix() + "/api/STTRN02/GetBatchSerials", data, { headers: headers })
            .map(function (response) { return _this.AutoSelectBatches(response.json(), requiredQty); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} currentBatchSerials
     * @param {?} qty
     * @return {?}
     */
    BatchSelectorService.prototype.AutoSelectBatches = /**
     * @param {?} currentBatchSerials
     * @param {?} qty
     * @return {?}
     */
    function (currentBatchSerials, qty) {
        var /** @type {?} */ result = [];
        var /** @type {?} */ totalQty = 0;
        debugger;
        for (var /** @type {?} */ i = 0; i < currentBatchSerials.length; i++) {
            if (parseFloat(currentBatchSerials[i].SelectedQty) > 0) {
                if (parseFloat(currentBatchSerials[i].SelectedQty) > currentBatchSerials[i].BalanceQuantity) {
                    return [];
                }
                else {
                    var /** @type {?} */ batchDate = null;
                    var /** @type {?} */ expiryDate = new Date("0001-01-01"); //min date for C#
                    if (currentBatchSerials[i].BatchDate != null) {
                        batchDate = new Date(currentBatchSerials[i].BatchDate);
                    }
                    if (currentBatchSerials[i].ExpiryDate != null) {
                        expiryDate = new Date(currentBatchSerials[i].ExpiryDate);
                    }
                    var /** @type {?} */ obj = new ProductBatchSerial();
                    obj.BatchDate = batchDate,
                        obj.BatchSerial = currentBatchSerials[i].BatchSerialNumber,
                        obj.BatchSerialPrice = parseFloat(currentBatchSerials[i].BatchPrice),
                        obj.BatchTransactionCost = currentBatchSerials[i].BatchCost,
                        obj.ExpiryDate = expiryDate,
                        obj.LotNumber = currentBatchSerials[i].LotNumber,
                        obj.ProductCode = currentBatchSerials[i].ProductCode,
                        obj.Quantity = parseFloat(currentBatchSerials[i].SelectedQty);
                    totalQty += parseFloat(currentBatchSerials[i].SelectedQty);
                    result.push(obj);
                }
            }
        }
        if (totalQty != qty) {
            return [];
        }
        return result;
    };
    BatchSelectorService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    BatchSelectorService.ctorParameters = function () { return [
        { type: Http, },
        { type: CommonService, },
        { type: DatetimeService, },
    ]; };
    return BatchSelectorService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var XontVenturaBatchSelectorComponent = (function () {
    function XontVenturaBatchSelectorComponent(batchSelectorService, datetimeService) {
        this.batchSelectorService = batchSelectorService;
        this.datetimeService = datetimeService;
        //@Input() taskCode: string;//mandatory
        //@Input() checkExpiryDate?: boolean = false;//optional
        this.allowChange = true;
        this.pageSize = 25;
        this.productCode = '';
        this.productDesc = '';
        this.onSelect = new EventEmitter();
        this.onCancel = new EventEmitter();
        this.batchSearch = "";
        this.currentBatchSerials = [];
        this.batchSelectorService.componentMethodCalled$
            .subscribe(function (error) {
            //this.batchSelectorMsgPrompt.show(error, this.taskCode);
            console.log(error);
        });
    }
    /**
     * @return {?}
     */
    XontVenturaBatchSelectorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        // if (localStorage.getItem("ControlData") !=null) {
        //     let controlData:any=JSON.parse(localStorage.getItem("ControlData"));
        //     this.decimalPlaces=controlData.DecimalPlaces;
        //     $("#hdnDecimalPlaces").val(controlData.DecimalPlaces);
        //     this.numberFormat = "1." + this.decimalPlaces.toString() + "-" + this.decimalPlaces.toString();
        // } else {
        //     $("#hdnDecimalPlaces").val(0);
        // }
        if (localStorage.getItem("PROMPT_MasterControlData") != null) {
            var /** @type {?} */ masterData = JSON.parse(localStorage.getItem("PROMPT_MasterControlData"));
            if (masterData.AllowPaging == "1") {
                this.pageSize = masterData.PageSize;
            }
            else {
                this.pageSize = masterData.ExtendedPageSize;
            }
        }
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} checkExpiryDate
     * @return {?}
     */
    XontVenturaBatchSelectorComponent.prototype.ShowBatchSerials1 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} checkExpiryDate
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, checkExpiryDate) {
        var _this = this;
        this.productCode = productCode;
        this.moveQty = requiredQty;
        this.selectedQty = requiredQty;
        this.batchSearch = "";
        this.busy = this.batchSelectorService.GetBatchSerials1(taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, checkExpiryDate)
            .subscribe(function (data) {
            _this.currentBatchSerials = data;
            _this.setSelectedQty();
            $('#' + _this.id).modal({ backdrop: "static" });
        }, function (err) {
            _this.batchSelectorMsgPrompt.show(err, taskCode);
        });
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} checkExpiryDate
     * @return {?}
     */
    XontVenturaBatchSelectorComponent.prototype.ShowBatchSerials2 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} checkExpiryDate
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, checkExpiryDate) {
        var _this = this;
        this.productCode = productCode;
        this.moveQty = requiredQty;
        this.selectedQty = requiredQty;
        this.batchSearch = "";
        this.busy = this.batchSelectorService.GetBatchSerials2(taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, checkExpiryDate)
            .subscribe(function (data) {
            _this.currentBatchSerials = data;
            _this.setSelectedQty();
            $('#' + _this.id).modal({ backdrop: "static" });
        }, function (err) {
            _this.batchSelectorMsgPrompt.show(err, taskCode);
        });
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} autoSelect
     * @param {?} checkExpiryDate
     * @return {?}
     */
    XontVenturaBatchSelectorComponent.prototype.ShowBatchSerials3 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} autoSelect
     * @param {?} checkExpiryDate
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, autoSelect, checkExpiryDate) {
        var _this = this;
        this.productCode = productCode;
        this.moveQty = requiredQty;
        this.selectedQty = requiredQty;
        this.batchSearch = "";
        var /** @type {?} */ reservedBatchSerials = [];
        this.busy = this.batchSelectorService.GetBatchSerials3(taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, autoSelect, checkExpiryDate)
            .subscribe(function (data) {
            _this.currentBatchSerials = data;
            _this.setSelectedQty();
            $('#' + _this.id).modal({ backdrop: "static" });
        }, function (err) {
            _this.batchSelectorMsgPrompt.show(err, taskCode);
        });
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @param {?} checkExpiryDate
     * @return {?}
     */
    XontVenturaBatchSelectorComponent.prototype.ShowBatchSerials4 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @param {?} checkExpiryDate
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, selectedBatchSerials, checkExpiryDate) {
        var _this = this;
        this.productCode = productCode;
        this.moveQty = requiredQty;
        this.selectedQty = requiredQty;
        this.batchSearch = "";
        this.busy = this.batchSelectorService.GetBatchSerials4(taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, selectedBatchSerials, checkExpiryDate)
            .subscribe(function (data) {
            _this.currentBatchSerials = data;
            _this.setSelectedQty();
            $('#' + _this.id).modal({ backdrop: "static" });
        }, function (err) {
            _this.batchSelectorMsgPrompt.show(err, taskCode);
        });
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @param {?} addAllocated
     * @return {?}
     */
    XontVenturaBatchSelectorComponent.prototype.ShowBatchSerials5 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @param {?} addAllocated
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, selectedBatchSerials, addAllocated) {
        var _this = this;
        this.productCode = productCode;
        this.moveQty = requiredQty;
        this.selectedQty = requiredQty;
        this.batchSearch = "";
        this.busy = this.batchSelectorService.GetBatchSerials5(taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, selectedBatchSerials, addAllocated)
            .subscribe(function (data) {
            _this.currentBatchSerials = data;
            _this.setSelectedQty();
            $('#' + _this.id).modal({ backdrop: "static" });
        }, function (err) {
            _this.batchSelectorMsgPrompt.show(err, taskCode);
        });
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @param {?} checkExpiryDate
     * @param {?} addAllocated
     * @return {?}
     */
    XontVenturaBatchSelectorComponent.prototype.ShowBatchSerials8 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} uom1
     * @param {?} conversionFactor
     * @param {?} reservedBatchSerials
     * @param {?} selectedBatchSerials
     * @param {?} checkExpiryDate
     * @param {?} addAllocated
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, selectedBatchSerials, checkExpiryDate, addAllocated) {
        var _this = this;
        this.productCode = productCode;
        this.moveQty = requiredQty;
        this.selectedQty = requiredQty;
        this.batchSearch = "";
        this.busy = this.batchSelectorService.GetBatchSerials8(taskCode, territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, uom1, conversionFactor, reservedBatchSerials, selectedBatchSerials, checkExpiryDate, addAllocated)
            .subscribe(function (data) {
            _this.currentBatchSerials = data;
            _this.setSelectedQty();
            $('#' + _this.id).modal({ backdrop: "static" });
        }, function (err) {
            _this.batchSelectorMsgPrompt.show(err, taskCode);
        });
    };
    /**
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} expireDate
     * @param {?=} uom1
     * @param {?=} conversionFactor
     * @param {?=} reservedBatchSerials
     * @param {?=} selectedBatchSerials
     * @return {?}
     */
    XontVenturaBatchSelectorComponent.prototype.ShowExpireBatch = /**
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} requiredQty
     * @param {?} uom
     * @param {?} expireDate
     * @param {?=} uom1
     * @param {?=} conversionFactor
     * @param {?=} reservedBatchSerials
     * @param {?=} selectedBatchSerials
     * @return {?}
     */
    function (territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, expireDate, uom1, conversionFactor, reservedBatchSerials, selectedBatchSerials) {
        var _this = this;
        this.productCode = productCode;
        this.moveQty = requiredQty;
        this.selectedQty = requiredQty;
        this.batchSearch = "";
        debugger;
        if (uom1 == undefined && conversionFactor == undefined && reservedBatchSerials == undefined && selectedBatchSerials == undefined) {
            this.busy = this.batchSelectorService.GetBatchSerials6(territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, expireDate)
                .subscribe(function (data) {
                _this.currentBatchSerials = data;
                _this.setSelectedQty();
                $('#' + _this.id).modal({ backdrop: "static" });
            }, function (err) {
                _this.batchSelectorMsgPrompt.show(err, "");
            });
        }
        else {
            this.busy = this.batchSelectorService.GetBatchSerials7(territoryCode, warehouseCode, locationCode, productCode, requiredQty, uom, expireDate, uom1, conversionFactor, reservedBatchSerials, selectedBatchSerials)
                .subscribe(function (data) {
                _this.currentBatchSerials = data;
                _this.setSelectedQty();
                $('#' + _this.id).modal({ backdrop: "static" });
            }, function (err) {
                _this.batchSelectorMsgPrompt.show(err, "");
            });
        }
    };
    /**
     * @param {?} item
     * @return {?}
     */
    XontVenturaBatchSelectorComponent.prototype.selectedQtyOnChange = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        //var selQty =0;
        // if (item.SelectedQty.toString().trim() != '') {
        //     //selQty = parseInt(item.SelectedQty.toString().trim());//V3001
        //     selQty = parseFloat(item.SelectedQty.toString().trim());//V3001
        // }
        if (item.SelectedQty > item.BalanceQuantity) {
            item.SelectedQty = item.BalanceQuantity;
        }
        if (item.SelectedQty == 0) {
            item.YesNo = false;
        }
        else {
            item.YesNo = true;
        }
        //item.SelectedQty = selQty.toFixed(this.decimalPlaces);
        this.setSelectedQty();
    };
    /**
     * @return {?}
     */
    XontVenturaBatchSelectorComponent.prototype.setSelectedQty = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ total = 0;
        for (var /** @type {?} */ i = 0; i < this.currentBatchSerials.length; i++) {
            //total += parseInt( this.currentBatchSerials[i].SelectedQty);//V3001
            total += parseFloat(this.currentBatchSerials[i].SelectedQty); //V3001
        }
        this.selectedQty = total; //parseFloat(total.toFixed(this.decimalPlaces));
    };
    /**
     * @param {?} item
     * @return {?}
     */
    XontVenturaBatchSelectorComponent.prototype.chkOnChange = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        item.YesNo = !item.YesNo;
        if (item.YesNo == true) {
            var /** @type {?} */ balance = this.moveQty - this.selectedQty;
            if (balance > 0) {
                if (balance > item.BalanceQuantity) {
                    item.SelectedQty = item.BalanceQuantity; //.toString();
                }
                else {
                    item.SelectedQty = balance; //.toString();
                }
            }
            else {
                item.YesNo = !item.YesNo; //change check status
            }
        }
        else {
            //item.SelectedQty = "0";
            item.SelectedQty = 0;
        }
        //item.SelectedQty=parseFloat(item.SelectedQty.toString()).toFixed(this.decimalPlaces);
        this.setSelectedQty();
    };
    /**
     * @param {?} currentBatchSerials
     * @param {?} qty
     * @return {?}
     */
    XontVenturaBatchSelectorComponent.prototype.AutoSelectBatchesForManual = /**
     * @param {?} currentBatchSerials
     * @param {?} qty
     * @return {?}
     */
    function (currentBatchSerials, qty) {
        var /** @type {?} */ result = [];
        var /** @type {?} */ totalQty = 0;
        debugger;
        for (var /** @type {?} */ i = 0; i < currentBatchSerials.length; i++) {
            if (parseFloat(currentBatchSerials[i].SelectedQty) > 0) {
                if (parseFloat(currentBatchSerials[i].SelectedQty) > currentBatchSerials[i].BalanceQuantity) {
                    return [];
                }
                else {
                    //var batchDate = null;
                    var /** @type {?} */ expiryDate = new Date("0001-01-01"); //min date for C#
                    // if (currentBatchSerials[i].BatchDate != "") {
                    //     batchDate = this.datetimeService.getDateTimeForString(currentBatchSerials[i].BatchDate);
                    // }
                    // if (currentBatchSerials[i].ExpiryDate != "") {
                    //     expiryDate = this.datetimeService.getDateTimeForString(currentBatchSerials[i].ExpiryDate);
                    // }
                    if (currentBatchSerials[i].ExpiryDate != null) {
                        expiryDate = currentBatchSerials[i].ExpiryDate;
                    }
                    var /** @type {?} */ obj = new ProductBatchSerial();
                    obj.BatchDate = currentBatchSerials[i].BatchDate,
                        obj.BatchSerial = currentBatchSerials[i].BatchSerialNumber,
                        obj.BatchSerialPrice = parseFloat(currentBatchSerials[i].BatchPrice),
                        obj.BatchTransactionCost = currentBatchSerials[i].BatchCost,
                        obj.ExpiryDate = expiryDate,
                        obj.LotNumber = currentBatchSerials[i].LotNumber,
                        obj.ProductCode = currentBatchSerials[i].ProductCode,
                        obj.Quantity = parseFloat(currentBatchSerials[i].SelectedQty);
                    totalQty += parseFloat(currentBatchSerials[i].SelectedQty);
                    result.push(obj);
                }
            }
        }
        if (totalQty != qty) {
            return [];
        }
        return result;
    };
    /**
     * @return {?}
     */
    XontVenturaBatchSelectorComponent.prototype.cancel = /**
     * @return {?}
     */
    function () {
        $('#' + this.id).modal('hide');
        this.onCancel.emit([]);
    };
    /**
     * @return {?}
     */
    XontVenturaBatchSelectorComponent.prototype.ok = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ selectedSerialsList = this.AutoSelectBatchesForManual(this.currentBatchSerials, this.moveQty);
        $('#' + this.id).modal('hide');
        this.onSelect.emit(selectedSerialsList);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    XontVenturaBatchSelectorComponent.prototype.onSearch = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event.keyCode == 13) {
            if (this.batchSearch.trim() != "") {
                for (var /** @type {?} */ i = 0; i < this.currentBatchSerials.length; i++) {
                    if (this.currentBatchSerials[i].BatchSerialNumber.toString().toUpperCase().trim() == this.batchSearch.toUpperCase().trim()) {
                        if (this.currentBatchSerials[i].YesNo == false) {
                            var /** @type {?} */ balance = this.moveQty - this.selectedQty;
                            if (balance > 0) {
                                if (balance > this.currentBatchSerials[i].BalanceQuantity) {
                                    this.currentBatchSerials[i].SelectedQty = this.currentBatchSerials[i].BalanceQuantity;
                                }
                                else {
                                    this.currentBatchSerials[i].SelectedQty = balance;
                                }
                                //this.currentBatchSerials[i].SelectedQty=this.currentBatchSerials[i].SelectedQty.toFixed(this.decimalPlaces);
                                this.currentBatchSerials[i].YesNo = true;
                                this.setSelectedQty();
                            }
                        }
                    }
                }
            }
        }
    };
    XontVenturaBatchSelectorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'xont-ventura-batchselector',
                    styles: ['.selectedRow {background-color: rgb(102, 153, 153);}'],
                    providers: [BatchSelectorService],
                    template: "<div [ngBusy]=\"busy\"></div> <xont-ventura-message-prompt [id]=\"'batchSelectorErrorPrompt'\" #batchSelectorMsgPrompt></xont-ventura-message-prompt> <div id=\"{{id}}\" class=\"modal fade\" role=\"dialog\"> <div class=\"modal-dialog\" style=\"width:800px\"> <!-- Modal content--> <div class=\"modal-content\" > <div class=\"modal-header\" style=\"padding-top: 5px;padding-right: 15px;padding-bottom:5px;padding-left: 15px;\"> <h5 class=\"modal-title\">Product Batch / Serials</h5> </div> <div class=\"modal-body\" style=\"padding-bottom: 0px;padding-top: 2px;\"> <table> <tr> <td width=\"90px\"> <label class=\"Captionstyle\">Product Code</label> </td> <td style=\"min-width:150px\"> <input type=\"text\" class=\"Textboxstyle\" name=\"txtProductCode\" [(ngModel)]=\"productCode\" style=\"width:140px\" disabled/> <label class=\"Captionstyle\"> {{productDesc}}</label> </td> <td width=\"100px\" style=\"padding-left:10px\"> <label class=\"Captionstyle\">Required Qty.</label> </td> <td> <input type=\"text\" class=\"Textboxstyle text-right\" name=\"txtMoveQty\" [ngModel]=\"moveQty | xont_quantity\"  style=\"width:140px\" disabled/> </td> </tr> <tr> <td> <label class=\"Captionstyle\">Batch/Serial</label> </td> <td> <input type=\"text\" class=\"Textboxstyle\" name=\"txtBatchSearch\" style=\"width:140px\" (keydown)=\"onSearch($event)\" [(ngModel)]=\"batchSearch\" placeholder=\"Search\"  /> </td> <td style=\"padding-left:10px\"> <label class=\"Captionstyle\">Selected Qty.</label> </td> <td> <input type=\"text\" class=\"Textboxstyle text-right\" name=\"txtSelectedQty\" [ngModel]=\"selectedQty | xont_quantity\" style=\"width:140px\" disabled /> </td> </tr> </table> <input type=\"hidden\" id=\"hdnDecimalPlaces\" /> <span class=\"Labelstyle\" [hidden]=\"currentBatchSerials?.length != 0\">No data found for the given criteria</span> <div class=\"DataTableContainer\"> <table id=\"tblWarehouse\" [hidden]=\"currentBatchSerials?.length == 0\" [mfData]=\"currentBatchSerials | batchFilter : batchSearch\" #batchSelectorDatatable=\"mfDataTable\" class=\"DataTableStyle HeaderFreezeStyle\" [mfRowsOnPage]=\"pageSize\" > <thead> <tr> <th width=\"7%\"> Select </th> <th width=\"12%\"> <mfDefaultSorter by=\"LotNumber\">Lot Number</mfDefaultSorter> </th> <th width=\"12%\"> <mfDefaultSorter by=\"BatchDate\">Batch Date</mfDefaultSorter> </th> <th width=\"14%\"> <mfDefaultSorter by=\"BatchSerialNumber\">Batch/Serial</mfDefaultSorter> </th> <th width=\"15%\"> <mfDefaultSorter by=\"BalanceQuantity\">Available Qty.</mfDefaultSorter> </th> <th width=\"16%\"> <mfDefaultSorter by=\"SelectedQty\">Select</mfDefaultSorter> </th> <th width=\"12%\"> <mfDefaultSorter by=\"ExpiryDate\">Expiry Date</mfDefaultSorter> </th> <th width=\"12%\"> <mfDefaultSorter by=\"BatchPrice\">Batch Price</mfDefaultSorter> </th> </tr> </thead> <tbody> <tr *ngFor=\"let item of batchSelectorDatatable.data\" [ngClass]=\"{'selectedRow': item.YesNo}\"> <td width=\"7%\" style=\"padding-top: 1px;padding-bottom: 1px;\" align=\"center\" > <span> <input type=\"checkbox\" [checked]=\"item.YesNo\" (change)=\"chkOnChange(item)\" [disabled]=\"!allowChange\" /> </span> </td> <td width=\"12%\" style=\"padding-top: 1px;padding-bottom: 1px;\"> {{item.LotNumber}} </td> <td width=\"12%\" style=\"padding-top: 1px;padding-bottom: 1px;\"> {{item.BatchDate | xont_date}} </td> <td width=\"14%\" style=\"padding-top: 1px;padding-bottom: 1px;\"> <span>{{item.BatchSerialNumber}}</span> </td> <td width=\"15%\" class=\"text-right\" style=\"padding-top: 1px;padding-bottom: 1px;\"> {{item.BalanceQuantity | xont_quantity }} </td> <td width=\"16%\" style=\"padding-top: 1px;padding-bottom: 1px;\"> <span> <input type=\"text\" name=\"SelectedQty\" [disabled]=\"!allowChange\"  class=\"Textboxstyle text-right\" [(ngModel)]=\"item.SelectedQty\" style=\"width:105px\" (blur)=\"selectedQtyOnChange(item)\" xont_quantity  /> </span> </td> <td width=\"12%\" style=\"padding-top: 1px;padding-bottom: 1px;\"> {{item.ExpiryDate | xont_date}} </td> <td width=\"12%\" class=\"text-right\" style=\"padding-top: 1px;padding-bottom: 1px;\"> {{item.BatchPrice | xont_value}} </td> </tr> </tbody> <tfoot> <tr> <td colspan=\"10\"> <mfBootstrapPaginator></mfBootstrapPaginator> </td> </tr> </tfoot> </table> </div> </div> <div class=\"modal-footer\" style=\"padding-top: 5px;padding-right: 15px;padding-bottom:0px;padding-left: 15px;\"> <button type=\"button\" class=\"MainButtonStyle\" [disabled]=\"moveQty!=selectedQty\" (click)=\"ok()\">OK</button> <button type=\"button\" class=\"MainButtonStyle\" (click)=\"cancel()\">Cancel</button> </div> </div> </div> </div> "
                },] },
    ];
    //V3001
    /** @nocollapse */
    XontVenturaBatchSelectorComponent.ctorParameters = function () { return [
        { type: BatchSelectorService, },
        { type: DatetimeService, },
    ]; };
    XontVenturaBatchSelectorComponent.propDecorators = {
        "id": [{ type: Input },],
        "allowChange": [{ type: Input },],
        "pageSize": [{ type: Input },],
        "productCode": [{ type: Input },],
        "productDesc": [{ type: Input },],
        "onSelect": [{ type: Output },],
        "onCancel": [{ type: Output },],
        "batchSelectorMsgPrompt": [{ type: ViewChild, args: ['batchSelectorMsgPrompt',] },],
    };
    return XontVenturaBatchSelectorComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var BatchEntryService = (function () {
    function BatchEntryService(http, commonService, datetimeService) {
        this.http = http;
        this.commonService = commonService;
        this.datetimeService = datetimeService;
        this.taskCode = '';
    }
    /**
     * @return {?}
     */
    BatchEntryService.prototype.getAPIPrefix = /**
     * @return {?}
     */
    function () {
        return this.commonService.getAPIPrefix(this.taskCode);
    };
    /**
     * @param {?} taskCode
     * @param {?} productCode
     * @return {?}
     */
    BatchEntryService.prototype.GetProduct = /**
     * @param {?} taskCode
     * @param {?} productCode
     * @return {?}
     */
    function (taskCode, productCode) {
        this.taskCode = taskCode;
        return this.http.get(this.getAPIPrefix() + "/api/STTRN02/GetProduct?productCode=" + productCode)
            .map(function (response) { return response.json(); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} productCode
     * @param {?} batchNo
     * @param {?} unitCost
     * @return {?}
     */
    BatchEntryService.prototype.HandleExistingBatchNo = function (taskCode, territoryCode, warehouseCode, locationCode, productCode, batchNo, unitCost) {
        this.taskCode = taskCode;
        var url = this.getAPIPrefix() + "/api/STTRN02/HandleExistingBatchNo?territoryCode=" + territoryCode +
            "&warehouseCode=" + warehouseCode + "&locationCode=" + locationCode +
            "&productCode=" + productCode + "&batchNo=" + batchNo + "&unitCost=" + unitCost;
        return this.http.get(url)
            .map(function (response) { return response.json(); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    BatchEntryService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    BatchEntryService.ctorParameters = function () { return [
        { type: Http, },
        { type: CommonService, },
        { type: DatetimeService, },
    ]; };
    return BatchEntryService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var XontVenturaBatchEntryComponent = (function () {
    function XontVenturaBatchEntryComponent(batchEntryService, datetimeService) {
        this.batchEntryService = batchEntryService;
        this.datetimeService = datetimeService;
        this.allowChange = true;
        this.onSelect = new EventEmitter();
        this.onCancel = new EventEmitter();
        this.selectedBatchSerials = [];
        this.txtProductCode = '';
        this.txtProductDesc = '';
        this.txtMoveQty = 0;
        this.txtUOM = '';
        this.txtSelectedQty = 0;
        this.batchForm = {
            txtFirstBatch: '',
            txtBatchQty: 0,
            txtUnitCost: 0,
            txtLotNumber: '',
            txtExpDate: ''
        };
        this.txtFirstBatchError = '';
        this.txtBatchQtyError = '';
        this.txtUnitCostError = '';
        this.txtLotNumberError = '';
        this.txtExpDateError = '';
        this.territoryCode = '';
        this.warehouseCode = '';
        this.locationCode = '';
        this.transDate = null;
        this.qty = 0;
        this.UOM = '';
        this.autoBatchNumberFlag = '';
        this.autoLotNumberFlag = '';
        this.expiryDateRequired = '';
        this.batchStatusFlag = '';
        this.expDateVisible = true;
        this.txtFirstBatchDisabled = true;
        this.txtLotNumberDisabled = true;
        this.title = '';
        this.isValid = true;
        this.mainErrorMsg = '';
        this.pageSize = 25;
        this.duplicateGeneratedBatchNo = '';
        this.duplicateGeneratedBatchNoError = '';
        this.taskCode = '';
    }
    /**
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        // if (localStorage.getItem("ControlData") !=null) {
        //     let controlData:any=JSON.parse(localStorage.getItem("ControlData"));
        //     this.decimalPlaces=controlData.DecimalPlaces;
        //     $("#hdnDecimalPlaces").val(controlData.DecimalPlaces);
        //     this.numberFormat = "1." + this.decimalPlaces.toString() + "-" + this.decimalPlaces.toString();
        // } else {
        //     $("#hdnDecimalPlaces").val(0);
        // }
        if (localStorage.getItem("PROMPT_MasterControlData") != null) {
            var /** @type {?} */ masterData = JSON.parse(localStorage.getItem("PROMPT_MasterControlData"));
            if (masterData.AllowPaging == "1") {
                this.pageSize = masterData.PageSize;
            }
            else {
                this.pageSize = masterData.ExtendedPageSize;
            }
        }
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} stockProduct
     * @param {?} transDate
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.ShowBatchEntry1 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} stockProduct
     * @param {?} transDate
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, stockProduct, transDate) {
        var _this = this;
        this.busy = this.batchEntryService.GetProduct(taskCode, stockProduct.ProductCode)
            .subscribe(function (product) {
            _this.ShowBatchEntry2(taskCode, territoryCode, warehouseCode, locationCode, stockProduct, product.Description, product.SerialNumberRequired, product.ExpiryDateRequired, product.AutoLotNumberGeneration, product.AutoBatchNumberFlag, transDate);
        }, function (err) {
            _this.batchEntryErrorPrompt.show(err, taskCode);
        });
    };
    /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} stockProduct
     * @param {?} productDesc
     * @param {?} serialNumberRequired
     * @param {?} expiryDateRequired
     * @param {?} autoLotNumberGeneration
     * @param {?} autoBatchNumberFlag
     * @param {?} transDate
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.ShowBatchEntry2 = /**
     * @param {?} taskCode
     * @param {?} territoryCode
     * @param {?} warehouseCode
     * @param {?} locationCode
     * @param {?} stockProduct
     * @param {?} productDesc
     * @param {?} serialNumberRequired
     * @param {?} expiryDateRequired
     * @param {?} autoLotNumberGeneration
     * @param {?} autoBatchNumberFlag
     * @param {?} transDate
     * @return {?}
     */
    function (taskCode, territoryCode, warehouseCode, locationCode, stockProduct, productDesc, serialNumberRequired, expiryDateRequired, autoLotNumberGeneration, autoBatchNumberFlag, transDate) {
        this.taskCode = taskCode;
        //V3003 add start
        debugger;
        this.clearContents();
        if (stockProduct.ProductBatchSerialList != undefined && stockProduct.ProductBatchSerialList != null) {
            for (var /** @type {?} */ i = 0; i < stockProduct.ProductBatchSerialList.length; i++) {
                this.txtSelectedQty += stockProduct.ProductBatchSerialList[i].Quantity;
                var /** @type {?} */ expiryDate = "";
                if (new Date(stockProduct.ProductBatchSerialList[i].ExpiryDate).getFullYear() != 1) {
                    //validate c# min date "0001-01-01"
                    expiryDate = this.datetimeService.getDisplayDate(new Date(stockProduct.ProductBatchSerialList[i].ExpiryDate));
                }
                var /** @type {?} */ obj = {
                    Selected: false,
                    ProductCode: stockProduct.ProductBatchSerialList[i].ProductCode,
                    BatchTransactionCost: stockProduct.ProductBatchSerialList[i].BatchTransactionCost,
                    Quantity: stockProduct.ProductBatchSerialList[i].Quantity,
                    BatchSerialPrice: stockProduct.ProductBatchSerialList[i].BatchSerialPrice,
                    BatchDate: this.datetimeService.getDisplayDate(transDate),
                    ExpiryDate: expiryDate,
                    LotNumber: stockProduct.ProductBatchSerialList[i].LotNumber,
                    BatchSerial: stockProduct.ProductBatchSerialList[i].BatchSerial,
                };
                this.selectedBatchSerials.push(obj);
            }
        }
        //V3003 add end
        //V3003 del start
        // if (stockProduct.ProductBatchSerialList != undefined && stockProduct.ProductBatchSerialList != null) {
        //     this.selectedBatchSerials = stockProduct.ProductBatchSerialList;
        // }
        // this.clearContents();
        //V3003 del end
        this.territoryCode = territoryCode;
        this.warehouseCode = warehouseCode;
        this.locationCode = locationCode;
        this.transDate = transDate;
        this.txtProductCode = stockProduct.ProductCode;
        this.qty = stockProduct.MovementQuantity;
        this.UOM = stockProduct.UOM;
        this.txtProductDesc = productDesc;
        this.autoBatchNumberFlag = autoBatchNumberFlag;
        this.autoLotNumberFlag = autoLotNumberGeneration;
        this.expiryDateRequired = expiryDateRequired;
        this.txtMoveQty = this.qty;
        //this.txtSelectedQty = 0;//V3003 del
        //Set Default Qty
        //this.batchForm.txtBatchQty = (this.txtMoveQty - this.txtSelectedQty).toFixed(this.decimalPlaces);
        this.batchForm.txtBatchQty = (this.txtMoveQty - this.txtSelectedQty);
        //Set the Batch/Serial Status: S=Serial, B=Batch
        if (serialNumberRequired == "1") {
            this.batchStatusFlag = "S";
        }
        else {
            this.batchStatusFlag = "B";
        }
        //Show/Hide expiery date based on Expiery Required Flag in Product
        if (expiryDateRequired == "1") {
            this.expDateVisible = true;
        }
        else {
            this.expDateVisible = false;
        }
        //Enable-Disable Batch Number based on Auto BatchNumber
        //This check is used as this flag may take different values for VRTRN01 to generate Batch Number (Eg. D:Date etc)
        if (this.autoBatchNumberFlag == "1") {
            this.txtFirstBatchDisabled = true;
            this.batchForm.txtFirstBatch = "*";
        }
        else {
            this.txtFirstBatchDisabled = false;
            this.batchForm.txtFirstBatch = "";
        }
        //AutoLot Number setting (Use AutoLotNumberGeneration in Product)
        //This check is used as this flag may take different values to generate Lot Number (Eg. D for Date etc)
        if (this.autoLotNumberFlag == "1") {
            //Get the Auto Lot number definition (when adding) and use the relevent number
            this.batchForm.txtLotNumber = "*";
            this.txtLotNumberDisabled = true;
        }
        else {
            this.txtLotNumberDisabled = false;
            this.batchForm.txtLotNumber = "";
        }
        //Display popup
        if (this.batchStatusFlag == "S") {
            this.title = "Product Serials";
        }
        else {
            this.title = "Product Batch Numbers";
        }
        this.txtUOM = this.UOM;
        this.validateBatch();
        //$('#' + this.id).draggable({ handle: ".modal-header" });
        $('#' + this.id).modal({ backdrop: "static" });
    };
    /**
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.setSelectedQty = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ selectedQty = 0;
        for (var /** @type {?} */ i = 0; i < this.selectedBatchSerials.length; i++) {
            selectedQty += this.selectedBatchSerials[i].Quantity;
        }
        this.txtSelectedQty = selectedQty; //parseFloat(selectedQty.toFixed(this.decimalPlaces)).toString();
    };
    /**
     * @param {?} date
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.onDateSelect = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        this.batchForm.txtExpDate = date;
        this.validateBatch();
    };
    /**
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.validateBatch = /**
     * @return {?}
     */
    function () {
        this.isValid = true;
        //start batch validate & validate serial number
        if (this.autoBatchNumberFlag == "0") {
            if (this.batchForm.txtFirstBatch.trim() == '') {
                this.isValid = false;
                this.txtFirstBatchError = '*';
            }
            else {
                if (this.batchStatusFlag == "S") {
                    var /** @type {?} */ serial = this.batchForm.txtFirstBatch.trim();
                    var /** @type {?} */ startIndex = this.minIndexOfAnyNumber(serial);
                    if (startIndex != -1) {
                        var /** @type {?} */ prefix = serial.substring(0, startIndex);
                        var /** @type {?} */ suffix = serial.substring(startIndex);
                        if (/^\d+$/.test(suffix)) {
                            this.txtFirstBatchError = '';
                        }
                        else {
                            this.isValid = false;
                            this.txtFirstBatchError = 'Invalid Serial number';
                        }
                    }
                    else {
                        this.isValid = false;
                        this.txtFirstBatchError = 'Invalid Serial number';
                    }
                }
                else {
                    this.txtFirstBatchError = '';
                }
            }
        }
        else {
            this.txtFirstBatchError = '';
        }
        //V3003 add start
        //validate Qty
        if (this.batchForm.txtBatchQty == 0) {
            this.isValid = false;
            this.txtBatchQtyError = '*';
        }
        else {
            this.txtBatchQtyError = '';
        }
        //V3003 add end
        //validate Uit cost
        // if (this.batchForm.txtUnitCost != '') {
        //     var pattern = /^(?!(-1(\.0{0,4})?|(\.0{0,4}))$)[0-9]{1,10}?(\.[0-9]{1,4})?$/;
        //     if (!pattern.test(this.batchForm.txtUnitCost)) {
        //         this.isValid = false;
        //         this.txtUnitCostError = 'Enter Valid Cost';
        //     } else {
        //         this.txtUnitCostError = '';
        //     }
        // } else {
        //     this.txtUnitCostError = '';
        // }
        //validate lot number
        if (this.autoLotNumberFlag == "0") {
            if (this.batchForm.txtLotNumber.trim() == '') {
                this.isValid = false;
                this.txtLotNumberError = '*';
            }
            else {
                this.txtLotNumberError = '';
            }
        }
        else {
            this.txtLotNumberError = '';
        }
        //validate expire date
        if (this.expiryDateRequired == '1') {
            if (this.batchForm.txtExpDate.trim() == '') {
                this.isValid = false;
                this.txtExpDateError = '*';
            }
            else {
                if (this.datetimeService.isValidDateString(this.batchForm.txtExpDate.trim())) {
                    this.txtExpDateError = '';
                }
                else {
                    this.isValid = false;
                    this.txtExpDateError = 'Invalid Date';
                }
            }
        }
        else {
            this.txtExpDateError = '';
        }
        //prevent auto serial
        if (this.batchStatusFlag == "S" && this.autoBatchNumberFlag == "1") {
            this.isValid = false;
        }
    };
    /**
     * @param {?} batchSerial
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.isSerialAlreadyExist = /**
     * @param {?} batchSerial
     * @return {?}
     */
    function (batchSerial) {
        var /** @type {?} */ result = false;
        if (this.autoBatchNumberFlag == "0") {
            for (var /** @type {?} */ i = 0; i < this.selectedBatchSerials.length; i++) {
                if (this.selectedBatchSerials[i].BatchSerial == batchSerial) {
                    result = true;
                    break;
                }
            }
        }
        return result;
    };
    /**
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.txtFirstBatch_OnChange = /**
     * @return {?}
     */
    function () {
        this.validateBatch();
    };
    /**
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.txtBatchQty_OnChange = /**
     * @return {?}
     */
    function () {
        this.validateBatch();
    };
    /**
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.txtUnitCost_OnChange = /**
     * @return {?}
     */
    function () {
        this.validateBatch();
    };
    /**
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.txtLotNumber_OnChange = /**
     * @return {?}
     */
    function () {
        this.validateBatch();
    };
    /**
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.txtExpDate_OnChange = /**
     * @return {?}
     */
    function () {
        this.validateBatch();
    };
    /**
     * @param {?} text
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.minIndexOfAnyNumber = /**
     * @param {?} text
     * @return {?}
     */
    function (text) {
        var /** @type {?} */ minStartIndex = -1;
        for (var /** @type {?} */ i = 0; i < 10; i++) {
            var /** @type {?} */ index = text.indexOf(i.toString());
            if (index >= 0) {
                if (minStartIndex == -1) {
                    minStartIndex = index;
                }
                else {
                    if (index < minStartIndex) {
                        minStartIndex = index;
                    }
                }
            }
        }
        return minStartIndex;
    };
    /**
     * @param {?} serial
     * @param {?} originalLength
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.formatNumber = /**
     * @param {?} serial
     * @param {?} originalLength
     * @return {?}
     */
    function (serial, originalLength) {
        var /** @type {?} */ result = "";
        if (serial.toString().length < originalLength) {
            var /** @type {?} */ offsetLength = originalLength - serial.toString().length;
            for (var /** @type {?} */ i = 0; i < offsetLength; i++) {
                result += "0";
            }
        }
        return (result + serial.toString());
    };
    /**
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.add = /**
     * @return {?}
     */
    function () {
        this.mainErrorMsg = '';
        var /** @type {?} */ firstBatchSerialNo = this.batchForm.txtFirstBatch.trim();
        if (!this.isSerialAlreadyExist(firstBatchSerialNo)) {
            // let cost: number = 0;
            // if (this.batchForm.txtUnitCost.trim() != '') {
            //     cost = parseFloat(this.batchForm.txtUnitCost.trim());
            // }
            var /** @type {?} */ cost = this.batchForm.txtUnitCost;
            // let batchQty: number = parseFloat(this.batchForm.txtBatchQty.trim());
            var /** @type {?} */ batchQty = this.batchForm.txtBatchQty;
            if (this.batchStatusFlag == 'S') {
                if (batchQty > 1) {
                    var /** @type {?} */ startIndex = this.minIndexOfAnyNumber(firstBatchSerialNo);
                    var /** @type {?} */ prefix = firstBatchSerialNo.substring(0, startIndex);
                    var /** @type {?} */ suffix = firstBatchSerialNo.substring(startIndex);
                    var /** @type {?} */ suffixNumber = parseInt(suffix);
                    var /** @type {?} */ array = [];
                    for (var /** @type {?} */ i = 0; i < batchQty; i++) {
                        var /** @type {?} */ num = suffixNumber + i;
                        var /** @type {?} */ nextSerial = prefix + this.formatNumber(num, suffix.length);
                        if (!this.isSerialAlreadyExist(nextSerial)) {
                            var /** @type {?} */ lotNumber = "";
                            if (this.autoLotNumberFlag == "0") {
                                lotNumber = this.batchForm.txtLotNumber.trim();
                            }
                            else {
                                lotNumber = "0";
                            }
                            var /** @type {?} */ obj = {
                                Selected: false,
                                ProductCode: this.txtProductCode.trim(),
                                BatchTransactionCost: cost,
                                Quantity: 1,
                                BatchSerialPrice: 0,
                                BatchDate: this.datetimeService.getDisplayDate(this.transDate),
                                ExpiryDate: this.batchForm.txtExpDate.trim(),
                                LotNumber: lotNumber,
                                BatchSerial: nextSerial
                            };
                            array.push(obj);
                        }
                        else {
                            array = [];
                            this.mainErrorMsg = 'Batch/Serial Number: ' + nextSerial + ' already exists';
                            break;
                        }
                    }
                    this.selectedBatchSerials = this.selectedBatchSerials.concat(array);
                    //console.log(this.selectedBatchSerials);
                }
                else {
                    var /** @type {?} */ lotNumber = "";
                    if (this.autoLotNumberFlag == "0") {
                        lotNumber = this.batchForm.txtLotNumber.trim();
                    }
                    else {
                        lotNumber = "0";
                    }
                    var /** @type {?} */ obj = {
                        Selected: false,
                        ProductCode: this.txtProductCode.trim(),
                        BatchTransactionCost: cost,
                        Quantity: 1,
                        BatchSerialPrice: 0,
                        BatchDate: this.datetimeService.getDisplayDate(this.transDate),
                        ExpiryDate: this.batchForm.txtExpDate.trim(),
                        LotNumber: lotNumber,
                        BatchSerial: firstBatchSerialNo
                    };
                    this.selectedBatchSerials.push(obj);
                }
            }
            else {
                var /** @type {?} */ batchSerial = "";
                var /** @type {?} */ lotNumber = "";
                if (this.autoBatchNumberFlag != "0") {
                    batchSerial = "*";
                }
                else {
                    batchSerial = firstBatchSerialNo;
                }
                if (this.autoLotNumberFlag == "0") {
                    lotNumber = this.batchForm.txtLotNumber.trim();
                }
                else {
                    lotNumber = "0";
                }
                var /** @type {?} */ obj = {
                    Selected: false,
                    ProductCode: this.txtProductCode.trim(),
                    BatchTransactionCost: cost,
                    Quantity: batchQty,
                    BatchSerialPrice: 0,
                    BatchDate: this.datetimeService.getDisplayDate(this.transDate),
                    ExpiryDate: this.batchForm.txtExpDate.trim(),
                    LotNumber: lotNumber,
                    BatchSerial: batchSerial
                };
                this.selectedBatchSerials.push(obj);
            }
        }
        else {
            this.mainErrorMsg = 'Batch/Serial Number: ' + this.batchForm.txtFirstBatch.trim() + ' already exists';
        }
        //console.log(this.selectedBatchSerials);
        this.cleanForm();
        this.setSelectedQty();
        this.validateBatch();
    };
    /**
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.ok = /**
     * @return {?}
     */
    function () {
        debugger;
        var /** @type {?} */ array = [];
        for (var /** @type {?} */ i = 0; i < this.selectedBatchSerials.length; i++) {
            var /** @type {?} */ obj = new ProductBatchSerial();
            obj.ProductCode = this.selectedBatchSerials[i].ProductCode;
            obj.BatchTransactionCost = this.selectedBatchSerials[i].BatchTransactionCost;
            obj.Quantity = this.selectedBatchSerials[i].Quantity;
            obj.BatchSerialPrice = this.selectedBatchSerials[i].BatchSerialPrice;
            obj.BatchDate = this.datetimeService.getDateTimeForString(this.selectedBatchSerials[i].BatchDate);
            obj.ExpiryDate = this.selectedBatchSerials[i].ExpiryDate.trim() == "" ? new Date("0001-01-01") : this.datetimeService.getDateTimeForString(this.selectedBatchSerials[i].ExpiryDate);
            obj.LotNumber = this.selectedBatchSerials[i].LotNumber;
            obj.BatchSerial = this.selectedBatchSerials[i].BatchSerial;
            array.push(obj);
        }
        this.onSelect.emit(array);
        $('#' + this.id).modal('hide');
    };
    /**
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.cancel = /**
     * @return {?}
     */
    function () {
        this.onCancel.emit([]);
        $('#' + this.id).modal('hide');
    };
    /**
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.removeAll = /**
     * @return {?}
     */
    function () {
        this.selectedBatchSerials = [];
        this.setSelectedQty();
    };
    /**
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.remove = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ selectedBatchSerialsCopy = [];
        for (var /** @type {?} */ i = 0; i < this.selectedBatchSerials.length; i++) {
            if (!this.selectedBatchSerials[i].Selected) {
                selectedBatchSerialsCopy.push(this.selectedBatchSerials[i]);
            }
        }
        this.selectedBatchSerials = selectedBatchSerialsCopy;
        this.setSelectedQty();
    };
    /**
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.clearContents = /**
     * @return {?}
     */
    function () {
        this.selectedBatchSerials = [];
        this.txtProductCode = '';
        this.txtProductDesc = '';
        this.txtMoveQty = 0;
        this.txtUOM = '';
        this.txtSelectedQty = 0;
        this.batchForm.txtFirstBatch = '';
        // this.batchForm.txtBatchQty= '';
        // this.batchForm.txtUnitCost = '';
        this.batchForm.txtBatchQty = 0;
        this.batchForm.txtUnitCost = 0;
        this.batchForm.txtLotNumber = '';
        this.batchForm.txtExpDate = '';
        this.txtFirstBatchError = '';
        this.txtBatchQtyError = '';
        this.txtUnitCostError = '';
        this.txtLotNumberError = '';
        this.txtExpDateError = '';
        this.mainErrorMsg = '';
    };
    /**
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.cleanForm = /**
     * @return {?}
     */
    function () {
        if (this.autoBatchNumberFlag == "1") {
            this.batchForm.txtFirstBatch = "*";
        }
        else {
            this.batchForm.txtFirstBatch = "";
        }
        if (this.autoLotNumberFlag == "1") {
            this.batchForm.txtLotNumber = "*";
        }
        else {
            this.batchForm.txtLotNumber = "";
        }
        // this.batchForm.txtBatchQty = '';
        // this.batchForm.txtUnitCost = '';
        this.batchForm.txtBatchQty = 0;
        this.batchForm.txtUnitCost = 0;
        this.batchForm.txtExpDate = '';
    };
    /**
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.txtBatchQty_blur = /**
     * @return {?}
     */
    function () {
        // if(this.batchForm.txtBatchQty.trim()==""){
        //     this.batchForm.txtBatchQty="0";
        //     this.batchForm.txtBatchQty=parseFloat(this.batchForm.txtBatchQty).toFixed(this.decimalPlaces);
        // }else{
        //     this.batchForm.txtBatchQty=parseFloat(this.batchForm.txtBatchQty).toFixed(this.decimalPlaces);
        // }
        this.validateBatch();
    };
    /**
     * @return {?}
     */
    XontVenturaBatchEntryComponent.prototype.txtUnitCost_blur = /**
     * @return {?}
     */
    function () {
        // if(this.batchForm.txtUnitCost.trim()==""){
        //     this.batchForm.txtUnitCost="0";
        //     this.batchForm.txtUnitCost=parseFloat(this.batchForm.txtUnitCost).toFixed(2);
        // }else{
        //     this.batchForm.txtUnitCost=parseFloat(this.batchForm.txtUnitCost).toFixed(2);
        // }
        this.validateBatch();
    };
	/**
     * @return {?}
     */
	XontVenturaBatchEntryComponent.prototype.roundToFourDecimalPlacesAndCompare = /**
 	* @param {?} value1
 	* @param {?} value2
 	* @return {?}
 	*/
	function (value1, value2) {
    	// Round both values to 4 decimal places
    	var roundedValue1 = Math.round(value1 * 10000) / 10000;
    	var roundedValue2 = Math.round(value2 * 10000) / 10000;
    
    	return roundedValue1 === roundedValue2;
	};
    /**
     * @return {void}
     */
    XontVenturaBatchEntryComponent.prototype.confirmDuplicateBatchPopup_Yes = /**
    * @return {void}
    */
    function () {
    this.batchForm.txtFirstBatch = this.duplicateGeneratedBatchNo;
    this.add();
    };
    /**
    * @return {void}
    */
    XontVenturaBatchEntryComponent.prototype.confirmDuplicateBatchPopup_No = /**
    * @return {void}
    */
    function () {
    this.duplicateGeneratedBatchNoError = "Batch/Serial Number already exists";
    };
    /**
     * @return {void}
     */
    XontVenturaBatchEntryComponent.prototype.validateAddClick = function () {
        var _this = this; // to handle 'this' inside callbacks
        this.busy = this.batchEntryService.HandleExistingBatchNo(this.taskCode, this.territoryCode,
            this.warehouseCode, this.locationCode, this.txtProductCode,
            this.batchForm.txtFirstBatch.trim(), this.batchForm.txtUnitCost)
            .subscribe(
                function (result) {
                    if (result.BatchNumberChangeForCostFlag.trim() === "1") {
                        if (result.Status === "InvalidBatchNo") {
                            _this.mainErrorMsg = result.Message;
                        }
                        else if (result.Status === "GenerateBatchNo") {
                            // Check for Existing and increment
                            if (result.BatchNumber.includes('_')) {
                                var parts = result.BatchNumber.split('_');
                                var prefix = parts[0];
                                var existingBatch = _this.selectedBatchSerials.find(function (batch) {
                                    return batch.BatchSerial.startsWith(prefix + '_');
                                });
    
                                if (existingBatch) {
                                    var existingParts = existingBatch.BatchSerial.split('_');
                                    var existingSuffix = existingParts[1];
                                    var suffixNumber = parseInt(existingSuffix);
                                    if (!isNaN(suffixNumber)) {
                                        var incrementedSuffix = suffixNumber + 1;
                                        _this.duplicateGeneratedBatchNo = prefix + "_" + incrementedSuffix;
                                    } else {
                                        // Handle case where suffix is not a number
                                        _this.duplicateGeneratedBatchNo = result.BatchNumber + "_1"; // Default behavior to add _1 if suffix isn't numeric
                                    }
                                } else {
                                    // If no existing batch starts with the same prefix, use the current batchNumber
                                    _this.duplicateGeneratedBatchNo = result.BatchNumber;
                                }
                            } else {
                                // If there's no underscore in the original batchNumber, perhaps handle differently or leave as is
                                _this.duplicateGeneratedBatchNo = result.BatchNumber;
                            }
    
                            _this.duplicateBatchPrompt.showConfirm(result.Message, "Yes", "No");
                            _this.mainErrorMsg = _this.duplicateGeneratedBatchNoError;
                        }
                        else {
                            _this.batchForm.txtFirstBatch = result.BatchNumber;
                            _this.add();
                        }
                    } else {
                        _this.add();
                    }
                },
                function (err) {
                    console.log(err);
                });
    };

    XontVenturaBatchEntryComponent.decorators = [
        { type: Component, args: [{
                    selector: 'xont-ventura-batchentry',
                    styles: ['.selectedRow {background-color: rgb(102, 153, 153);}'],
                    template: "<div [ngBusy]=\"busy\"></div> <xont-ventura-message-prompt [id]=\"'batchEntryErrorPrompt'\" #batchSelectorMsgPrompt></xont-ventura-message-prompt> <div id=\"{{id}}\" class=\"modal fade\" role=\"dialog\"> <div class=\"modal-dialog\" > <!-- Modal content--> <div class=\"modal-content\"> <div class=\"modal-header\" style=\"padding-top: 5px;padding-right: 15px;padding-bottom:5px;padding-left: 15px;\"> <h5 class=\"modal-title\">{{title}}</h5> </div> <div class=\"modal-body\"> <table > <tr> <td width=\"118\"> <span id=\"ucProductBatchEntry1_lblProduct\" class=\"Captionstyle\">Product Code</span> </td> <td style=\"padding-right: 10px;\"> <input name=\"txtProductCode\" type=\"text\" [(ngModel)]=\"txtProductCode\"  disabled id=\"txtProductCode\" class=\"Textboxstyle\" style=\"width:120px;\"> </td> <td> <input name=\"txtProductDesc\" type=\"text\" [(ngModel)]=\"txtProductDesc\" disabled id=\"txtProductDesc\" class=\"Textboxstyle\" style=\"width:250px;\"> </td> </tr> <tr> <td> <span id=\"lblRequiredQty\" class=\"Captionstyle\">Required Qty</span> </td> <td> <input name=\"txtMoveQty\" type=\"text\" disabled id=\"txtMoveQty\" [ngModel]=\"txtMoveQty | xont_quantity\"  class=\"Textboxstyle\" style=\"width:120px;text-align:right\" /> </td> <td> <input name=\"txtUOM\" type=\"text\" disabled id=\"txtUOM\" [(ngModel)]=\"txtUOM\" class=\"Textboxstyle\" style=\"width:250px;\"> </td> </tr> <tr> <td> <span id=\"ucProductBatchEntry1_lblSelectedQty\" class=\"Captionstyle\">Selected Qty</span> </td> <td> <input name=\"txtSelectedQty\" type=\"text\" [ngModel]=\"txtSelectedQty | xont_quantity\"  disabled id=\"txtSelectedQty\" class=\"Textboxstyle\" style=\"width:120px;text-align:right\"> </td> <td> &nbsp; </td> </tr> </table> <input type=\"hidden\" id=\"hdnDecimalPlaces\" /> <table style=\"margin-top: 10px;\"> <tr > <td width=\"118\"> <span id=\"lblStartBatchSerial\" class=\"Captionstyle\">Starting Batch S/N</span> </td> <td style=\"padding-right: 10px;\"> <input name=\"txtFirstBatch\" type=\"text\"  maxlength=\"20\" [(ngModel)]=\"batchForm.txtFirstBatch\" id=\"txtFirstBatch\" [disabled]=\"txtFirstBatchDisabled\" (ngModelChange)=\"txtFirstBatch_OnChange()\" class=\"Textboxstyle\" style=\"width:120px;\"> </td> <td> <span  class=\"Errormessagetextstyle\" >{{txtFirstBatchError}}</span> </td> </tr> <tr> <td> <span id=\"lblQty\" class=\"Captionstyle\">Qty</span> </td> <td> <input name=\"txtBatchQty\" type=\"text\"  id=\"txtBatchQty\" [(ngModel)]=\"batchForm.txtBatchQty\" xont_quantity  (blur)=\"txtBatchQty_blur()\"  class=\"Textboxstyle\" style=\"width:120px;text-align:right\"  > </td> <td> <span class=\"Errormessagetextstyle\">{{txtBatchQtyError}}</span> </td> </tr> <tr> <td> <span id=\"lblUnitCost\" class=\"Captionstyle\">Unit Cost</span> </td> <td> <input name=\"txtUnitCost\" type=\"text\" id=\"txtUnitCost\" [(ngModel)]=\"batchForm.txtUnitCost\" xont_value  (blur)=\"txtUnitCost_blur()\" class=\"Textboxstyle\" style=\"width:120px;text-align:right\" > </td> <td> <span  class=\"Errormessagetextstyle \">{{txtUnitCostError}}</span> </td> </tr> <tr> <td> <span id=\"lblLotNumber\" class=\"Captionstyle\">Lot Number</span> </td> <td> <input name=\"txtLotNumber\" type=\"text\"  maxlength=\"20\" [(ngModel)]=\"batchForm.txtLotNumber\" id=\"txtLotNumber\"  class=\"Textboxstyle\" [disabled]=\"txtLotNumberDisabled\" (ngModelChange)=\"txtLotNumber_OnChange()\" style=\"width:120px;\"> </td> <td> <span  class=\"Errormessagetextstyle\">{{txtLotNumberError}}</span> </td> </tr> <tr *ngIf=\"expDateVisible\"> <td> <span id=\"lblExpiryDate\" class=\"Captionstyle\">Expiry Date</span> </td> <td> <input name=\"txtExpDate\" type=\"text\" maxlength=\"10\" [(ngModel)]=\"batchForm.txtExpDate\" id=\"txtExpDate\" (ngModelChange)=\"txtExpDate_OnChange()\" class=\"Textboxstyle\" style=\"width:120px;\"> <xont-ventura-datepicker [id]=\"id+'datepickerFrom'\" (onDateSelect)=\"onDateSelect($event)\"></xont-ventura-datepicker> </td> <td> <span class=\"Errormessagetextstyle\">{{txtExpDateError}}</span> </td> </tr> </table> <span class=\"Labelstyle\" [hidden]=\"selectedBatchSerials?.length != 0\">Enter the Batch/ Serial details</span> <div class=\"PromptDataTableContainer\" style=\"max-height:200px\"> <table [hidden]=\"selectedBatchSerials?.length == 0\" [mfData]=\"selectedBatchSerials\" #batchEntryDatatable=\"mfDataTable\" class=\"PromptDataTableStyle\" [mfRowsOnPage]=\"pageSize\"> <thead> <tr> <th > Select </th> <th > Lot Number </th> <th > Batch Date </th> <th > Batch/Serial </th> <th > Qty. </th> <th > Unit Price </th> <th > Expiry Date </th> </tr> </thead> <tbody> <tr *ngFor=\"let item of batchEntryDatatable.data\" [ngClass]=\"{'selectedRow': item.Selected}\"> <td   class=\"text-center\" style=\"padding-top: 1px;padding-bottom: 1px;\"> <span> <input type=\"checkbox\" [checked]=\"item.Selected\" (change)=\"item.Selected = !item.Selected\" [disabled]=\"!allowChange\" /> </span> </td> <td  style=\"padding-top: 1px;padding-bottom: 1px;\"> {{item.LotNumber}} </td> <td style=\"padding-top: 1px;padding-bottom: 1px;\"> {{item.BatchDate}} </td> <td style=\"padding-top: 1px;padding-bottom: 1px;\"> {{item.BatchSerial}} </td> <td class=\"text-right\" style=\"padding-top: 1px;padding-bottom: 1px;\"> {{item.Quantity    | xont_quantity }} </td> <td class=\"text-right\" style=\"padding-top: 1px;padding-bottom: 1px;\"> {{item.BatchTransactionCost}} </td> <td style=\"padding-top: 1px;padding-bottom: 1px;\"> {{item.ExpiryDate}} </td> </tr> </tbody> <tfoot> <tr> <td colspan=\"7\"> <mfBootstrapPaginator></mfBootstrapPaginator> </td> </tr> </tfoot> </table> </div> </div> <div class=\"modal-footer\" style=\"padding-top: 5px;padding-right: 15px;padding-bottom: 0px;padding-left: 15px;\"> <table> <tr> <td> <button type=\"button\" class=\"MainButtonStyle\" [disabled]=\"!isValid || !allowChange\" (click)=\"validateAddClick()\">Add</button> <button type=\"button\" class=\"MainButtonStyle\" [disabled]=\"!allowChange\" (click)=\"remove()\">Remove</button> <button type=\"button\" class=\"MainButtonStyle\" [disabled]=\"!allowChange\" (click)=\"removeAll()\">Remove All</button> <button type=\"button\" class=\"MainButtonStyle\" [disabled]=\"!roundToFourDecimalPlacesAndCompare(txtMoveQty, txtSelectedQty)\" (click)=\"ok()\">OK</button> <button type=\"button\" class=\"MainButtonStyle\" (click)=\"cancel()\">Cancel</button> </td> </tr> <tr> <td class=\"text-left\"> <span class=\"Errormessagetextstyle\">{{mainErrorMsg}}</span> </td> </tr> </table> </div> </div> </div> </div> <xont-ventura-message-prompt [id]=\"'duplicateBatchPrompt'\" #duplicateBatchPrompt messageType=\"confirm\" (onOK)=\"confirmDuplicateBatchPopup_Yes()\" (onCancel)=\"confirmDuplicateBatchPopup_No()\"></xont-ventura-message-prompt>",
                    providers: [BatchEntryService]
                },] },
    ];
    /** @nocollapse */
    XontVenturaBatchEntryComponent.ctorParameters = function () { return [
        { type: BatchEntryService, },
        { type: DatetimeService, },
    ]; };
    XontVenturaBatchEntryComponent.propDecorators = {
        "id": [{ type: Input },],
        "allowChange": [{ type: Input },],
        "batchEntryErrorPrompt": [{ type: ViewChild, args: ['batchEntryErrorPrompt',] },],
        "duplicateBatchPrompt": [{ type: core.ViewChild, args: ['duplicateBatchPrompt',] }],
        "onSelect": [{ type: Output },],
        "onCancel": [{ type: Output },],
    };
    return XontVenturaBatchEntryComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var BatchFilterPipe = (function () {
    function BatchFilterPipe() {
    }
    /**
     * @param {?} items
     * @param {?} query
     * @return {?}
     */
    BatchFilterPipe.prototype.transform = /**
     * @param {?} items
     * @param {?} query
     * @return {?}
     */
    function (items, query) {
        if (!items) {
            return [];
        }
        if (!query) {
            return items;
        }
        //console.log(items, query);
        var /** @type {?} */ filteredList = [];
        var /** @type {?} */ selectedList = [];
        if (query.trim() != '') {
            for (var /** @type {?} */ i = 0; i < items.length; i++) {
                if (items[i].YesNo == true) {
                    selectedList.push(items[i]);
                }
                else {
                    if ((items[i].BatchSerialNumber.toUpperCase().trim()).startsWith(query.toUpperCase().trim()) == true) {
                        filteredList.push(items[i]);
                    }
                }
            }
            filteredList = filteredList.concat(selectedList);
            return filteredList;
        }
        return items;
    };
    BatchFilterPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'batchFilter'
                },] },
    ];
    return BatchFilterPipe;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var StockProduct = (function () {
    function StockProduct() {
        this.AllocatedQty = 0;
        this.AutoBatchNumberFlag = "";
        this.BatchSerialControlItem = "";
        this.Cost = 0;
        this.FreeIssue = 0;
        this.LineType = "";
        this.Location = "";
        this.MovementQuantity = 0;
        this.MovementReason = "";
        this.Price = 0;
        this.ProductBatchSerialList = [];
        this.ProductCode = "";
        this.Territory = "";
        this.TRNTypeDetailNumber = 0;
        this.UOM = "";
        this.UOMType = "";
        this.Warehouse = "";
    }
    return StockProduct;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var XontVenturaBatchSelectorModule = (function () {
    function XontVenturaBatchSelectorModule() {
    }
    /**
     * @return {?}
     */
    XontVenturaBatchSelectorModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: XontVenturaBatchSelectorModule,
            providers: [BatchSelectorService]
        };
    };
    XontVenturaBatchSelectorModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule, BrowserAnimationsModule, HttpModule, XontVenturaMessagePromptModule, BusyModule, DataTableModule, FormsModule, XontVenturaDatepickerModule, XontVenturaValidatorsModule
                    ],
                    declarations: [
                        XontVenturaBatchSelectorComponent, BatchFilterPipe, XontVenturaBatchEntryComponent
                    ],
                    exports: [
                        XontVenturaBatchSelectorComponent, XontVenturaBatchEntryComponent
                    ],
                    providers: [CommonService, DatetimeService],
                },] },
    ];
    return XontVenturaBatchSelectorModule;
}());

export { XontVenturaBatchSelectorModule, BatchSelectorService, ProductBatchSerial, XontVenturaBatchSelectorComponent, XontVenturaBatchEntryComponent, StockProduct };
