// xont-ventura-multiple-prompt/xont-ventura-multiple-prompt.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('angular2-datatable'), require('@angular/forms'), require('lodash')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/common', 'angular2-datatable', '@angular/forms', 'lodash'], factory) :
	(factory((global['xont-ventura-multiple-prompt'] = {}),global._angular_core,global._angular_common,global.angular2Datatable,global._angular_forms,global._));
}(this, (function (exports,_angular_core,_angular_common,angular2Datatable,_angular_forms,_) { 'use strict';

var DataFilterPipe = (function () {
    function DataFilterPipe() {
    }
    /**
     * @param {?} items
     * @param {?} filter
     * @return {?}
     */
    DataFilterPipe.prototype.transform = function (items, filter$$1) {
        if (!items) {
            return [];
        }
        if (!filter$$1) {
            return items;
        }
        for (var /** @type {?} */ i = 0; i < filter$$1.length; i++) {
            if (filter$$1[i].searchString.trim() != '') {
                items = _.filter(items, function (row) { return row[filter$$1[i].field].toUpperCase().indexOf(filter$$1[i].searchString.trim().toUpperCase()) > -1; });
            }
        }
        return items;
    };
    return DataFilterPipe;
}());
DataFilterPipe.decorators = [
    { type: _angular_core.Pipe, args: [{
                name: 'dataFilter'
            },] },
];
/**
 * @nocollapse
 */
DataFilterPipe.ctorParameters = function () { return []; };

var XontVenturaMultiplePromptComponent = (function () {
    function XontVenturaMultiplePromptComponent() {
        this.rowsOnPage = 25;
        this.sortBy = "";
        this.columnDefs = [];
        this.title = "";
        this.promptFilter = [];
        this.type = "single";
        this.checkedAll = false;
        this.message = "";
        this.onSelect = new _angular_core.EventEmitter();
        this.onCancel = new _angular_core.EventEmitter();
    }
    /**
     * @return {?}
     */
    XontVenturaMultiplePromptComponent.prototype.ngOnInit = function () {
        if (localStorage.getItem("PROMPT_MasterControlData") != null) {
            var /** @type {?} */ masterData = JSON.parse(localStorage.getItem("PROMPT_MasterControlData"));
            if (masterData.AllowPaging == "1") {
                this.rowsOnPage = masterData.PageSize;
            }
            else {
                this.rowsOnPage = masterData.ExtendedPageSize;
            }
        }
    };
    /**
     * @param {?} gridOption
     * @return {?}
     */
    XontVenturaMultiplePromptComponent.prototype.show = function (gridOption) {
        this.type = "single";
        this.checkedAll = false;
        gridOption.dataSet.forEach(function (obj) { obj._checked = false; });
        this.data = gridOption.dataSet;
        this.dataBK = gridOption.dataSet;
        this.columnDefs = gridOption.columnDefs;
        // this.sortBy=this.columnDefs[0].field;
        // this.data.sort((a: any, b: any) => {
        //     if (a[this.sortBy].toUpperCase() < b[this.sortBy].toUpperCase() ){
        //         return -1;
        //     }else if( a[this.sortBy].toUpperCase() > b[this.sortBy].toUpperCase() ){
        //         return 1;
        //     }else{
        //         return 0;
        //     }
        // });
        this.backupData = gridOption.dataSet;
        this.promptFilter = gridOption.columnDefs;
        for (var /** @type {?} */ i = 0; i < this.promptFilter.length; i++) {
            this.promptFilter[i]['searchString'] = '';
        }
        this.title = gridOption.title;
        $('#' + this.id).modal({ backdrop: "static" });
    };
    /**
     * @param {?} type
     * @return {?}
     */
    XontVenturaMultiplePromptComponent.prototype.changeType = function (type) {
        this.message = "";
        this.type = type;
        this.checkedAll = false;
        this.data = this.dataBK;
        this.unselectAll();
    };
    /**
     * @return {?}
     */
    XontVenturaMultiplePromptComponent.prototype.unselectAll = function () {
        for (var /** @type {?} */ i = 0; i < this.data.length; i++) {
            this.data[i]._checked = false;
        }
    };
    /**
     * @return {?}
     */
    XontVenturaMultiplePromptComponent.prototype.selectAll = function () {
        for (var /** @type {?} */ i = 0; i < this.data.length; i++) {
            this.data[i]._checked = true;
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    XontVenturaMultiplePromptComponent.prototype.onKey = function (event) {
        this.data = new DataFilterPipe().transform(this.backupData, this.promptFilter);
        this.message = "";
    };
    /**
     * @return {?}
     */
    XontVenturaMultiplePromptComponent.prototype.cancel = function () {
        $('#' + this.id).modal('hide');
        this.data = [];
        this.onCancel.emit();
    };
    /**
     * @return {?}
     */
    XontVenturaMultiplePromptComponent.prototype.ok = function () {
        var /** @type {?} */ array = [];
        this.data.forEach(function (obj) {
            if (obj._checked) {
                array.push(obj);
            }
        });
        this.message = "";
        if (array.length == 0) {
            $('#' + this.id).modal('hide');
            return;
        }
        if (this.type == "single") {
            if (array.length != 1) {
                this.message = "Please select single record.";
                return;
            }
        }
        else if (this.type == "range") {
            if (array.length != 2) {
                this.message = "Please select two records.";
                return;
            }
        }
        this.onSelect.emit({ Type: this.type, Data: array });
        $('#' + this.id).modal('hide');
    };
    /**
     * @param {?} item
     * @return {?}
     */
    XontVenturaMultiplePromptComponent.prototype.selectRow = function (item) {
        this.message = "";
        if (this.type == "single") {
            if (item._checked == true) {
                item._checked = false;
            }
            else {
                this.unselectAll();
                item._checked = true;
            }
        }
        else if (this.type == "multiple") {
            item._checked = !item._checked;
        }
        else {
            var /** @type {?} */ filterList = this.data.filter(function (a) { return a._checked == true; });
            if (item._checked == false) {
                if (filterList.length > 1) {
                    filterList[1]._checked = false;
                }
            }
            item._checked = !item._checked;
        }
    };
    /**
     * @return {?}
     */
    XontVenturaMultiplePromptComponent.prototype.chkSelectAll_Change = function () {
        if (this.checkedAll == true) {
            this.selectAll();
        }
        else {
            this.unselectAll();
        }
    };
    return XontVenturaMultiplePromptComponent;
}());
XontVenturaMultiplePromptComponent.decorators = [
    { type: _angular_core.Component, args: [{
                selector: 'xont-ventura-multiple-prompt',
                template: "\n  <div id=\"{{id}}\" class=\"modal fade\" role=\"dialog\">\n    <div class=\"modal-dialog \">\n\n        <div class=\"modal-content\">\n            <div class=\"modal-header\" style=\"padding-top: 5px;padding-right: 15px;padding-bottom:5px;padding-left: 15px;\">\n                <h5 class=\"modal-title\">{{title}}</h5>\n            </div>\n            <div class=\"modal-body\" style=\"padding-top: 5px;padding-right: 15px;padding-bottom:15px;padding-left: 15px;\">\n                <table width=\"100%\">\n                    <tr >\n                        \n                        <td align=\"center\">\n                        <input type=\"radio\" id=\"rdoSingle\" name=\"type\" [checked]=\"type=='single'\" (change)=\"changeType('single')\" /><label id=\"lblSingle\" for=\"rdoSingle\" class=\"Captionstyle\">Single</label> &nbsp;\n                        <input type=\"radio\" id=\"rdoMultiple\" name=\"type\" [checked]=\"type=='multiple'\" (change)=\"changeType('multiple')\"  /><label id=\"lblMultiple\" for=\"rdoMultiple\" class=\"Captionstyle\">Multiple</label> &nbsp;\n                        <input type=\"radio\" id=\"rdoRange\" name=\"type\" [checked]=\"type=='range'\" (change)=\"changeType('range')\"  /><label id=\"lblRange\" for=\"rdoRange\" class=\"Captionstyle\">Range</label>\n                        </td>\n                        \n                    </tr>\n                </table>\n\n                <div class=\"PromptDataTableContainer\">\n\n\n                    <table class=\"PromptDataTableStyle\" [mfData]=\"data | dataFilter\" #mf=\"mfDataTable\"\n                           [mfRowsOnPage]=\"rowsOnPage\">\n                        <thead>\n                            <tr>\n                                <th  align=\"center\">\n                                    <input type=\"checkbox\" id=\"chkSelectAll\" [hidden]=\"type!='multiple'\" [checked]=\"checkedAll\" (change)=\"checkedAll=!checkedAll;chkSelectAll_Change()\">\n                                </th>\n\n                                <th *ngFor=\"let column of columnDefs \">\n                                    {{column.headerName}}\n                                </th>\n                            </tr>\n                            <tr>\n                                <td align=\"center\"><i class=\"fa fa-search\" aria-hidden=\"true\"></i></td>\n                                <td *ngFor=\"let column of promptFilter \">\n                                    <input style=\"width:100%\" *ngIf=\"column.filter\" [(ngModel)]=\"column.searchString\" type=\"text\" (keyup)=\"onKey($event)\" class=\"Textboxstyle\">\n                                </td>\n\n                            </tr>\n\n                        </thead>\n                        <tbody>\n                            <tr *ngFor=\"let item of mf.data\" [ngClass]=\"{'selectedRow': item._checked}\">\n                                <td align=\"center\" style=\"padding-top: 1px;padding-bottom: 1px;\">\n                                    <input [type]=\"type=='single'?'radio':'checkbox'\"  [checked]=\"item._checked\" (change)=\"selectRow(item)\">\n\n                                </td>\n\n                                <td *ngFor=\"let column of columnDefs\" style=\"padding-top: 1px;padding-bottom: 1px;\">{{item[column.field]}}</td>\n\n                            </tr>\n                        </tbody>\n                        <tfoot>\n                            <tr>\n                                <td [colSpan]=\"columnDefs.length+1\">\n                                    <mfBootstrapPaginator ></mfBootstrapPaginator>\n                                </td>\n                            </tr>\n                        </tfoot>\n                    </table>\n\n                </div>\n            </div>\n            <div class=\"modal-footer\" style=\"padding-top: 5px;padding-right: 15px;padding-bottom:0px;padding-left: 15px;\">\n                <span style=\"float: left\" class=\"Errormessagetextstyle\">{{message}}</span>\n                <button type=\"button\" class=\"MainButtonStyle\" [disabled]=\"\" (click)=\"ok()\">OK</button>\n                <button type=\"button\" class=\"MainButtonStyle\" (click)=\"cancel()\">Cancel</button>\n            </div>\n        </div>\n\n    </div>\n</div>\n"
            },] },
];
/**
 * @nocollapse
 */
XontVenturaMultiplePromptComponent.ctorParameters = function () { return []; };
XontVenturaMultiplePromptComponent.propDecorators = {
    'id': [{ type: _angular_core.Input },],
    'onSelect': [{ type: _angular_core.Output },],
    'onCancel': [{ type: _angular_core.Output },],
};

var XontVenturaMultiplePromptModule = (function () {
    function XontVenturaMultiplePromptModule() {
    }
    /**
     * @return {?}
     */
    XontVenturaMultiplePromptModule.forRoot = function () {
        return {
            ngModule: XontVenturaMultiplePromptModule
        };
    };
    return XontVenturaMultiplePromptModule;
}());
XontVenturaMultiplePromptModule.decorators = [
    { type: _angular_core.NgModule, args: [{
                imports: [
                    _angular_common.CommonModule, angular2Datatable.DataTableModule, _angular_forms.FormsModule
                ],
                declarations: [
                    XontVenturaMultiplePromptComponent,
                    DataFilterPipe
                ],
                exports: [
                    XontVenturaMultiplePromptComponent,
                    DataFilterPipe
                ]
            },] },
];
/**
 * @nocollapse
 */
XontVenturaMultiplePromptModule.ctorParameters = function () { return []; };

exports.XontVenturaMultiplePromptModule = XontVenturaMultiplePromptModule;
exports.XontVenturaMultiplePromptComponent = XontVenturaMultiplePromptComponent;
exports.DataFilterPipe = DataFilterPipe;

Object.defineProperty(exports, '__esModule', { value: true });

})));
