// xont-ventura-gridexport/index.js
import { Component, Input, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

var XontVenturaGridExportComponent = (function () {
    function XontVenturaGridExportComponent() {
        this.cellText = '';
        this.columnList = [];
    }
    /**
     * @param {?} tableID
     * @param {?} tableName
     * @return {?}
     */
    XontVenturaGridExportComponent.prototype.btnExcelExport_OnClick = function (tableID, tableName) {
        var /** @type {?} */ xml = this.htmlTableToXML(tableID.trim(), tableName.trim());
        if (this.isIE()) {
            if (window.navigator.msSaveBlob) {
                //var xml = this.htmlTableToXML('tblWarehouse');
                var /** @type {?} */ blob = new Blob([xml], {
                    type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'
                });
                navigator.msSaveBlob(blob, tableName + '.xls');
            }
        }
        else {
            var /** @type {?} */ uri = 'data:application/vnd.ms-excel,' + encodeURI(xml);
            var /** @type {?} */ downloadLink = document.createElement("a");
            downloadLink.href = uri;
            downloadLink.download = tableName + ".xls";
            document.body.appendChild(downloadLink);
            downloadLink.click();
            document.body.removeChild(downloadLink);
            //window.open("data:application/vnd.ms-excel," + encodeURI(xml));// can not set file name with this method.
        }
    };
    /**
     * @param {?} tableName
     * @return {?}
     */
    XontVenturaGridExportComponent.prototype.emitXmlHeader = function (tableName) {
        return '<?xml version="1.0"?>\n' +
            '<ss:Workbook xmlns:ss="urn:schemas-microsoft-com:office:spreadsheet">\n' +
            '<ss:Styles>\n' +
            '<ss:Style ss:ID="columnheaders">\n' +
            '<ss:Font ss:Bold="1" />\n' +
            '</ss:Style>\n' +
            '</ss:Styles>\n' +
            '<ss:Worksheet ss:Name="' + tableName + '">\n' +
            '<ss:Table >\n\n';
    };
    /**
     * @return {?}
     */
    XontVenturaGridExportComponent.prototype.emitXmlFooter = function () {
        return '\n</ss:Table>\n' +
            '</ss:Worksheet>\n' +
            '</ss:Workbook>\n';
    };
    /**
     * @return {?}
     */
    XontVenturaGridExportComponent.prototype.isIE = function () {
        var /** @type {?} */ ua = window.navigator.userAgent;
        var /** @type {?} */ msie = ua.indexOf('MSIE ');
        if (msie > 0) {
            // IE 10 or older 
            return true;
        }
        var /** @type {?} */ trident = ua.indexOf('Trident/');
        if (trident > 0) {
            // IE 11 
            return true;
        }
        var /** @type {?} */ edge = ua.indexOf('Edge/');
        if (edge > 0) {
            // Edge (IE 12+) 
            return true;
        }
        // other browser
        return false;
    };
    /**
     * @param {?} tableID
     * @param {?} tableName
     * @return {?}
     */
    XontVenturaGridExportComponent.prototype.htmlTableToXML = function (tableID, tableName) {
        //write column header row start
        var /** @type {?} */ headerRow = $('#' + tableID + ' > thead' + ' > tr > th');
        var /** @type {?} */ xml = '';
        xml += this.emitXmlHeader(tableName);
        for (var /** @type {?} */ i = 0; i < headerRow.length; i++) {
            if ($(headerRow[i]).width() != -1) {
                xml += '<ss:Column ss:AutoFitWidth="0" ss:Width="' + $(headerRow[i]).width() + '"/>\n';
            }
            else {
                xml += '<ss:Column ss:AutoFitWidth="0" />\n'; //no column text
            }
        }
        xml += '<ss:Row>\n';
        for (var /** @type {?} */ i = 0; i < headerRow.length; i++) {
            var /** @type {?} */ aaa = headerRow[i].getElementsByTagName('a');
            var /** @type {?} */ headerText;
            if (aaa.length > 0) {
                headerText = headerRow[i].getElementsByTagName('a')[0].textContent.toString().trim();
            }
            else {
                //console.log(headerRow[i]);
                headerText = headerRow[i].textContent.toString().trim();
            }
            xml += '  <ss:Cell ss:StyleID="columnheaders">\n';
            xml += '    <ss:Data ss:Type="String">' + headerText.trim() + '</ss:Data>\n';
            xml += '  </ss:Cell>\n';
        }
        xml += '</ss:Row>\n';
        // write column header row end
        // write table body start
        var /** @type {?} */ bodyRows = $('#' + tableID + ' > tbody' + ' > tr ');
        for (var /** @type {?} */ i = 0; i < bodyRows.length; i++) {
            xml += '<ss:Row>\n';
            var /** @type {?} */ cellList = bodyRows[i].getElementsByTagName('td');
            for (var /** @type {?} */ j = 0; j < cellList.length; j++) {
                this.cellText = '';
                this.getDeepText($(cellList[j]));
                xml += '  <ss:Cell>\n';
                xml += '    <ss:Data ss:Type="';
                if (this.cellText.trim() != '') {
                    if (isNaN(this.cellText.trim())) {
                        xml += 'String';
                    }
                    else {
                        xml += 'Number';
                    }
                }
                else {
                    xml += 'String';
                }
                xml += '">';
                xml += this.cellText.trim() + '</ss:Data>\n';
                xml += '  </ss:Cell>\n';
            }
            xml += '</ss:Row>\n';
        }
        //write table body end
        xml += this.emitXmlFooter();
        return xml;
    };
    /**
     * @param {?} DOMObject
     * @return {?}
     */
    XontVenturaGridExportComponent.prototype.getDeepText = function (DOMObject) {
        if (DOMObject[0].nodeName == "SELECT") {
            var /** @type {?} */ x = DOMObject[0].selectedIndex;
            var /** @type {?} */ y = DOMObject[0].options;
            this.cellText += y[x].text.toString().trim();
        }
        else if (DOMObject[0].children.length > 0) {
            for (var /** @type {?} */ i = 0; i < DOMObject[0].children.length; i++) {
                this.getDeepText($(DOMObject[0].children[i]));
            }
        }
        else {
            if (DOMObject[0].nodeName == "INPUT") {
                if (DOMObject[0].type == "radio") {
                    if (DOMObject[0].checked) {
                        this.cellText += "Selected";
                    }
                    else {
                        this.cellText += "Unselected";
                    }
                }
                else if (DOMObject[0].type == "checkbox") {
                    if (DOMObject[0].checked) {
                        this.cellText += "Checked";
                    }
                    else {
                        this.cellText += "Unchecked";
                    }
                }
                else {
                    this.cellText += DOMObject[0].value.toString().trim();
                }
            }
            else {
                this.cellText += DOMObject[0].innerText.toString().trim();
            }
        }
    };
    /**
     * @param {?} tableID
     * @param {?} tableName
     * @return {?}
     */
    XontVenturaGridExportComponent.prototype.btnPDFExport_OnClick = function (tableID, tableName) {
        var /** @type {?} */ doc = new jsPDF('p', 'pt', 'a4');
        var /** @type {?} */ options = {
            theme: 'grid',
            styles: { overflow: 'linebreak', fontSize: 8, lineColor: '#AAAAAA' },
            headerStyles: { fillColor: "#006699" }
        };
        var /** @type {?} */ data = this.getJSONColumnRows(tableID);
        doc.autoTable(data.Columns, data.Rows, options);
        doc.save(tableName + ".pdf");
    };
    /**
     * @param {?} tableID
     * @return {?}
     */
    XontVenturaGridExportComponent.prototype.getJSONColumnRows = function (tableID) {
        var /** @type {?} */ headerRow = $('#' + tableID + ' > thead' + ' > tr > th');
        var /** @type {?} */ cols = [];
        var /** @type {?} */ rows = [];
        this.columnList = [];
        for (var /** @type {?} */ i = 0; i < headerRow.length; i++) {
            var /** @type {?} */ aaa = headerRow[i].getElementsByTagName('a');
            var /** @type {?} */ headerText;
            if (aaa.length > 0) {
                headerText = headerRow[i].getElementsByTagName('a')[0].textContent.toString().trim();
            }
            else {
                headerText = headerRow[i].textContent.toString().trim();
            }
            cols.push({ title: headerText.trim(), dataKey: "col" + i });
            this.columnList.push("col" + i);
        }
        var /** @type {?} */ bodyRows = $('#' + tableID + ' > tbody' + ' > tr ');
        for (var /** @type {?} */ i = 0; i < bodyRows.length; i++) {
            var /** @type {?} */ cellList = bodyRows[i].getElementsByTagName('td');
            var /** @type {?} */ row = {};
            for (var /** @type {?} */ j = 0; j < cellList.length; j++) {
                this.cellText = '';
                this.getDeepText($(cellList[j]));
                row[this.columnList[j]] = this.cellText.trim();
            }
            rows.push(row);
        }
        return { Columns: cols, Rows: rows };
    };
    /**
     * @param {?} tableID
     * @param {?} tableName
     * @return {?}
     */
    XontVenturaGridExportComponent.prototype.btnWordExport_OnClick = function (tableID, tableName) {
        var /** @type {?} */ link, /** @type {?} */ blob, /** @type {?} */ url;
        blob = new Blob(['\ufeff', this.getHTMLTable(tableID)], {
            type: 'application/msword'
        });
        if (this.isIE()) {
            if (window.navigator.msSaveBlob) {
                navigator.msSaveOrOpenBlob(blob, tableName + '.doc');
            }
        }
        else {
            url = URL.createObjectURL(blob);
            link = document.createElement('A');
            link.href = url;
            link.download = tableName;
            document.body.appendChild(link);
            link.click();
            document.body.removeChild(link);
        }
    };
    /**
     * @param {?} tableID
     * @return {?}
     */
    XontVenturaGridExportComponent.prototype.getHTMLTable = function (tableID) {
        var /** @type {?} */ css = '<style>' +
            '@page WordSection1{size: 841.95pt 595.35pt;mso-page-orientation: landscape}' +
            'div.WordSection1 {page: WordSection1;}' +
            'table{border-collapse:collapse;font-family:Trebuchet MS;}' +
            'td{padding: 5px;border: 1px solid #AAAAAA}' +
            'th{background-color: #006699;color: #ffffff;padding: 5px;border: 1px solid #AAAAAA;}' +
            '</style>';
        //write column header row start
        var /** @type {?} */ headerRow = $('#' + tableID + ' > thead' + ' > tr > th');
        var /** @type {?} */ html = css + '<div class="WordSection1"><table><tr>';
        for (var /** @type {?} */ i = 0; i < headerRow.length; i++) {
            var /** @type {?} */ aaa = headerRow[i].getElementsByTagName('a');
            var /** @type {?} */ headerText;
            if (aaa.length > 0) {
                headerText = headerRow[i].getElementsByTagName('a')[0].textContent.toString().trim();
            }
            else {
                headerText = headerRow[i].textContent.toString().trim();
            }
            html += '    <th>' + headerText.trim() + '</th>';
        }
        html += '</tr>';
        // write column header row end
        // write table body start
        var /** @type {?} */ bodyRows = $('#' + tableID + ' > tbody' + ' > tr ');
        for (var /** @type {?} */ i = 0; i < bodyRows.length; i++) {
            html += '<tr>';
            var /** @type {?} */ cellList = bodyRows[i].getElementsByTagName('td');
            for (var /** @type {?} */ j = 0; j < cellList.length; j++) {
                html += '<td>';
                this.cellText = '';
                this.getDeepText($(cellList[j]));
                html += this.cellText.trim();
                html += '</td>';
            }
            html += '</tr>';
        }
        //write table body end
        html += '</table></div>';
        return html;
    };
    return XontVenturaGridExportComponent;
}());
XontVenturaGridExportComponent.decorators = [
    { type: Component, args: [{
                selector: 'xont-ventura-gridexport',
                template: "<span class='Linkboldtext'>{{gridName}}&nbsp;&nbsp;</span>\n<button title=\"Excel\" class=\"btn btn-xs\" style=\"color: white;background-color:#20744b;padding: 0px 4px;font-size: 14px;\" (click)=\"btnExcelExport_OnClick(gridId,gridName)\"><i class=\"fa fa-file-excel-o\" aria-hidden=\"true\"></i></button>\n<button title=\"PDF\" class=\"btn btn-xs\" style=\"color: white;background-color:#bb0706;padding: 0px 4px;font-size: 14px;\" (click)=\"btnPDFExport_OnClick(gridId,gridName)\"><i class=\"fa fa-file-pdf-o\" aria-hidden=\"true\"></i></button>\n<button title=\"Word\" class=\"btn btn-xs\" style=\"color: white;background-color:#2d5fa2;padding: 0px 4px;font-size: 14px;\" (click)=\"btnWordExport_OnClick(gridId,gridName)\"><i class=\"fa fa-file-word-o\" aria-hidden=\"true\"></i></button>"
            },] },
];
/**
 * @nocollapse
 */
XontVenturaGridExportComponent.ctorParameters = function () { return []; };
XontVenturaGridExportComponent.propDecorators = {
    'id': [{ type: Input },],
    'gridName': [{ type: Input },],
    'gridId': [{ type: Input },],
};

var XontVenturaGridExportModule = (function () {
    function XontVenturaGridExportModule() {
    }
    /**
     * @return {?}
     */
    XontVenturaGridExportModule.forRoot = function () {
        return {
            ngModule: XontVenturaGridExportModule
        };
    };
    return XontVenturaGridExportModule;
}());
XontVenturaGridExportModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    XontVenturaGridExportComponent,
                ],
                exports: [
                    XontVenturaGridExportComponent,
                ]
            },] },
];
/**
 * @nocollapse
 */
XontVenturaGridExportModule.ctorParameters = function () { return []; };

export { XontVenturaGridExportModule, XontVenturaGridExportComponent };
