// xont-ventura-single-prompt/index.js
import { Component, EventEmitter, Injectable, Input, NgModule, Output, Pipe, ViewChild } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DataTableModule } from 'angular2-datatable';
import { FormsModule } from '@angular/forms';
import { CommonService } from 'xont-ventura-services';
import { XontVenturaMessagePromptModule } from 'xont-ventura-message-prompt';
import { BusyModule } from 'angular2-busy';
import { filter } from 'lodash';
import * as _ from 'lodash';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/catch';
import { Observable } from 'rxjs/Rx';
import { Headers, Http } from '@angular/http';

var DataFilterPipe = (function () {
    function DataFilterPipe() {
    }
    /**
     * @param {?} items
     * @param {?} filter
     * @return {?}
     */
    DataFilterPipe.prototype.transform = function (items, filter$$1) {
        if (!items) {
            return [];
        }
        if (!filter$$1) {
            return items;
        }
        //console.log(filter);
        for (var /** @type {?} */ i = 0; i < filter$$1.length; i++) {
            if (filter$$1[i].searchString.trim() != '') {
                items = filter(items, function (row) { return row[filter$$1[i].field].toString().toUpperCase().indexOf(filter$$1[i].searchString.trim().toUpperCase()) > -1; });
            }
        }
        return items;
    };
    return DataFilterPipe;
}());
DataFilterPipe.decorators = [
    { type: Pipe, args: [{
                name: 'dataFilter'
            },] },
];
/**
 * @nocollapse
 */
DataFilterPipe.ctorParameters = function () { return []; };

var XontVenturaSinglePromptService = (function () {
    /**
     * @param {?} http
     * @param {?} commonService
     */
    function XontVenturaSinglePromptService(http, commonService) {
        this.http = http;
        this.commonService = commonService;
        this.taskCode = '';
    }
    /**
     * @return {?}
     */
    XontVenturaSinglePromptService.prototype.getAPIPrefix = function () {
        return this.commonService.getAPIPrefix(this.taskCode);
    };
    /**
     * @param {?} taskCode
     * @param {?} ListFromID
     * @param {?} ExternalParameters
     * @return {?}
     */
    XontVenturaSinglePromptService.prototype.GetPredefinedPrompt = function (taskCode, ListFromID, ExternalParameters) {
        this.taskCode = taskCode;
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.getAPIPrefix() + "/api/Prompt/GetPredefinedPrompt", { listFromID: ListFromID, externalParameters: ExternalParameters }, { headers: headers })
            .map(function (response) { return response.json(); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    /**
     * @param {?} data
     * @param {?} taskCode
     * @return {?}
     */
    XontVenturaSinglePromptService.prototype.GetDataForMasterClassificationPopup = function (data, taskCode) {
        this.taskCode = taskCode;
        var /** @type {?} */ headers = new Headers();
        headers.append('Content-Type', 'application/json');
        return this.http.post(this.getAPIPrefix() + "/api/Prompt/GetDataForMasterClassificationPopup", data, { headers: headers })
            .map(function (response) { return response.json(); })
            .catch(function (error) { return Observable.throw(error.json()); });
    };
    return XontVenturaSinglePromptService;
}());
//public handleError(error: any) {
//    this.componentMethodCallSource.next(error);
//    return Observable.throw(error);
//}
XontVenturaSinglePromptService.decorators = [
    { type: Injectable },
];
/**
 * @nocollapse
 */
XontVenturaSinglePromptService.ctorParameters = function () { return [
    { type: Http, },
    { type: CommonService, },
]; };

var XontVenturaSinglePromptComponent = (function () {
    /**
     * @param {?} promptService
     */
    function XontVenturaSinglePromptComponent(promptService) {
        this.promptService = promptService;
        this.rowsOnPage = 10;
        this.sortBy = "code";
        this.sortOrder = "asc";
        this.columnDefs = [];
        this.title = "";
        this.promptFilter = [];
        this.selectedItem = null;
        this.taskCode = "";
        this.onSelect = new EventEmitter();
        this.onCancel = new EventEmitter();
        this.errorPromptID = this.id + "Error";
        this.sortByWordLength = function (a) {
            return a.city.length;
        };
    }
    /**
     * @param {?} num
     * @return {?}
     */
    XontVenturaSinglePromptComponent.prototype.toInt = function (num) {
        return +num;
    };
    /**
     * @param {?} item
     * @return {?}
     */
    XontVenturaSinglePromptComponent.prototype.remove = function (item) {
        var /** @type {?} */ index = this.data.indexOf(item);
        if (index > -1) {
            this.data.splice(index, 1);
        }
    };
    /**
     * @param {?} item
     * @return {?}
     */
    XontVenturaSinglePromptComponent.prototype.selectOne = function (item) {
        this.selectedRowNum = item._indexNumber;
        this.selectedItem = item;
        //this.onSelect.emit(item);
    };
    /**
     * @param {?} gridOption
     * @return {?}
     */
    XontVenturaSinglePromptComponent.prototype.show = function (gridOption) {
        this.selectedRowNum = -1;
        this.selectedItem = null;
        var /** @type {?} */ num = 0;
        gridOption.dataSet.forEach(function (obj) {
            obj._indexNumber = num;
            num++;
        });
        this.data = gridOption.dataSet;
        this.backupData = gridOption.dataSet;
        this.columnDefs = gridOption.columnDefs;
        this.promptFilter = gridOption.columnDefs;
        for (var /** @type {?} */ i = 0; i < this.promptFilter.length; i++) {
            this.promptFilter[i]['searchString'] = '';
        }
        //console.log(this.promptFilter);
        this.title = gridOption.title;
        $('#' + this.id).modal({ backdrop: "static" });
    };
    /**
     * @return {?}
     */
    XontVenturaSinglePromptComponent.prototype.cancel = function () {
        $('#' + this.id).modal('hide');
        this.onCancel.emit();
    };
    /**
     * @return {?}
     */
    XontVenturaSinglePromptComponent.prototype.ok = function () {
        if (this.selectedItem != null) {
            this.onSelect.emit(this.selectedItem);
        }
        $('#' + this.id).modal('hide');
    };
    /**
     * @param {?} event
     * @return {?}
     */
    XontVenturaSinglePromptComponent.prototype.onKey = function (event) {
        this.data = new DataFilterPipe().transform(this.backupData, this.promptFilter);
    };
    /**
     * @param {?} promptTitle
     * @param {?} listFromID
     * @param {?} externalParameters
     * @return {?}
     */
    XontVenturaSinglePromptComponent.prototype.ShowPredefinedPrompt1 = function (promptTitle, listFromID, externalParameters) {
        var _this = this;
        this.mdpBusy = this.promptService.GetPredefinedPrompt(this.taskCode, listFromID, externalParameters)
            .subscribe(function (data) {
            //console.log(data);
            var /** @type {?} */ columnHeaders = [];
            if (data.length > 0) {
                var /** @type {?} */ columnNames = Object.keys(data[0]);
                for (var /** @type {?} */ i = 0; i < columnNames.length; i++) {
                    columnHeaders.push({ headerName: columnNames[i], field: columnNames[i], filter: false });
                }
            }
            var /** @type {?} */ gridOption = {
                title: promptTitle,
                columnDefs: columnHeaders,
                dataSet: data
            };
            _this.show(gridOption);
        }, function (err) {
            console.log(err);
            _this.singlePromptErrorPrompt.show(err, _this.taskCode);
        });
    };
    /**
     * @param {?} promptTitle
     * @param {?} listFromID
     * @param {?} externalParameters
     * @param {?} columnHeadings
     * @return {?}
     */
    XontVenturaSinglePromptComponent.prototype.ShowPredefinedPrompt2 = function (promptTitle, listFromID, externalParameters, columnHeadings) {
        var _this = this;
        this.mdpBusy = this.promptService.GetPredefinedPrompt(this.taskCode, listFromID, externalParameters)
            .subscribe(function (data) {
            for (var /** @type {?} */ i = 0; i < data.length; i++) {
                var /** @type {?} */ columnNames = Object.keys(data[i]);
                for (var /** @type {?} */ j = 0; j < columnNames.length; j++) {
                    data[i][columnHeadings[j]] = data[i][columnNames[j]];
                    delete data[i][columnNames[j]];
                }
            }
            var /** @type {?} */ columnHeaders = [];
            for (var /** @type {?} */ i = 0; i < columnHeadings.length; i++) {
                columnHeaders.push({ headerName: columnHeadings[i], field: columnHeadings[i], filter: false });
            }
            var /** @type {?} */ gridOption = {
                title: promptTitle,
                columnDefs: columnHeaders,
                dataSet: data
            };
            _this.show(gridOption);
        }, function (err) {
            console.log(err);
            _this.singlePromptErrorPrompt.show(err, _this.taskCode);
        });
    };
    /**
     * @param {?} promptTitle
     * @param {?} masterGroupType
     * @return {?}
     */
    XontVenturaSinglePromptComponent.prototype.ShowMasterClassificationPopup1 = function (promptTitle, masterGroupType) {
        var _this = this;
        var /** @type {?} */ data = {
            MethodNumber: "1",
            MasterGroupType: masterGroupType
        };
        this.mdpBusy = this.promptService.GetDataForMasterClassificationPopup(data, this.taskCode)
            .subscribe(function (data) {
            var /** @type {?} */ columnHeaders = [];
            if (data.length > 0) {
                var /** @type {?} */ columnNames = Object.keys(data[0]);
                for (var /** @type {?} */ i = 0; i < columnNames.length; i++) {
                    columnHeaders.push({ headerName: columnNames[i], field: columnNames[i], filter: true });
                }
            }
            var /** @type {?} */ gridOption = {
                title: promptTitle,
                columnDefs: columnHeaders,
                dataSet: data
            };
            _this.show(gridOption);
        }, function (err) {
            console.log(err);
            _this.singlePromptErrorPrompt.show(err, _this.taskCode);
        });
    };
    /**
     * @param {?} promptTitle
     * @param {?} masterGroupType
     * @param {?} promptParameters
     * @return {?}
     */
    XontVenturaSinglePromptComponent.prototype.ShowMasterClassificationPopup2 = function (promptTitle, masterGroupType, promptParameters) {
        var _this = this;
        var /** @type {?} */ data = {
            MethodNumber: "2",
            MasterGroupType: masterGroupType,
            PromptParameters: promptParameters
        };
        this.mdpBusy = this.promptService.GetDataForMasterClassificationPopup(data, this.taskCode)
            .subscribe(function (data) {
            var /** @type {?} */ columnHeaders = [];
            if (data.length > 0) {
                var /** @type {?} */ columnNames = Object.keys(data[0]);
                for (var /** @type {?} */ i = 0; i < columnNames.length; i++) {
                    columnHeaders.push({ headerName: columnNames[i], field: columnNames[i], filter: true });
                }
            }
            var /** @type {?} */ gridOption = {
                title: promptTitle,
                columnDefs: columnHeaders,
                dataSet: data
            };
            _this.show(gridOption);
        }, function (err) {
            console.log(err);
            _this.singlePromptErrorPrompt.show(err, _this.taskCode);
        });
    };
    /**
     * @param {?} promptTitle
     * @param {?} masterGroupType
     * @param {?} promptParameters
     * @param {?} columnHeadings
     * @return {?}
     */
    XontVenturaSinglePromptComponent.prototype.ShowMasterClassificationPopup3 = function (promptTitle, masterGroupType, promptParameters, columnHeadings) {
        var _this = this;
        var /** @type {?} */ data = {
            MethodNumber: "3",
            MasterGroupType: masterGroupType,
            PromptParameters: promptParameters,
            ColumnHeadings: columnHeadings
        };
        this.mdpBusy = this.promptService.GetDataForMasterClassificationPopup(data, this.taskCode)
            .subscribe(function (data) {
            var /** @type {?} */ columnHeaders = [];
            if (data.length > 0) {
                var /** @type {?} */ columnNames = Object.keys(data[0]);
                for (var /** @type {?} */ i = 0; i < columnNames.length; i++) {
                    columnHeaders.push({ headerName: columnNames[i], field: columnNames[i], filter: true });
                }
            }
            var /** @type {?} */ gridOption = {
                title: promptTitle,
                columnDefs: columnHeaders,
                dataSet: data
            };
            _this.show(gridOption);
        }, function (err) {
            console.log(err);
            _this.singlePromptErrorPrompt.show(err, _this.taskCode);
        });
    };
    /**
     * @param {?} promptTitle
     * @param {?} masterGroupType
     * @param {?} promptParameters
     * @param {?} columnHeadings
     * @param {?} boundFieldNames
     * @return {?}
     */
    XontVenturaSinglePromptComponent.prototype.ShowMasterClassificationPopup4 = function (promptTitle, masterGroupType, promptParameters, columnHeadings, boundFieldNames) {
        this.ShowMasterClassificationPopup5(promptTitle, masterGroupType, promptParameters, "Active", columnHeadings, boundFieldNames);
    };
    /**
     * @param {?} promptTitle
     * @param {?} masterGroupType
     * @param {?} promptParameters
     * @param {?} statusFlag
     * @param {?} columnHeadings
     * @param {?} boundFieldNames
     * @return {?}
     */
    XontVenturaSinglePromptComponent.prototype.ShowMasterClassificationPopup5 = function (promptTitle, masterGroupType, promptParameters, statusFlag, columnHeadings, boundFieldNames) {
        this.ShowMasterClassificationPopup6(promptTitle, masterGroupType, promptParameters, statusFlag, columnHeadings, boundFieldNames, null);
    };
    /**
     * @param {?} promptTitle
     * @param {?} masterGroupType
     * @param {?} promptParameters
     * @param {?} statusFlag
     * @param {?} columnHeadings
     * @param {?} boundFieldNames
     * @param {?} filterExpression
     * @return {?}
     */
    XontVenturaSinglePromptComponent.prototype.ShowMasterClassificationPopup6 = function (promptTitle, masterGroupType, promptParameters, statusFlag, columnHeadings, boundFieldNames, filterExpression) {
        var _this = this;
        var /** @type {?} */ data = {
            MethodNumber: "4",
            MasterGroupType: masterGroupType,
            PromptParameters: promptParameters,
            StatusFlag: statusFlag,
            BoundFieldNames: boundFieldNames,
            FilterExpression: filterExpression,
            ColumnHeadings: columnHeadings
        };
        this.mdpBusy = this.promptService.GetDataForMasterClassificationPopup(data, this.taskCode)
            .subscribe(function (data) {
            var /** @type {?} */ columnHeaders = [];
            if (data.length > 0) {
                var /** @type {?} */ columnNames = Object.keys(data[0]);
                for (var /** @type {?} */ i = 0; i < columnNames.length; i++) {
                    columnHeaders.push({ headerName: columnNames[i], field: columnNames[i], filter: true });
                }
            }
            var /** @type {?} */ gridOption = {
                title: promptTitle,
                columnDefs: columnHeaders,
                dataSet: data
            };
            _this.show(gridOption);
        }, function (err) {
            console.log(err);
            _this.singlePromptErrorPrompt.show(err, _this.taskCode);
        });
    };
    return XontVenturaSinglePromptComponent;
}());
XontVenturaSinglePromptComponent.decorators = [
    { type: Component, args: [{
                selector: 'xont-ventura-single-prompt',
                styles: ['.selectedRow {background-color: rgb(102, 153, 153);}'],
                template: "  \n<div [ngBusy]=\"mdpBusy\"></div>\n<xont-ventura-message-prompt [id]=\"errorPromptID\" #singlePromptErrorPrompt></xont-ventura-message-prompt>\n\n\n  <div id=\"{{id}}\" class=\"modal fade\" role=\"dialog\">\n        <div class=\"modal-dialog \">\n\n            <div class=\"modal-content\">\n                <div class=\"modal-header\">\n                    <h5 class=\"modal-title\">{{title}}</h5>\n                </div>\n                <div class=\"modal-body\" >\n                               \n                                <span class=\"Labelstyle\" [hidden]=\"backupData?.length != 0\">No Data found to display</span>\n                                <div  class=\"PromptDataTableContainer\">\n                                    \n\n                                    <table class=\"PromptDataTableStyle\" [hidden]=\"backupData?.length == 0\" [mfData]=\"data | dataFilter\" #mf=\"mfDataTable\"\n                                           [mfRowsOnPage]=\"rowsOnPage\" [(mfSortBy)]=\"sortBy\" [(mfSortOrder)]=\"sortOrder\">\n                                        <thead>\n                                            <tr>\n                                                <th style=\"width: 10%\"></th>\n\n                                                <th *ngFor=\"let column of columnDefs \" >\n                                                    <mfDefaultSorter by=\"{{column.field}}\">{{column.headerName}}</mfDefaultSorter>\n                                                </th>\n                                            </tr>\n                                            <tr>\n                                                <td align=\"center\"><i class=\"fa fa-search\" aria-hidden=\"true\"></i></td>\n                                                <td *ngFor=\"let column of promptFilter \">\n                                                    <input style=\"width:100%\" *ngIf=\"column.filter\" [(ngModel)]=\"column.searchString\" type=\"text\" (keyup)=\"onKey($event)\"  class=\"Textboxstyle\" >\n                                                </td>\n                                                \n                                            </tr>\n                                        \n                                        </thead>\n                                        <tbody>\n                                            <tr *ngFor=\"let item of mf.data\"  [ngClass]=\"{'selectedRow': item._indexNumber==selectedRowNum}\">\n                                                <td align=\"center\">\n                                                 <input type=\"radio\" name=\"rd\" value=\"item._indexNumber\" (click)=\"selectOne(item)\" [checked]=\"item._indexNumber==selectedRowNum\">\n                                                \n                                                </td>\n\n                                                <td *ngFor=\"let column of columnDefs\">{{item[column.field]}}</td>\n                                            \n                                            </tr>\n                                        </tbody>\n                                        <tfoot>\n                                            <tr>\n                                                <td [colSpan]=\"columnDefs.length+1\">\n                                                    <mfBootstrapPaginator ></mfBootstrapPaginator>\n                                                </td>\n                                            </tr>\n                                        </tfoot>\n                                    </table>\n                                \n                                </div>\n                        </div>\n                        <div class=\"modal-footer\">\n                            <button type=\"button\" class=\"MainButtonStyle\" (click)=\"ok()\" >OK</button>\n                            <button type=\"button\" class=\"MainButtonStyle\" (click)=\"cancel()\" >Cancel</button>\n                        </div>\n            </div>\n\n        </div>\n    </div>\n"
            },] },
];
/**
 * @nocollapse
 */
XontVenturaSinglePromptComponent.ctorParameters = function () { return [
    { type: XontVenturaSinglePromptService, },
]; };
XontVenturaSinglePromptComponent.propDecorators = {
    'id': [{ type: Input },],
    'taskCode': [{ type: Input },],
    'onSelect': [{ type: Output },],
    'onCancel': [{ type: Output },],
    'singlePromptErrorPrompt': [{ type: ViewChild, args: ['singlePromptErrorPrompt',] },],
};

var XontVenturaSinglePromptModule = (function () {
    function XontVenturaSinglePromptModule() {
    }
    /**
     * @return {?}
     */
    XontVenturaSinglePromptModule.forRoot = function () {
        return {
            ngModule: XontVenturaSinglePromptModule
        };
    };
    return XontVenturaSinglePromptModule;
}());
XontVenturaSinglePromptModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule, DataTableModule, FormsModule, BusyModule, XontVenturaMessagePromptModule
                ],
                declarations: [
                    XontVenturaSinglePromptComponent,
                    DataFilterPipe
                ],
                exports: [
                    XontVenturaSinglePromptComponent,
                    DataFilterPipe
                ],
                providers: [CommonService, XontVenturaSinglePromptService]
            },] },
];
/**
 * @nocollapse
 */
XontVenturaSinglePromptModule.ctorParameters = function () { return []; };

export { XontVenturaSinglePromptModule, XontVenturaSinglePromptComponent, DataFilterPipe };
