// xont-ventura-gridloader/index.js
import { Component, EventEmitter, NgModule, Output } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';

var XontVenturaGridLoaderComponent = (function () {
    function XontVenturaGridLoaderComponent() {
        this.RowStart = 0;
        this.RowEnd = 0;
        this.CurrentPage = 1;
        this.TotalPage = 0;
        this.TaskCode = '';
        this.MaxLen = 1;
        this.LastCurrentPage = 1;
        this.onChange = new EventEmitter();
    }
    /**
     * @param {?} taskCode
     * @return {?}
     */
    XontVenturaGridLoaderComponent.prototype.init = function (taskCode) {
        this.TaskCode = taskCode;
    };
    /**
     * @return {?}
     */
    XontVenturaGridLoaderComponent.prototype.getPageSize = function () {
        var /** @type {?} */ data = JSON.parse(localStorage.getItem(this.TaskCode + '_MasterControlData'));
        if (data.AllowPaging == "1") {
            return data.PageSize;
        }
        else {
            return data.ExtendedPageSize;
        }
    };
    /**
     * @return {?}
     */
    XontVenturaGridLoaderComponent.prototype.getRowStart = function () {
        if (this.RowStart < 1) {
            this.RowStart = 1;
        }
        return this.RowStart;
    };
    /**
     * @return {?}
     */
    XontVenturaGridLoaderComponent.prototype.getRowEnd = function () {
        if (this.RowEnd < 1) {
            if (this.RowStart < 1) {
                this.RowStart = 1;
            }
            var /** @type {?} */ loadSize = this.getLoadSize();
            this.RowEnd = this.RowStart + loadSize - 1;
        }
        return this.RowEnd;
    };
    /**
     * @param {?} rowTotal
     * @return {?}
     */
    XontVenturaGridLoaderComponent.prototype.setRowCount = function (rowTotal) {
        var /** @type {?} */ loadSize = this.getLoadSize();
        this.TotalPage = Math.ceil(rowTotal / loadSize);
        this.MaxLen = this.TotalPage.toString().length;
    };
    /**
     * @return {?}
     */
    XontVenturaGridLoaderComponent.prototype.getLoadSize = function () {
        var /** @type {?} */ data = JSON.parse(localStorage.getItem(this.TaskCode + '_MasterControlData'));
        if (data.AllowPaging == "1") {
            return data.LoadSize;
        }
        else {
            return data.ExtendedPageSize;
        }
    };
    /**
     * @return {?}
     */
    XontVenturaGridLoaderComponent.prototype.emit = function () {
        var /** @type {?} */ loadSize = this.getLoadSize();
        this.RowStart = ((this.CurrentPage - 1) * loadSize) + 1;
        this.RowEnd = this.RowStart + loadSize - 1;
        this.onChange.emit();
    };
    /**
     * @return {?}
     */
    XontVenturaGridLoaderComponent.prototype.btnFirst_OnClick = function () {
        if (this.CurrentPage > 1) {
            this.CurrentPage = 1;
            this.LastCurrentPage = 1;
            this.emit();
        }
    };
    /**
     * @return {?}
     */
    XontVenturaGridLoaderComponent.prototype.btnPrev_OnClick = function () {
        if (this.CurrentPage > 1) {
            this.CurrentPage = this.CurrentPage - 1;
            this.LastCurrentPage = this.CurrentPage;
            this.emit();
        }
    };
    /**
     * @return {?}
     */
    XontVenturaGridLoaderComponent.prototype.btnNext_OnClick = function () {
        if (this.CurrentPage < this.TotalPage) {
            this.CurrentPage++;
            this.LastCurrentPage = this.CurrentPage;
            this.emit();
        }
    };
    /**
     * @return {?}
     */
    XontVenturaGridLoaderComponent.prototype.btnLast_OnClick = function () {
        if (this.CurrentPage < this.TotalPage) {
            this.CurrentPage = this.TotalPage;
            this.LastCurrentPage = this.TotalPage;
            this.emit();
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    XontVenturaGridLoaderComponent.prototype.currentPage_OnKeypress = function (event) {
        if (event.keyCode >= 48 && event.keyCode <= 57) {
            return true;
        }
        else {
            return false;
        }
    };
    /**
     * @return {?}
     */
    XontVenturaGridLoaderComponent.prototype.currentPage_OnBlur = function () {
        //console.log(this.CurrentPage);
        if (this.CurrentPage != this.LastCurrentPage) {
            if (this.CurrentPage > 0 && this.CurrentPage <= this.TotalPage) {
                this.LastCurrentPage = this.CurrentPage;
                this.emit();
            }
            else {
                this.CurrentPage = this.LastCurrentPage;
            }
        }
    };
    /**
     * @param {?} num
     * @return {?}
     */
    XontVenturaGridLoaderComponent.prototype.setCurrentPage = function (num) {
        this.CurrentPage = num;
    };
    return XontVenturaGridLoaderComponent;
}());
XontVenturaGridLoaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'xont-ventura-gridloader',
                template: "\n    <button type=\"button\" name=\"btnFirst\" id=\"btnFirst\" (click)=\"btnFirst_OnClick()\" class=\"skipfwrdleft\"></button>\n    <button type=\"button\" name=\"btnPrev\" id=\"btnPrev\" (click)=\"btnPrev_OnClick()\" class=\"fwrdleft\"></button>\n    <input name=\"txtCurrentPage\" type=\"text\" [(ngModel)]=\"CurrentPage\" (keypress)=\"currentPage_OnKeypress($event)\" (blur)=\"currentPage_OnBlur()\" [maxlength]=\"MaxLen\"  class=\"Textboxstyle\" id=\"txtCurrentPage\"  style=\"width:40px;height:22px\" ><span class=\"Captionstyle\">&nbsp;OF</span> \n    <input name=\"txtTotalPages\" type=\"text\" [(ngModel)]=\"TotalPage\" disabled id=\"txtTotalPages\" tabindex=\"-1\" class=\"Textboxstyle\" style=\"width:40px;height:22px\">\n    <button type=\"button\" name=\"btnNext\" id=\"btnNext\" (click)=\"btnNext_OnClick()\" class=\"fwrdright\"></button>\n    <button type=\"button\" name=\"btnLast\" id=\"btnLast\" (click)=\"btnLast_OnClick()\" class=\"skipfwrdright\"></button>\n  "
            },] },
];
/**
 * @nocollapse
 */
XontVenturaGridLoaderComponent.ctorParameters = function () { return []; };
XontVenturaGridLoaderComponent.propDecorators = {
    'onChange': [{ type: Output },],
};

var XontVenturaGridLoaderModule = (function () {
    function XontVenturaGridLoaderModule() {
    }
    /**
     * @return {?}
     */
    XontVenturaGridLoaderModule.forRoot = function () {
        return {
            ngModule: XontVenturaGridLoaderModule
        };
    };
    return XontVenturaGridLoaderModule;
}());
XontVenturaGridLoaderModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule, FormsModule
                ],
                declarations: [
                    XontVenturaGridLoaderComponent
                ],
                exports: [
                    XontVenturaGridLoaderComponent
                ]
            },] },
];
/**
 * @nocollapse
 */
XontVenturaGridLoaderModule.ctorParameters = function () { return []; };

export { XontVenturaGridLoaderModule, XontVenturaGridLoaderComponent };
